// @jup-ag/react-hook/dist/index.mjs
// src/index.tsx
import {
  createJupiterApiClient
} from "@jup-ag/api";
import {
  JUPITER_FEE_OWNER,
  JUPITER_V6_ERRORS,
  LAMPORTS_PER_SIGNATURE,
  MARKETS_URL,
  Owner,
  SwapMode,
  TOKEN_LIST_URL,
  calculateTransactionDepositAndFeeV2,
  executeTransaction,
  getReferralTokenAccount,
  getTransactionBlockhash
} from "@jup-ag/common";
import { TransactionError, checkTxSize } from "@mercurial-finance/optimist";
import { TOKEN_PROGRAM_ID as TOKEN_PROGRAM_ID2, NATIVE_MINT as WRAPPED_SOL_MINT } from "@solana/spl-token";
import { PublicKey as PublicKey3, Transaction, VersionedTransaction } from "@solana/web3.js";
import JSBI2 from "jsbi";
import promiseRetry from "promise-retry";
import React, { createContext, useCallback, useContext, useEffect as useEffect2, useMemo, useRef, useState as useState2 } from "react";
import { create } from "superstruct";

// src/debug.ts
function getQuoteParamsOverrides() {
  const overrides = {};
  const quoteParamsJson = localStorage.getItem("quoteParams");
  if (quoteParamsJson) {
    const quoteArgs = JSON.parse(quoteParamsJson);
    Object.assign(overrides, quoteArgs);
  }
  const quoteArgsJson = localStorage.getItem("quoteArgs");
  if (quoteArgsJson) {
    const quoteArgs = JSON.parse(quoteArgsJson);
    for (const [name, value] of Object.entries(quoteArgs)) {
      overrides[`quoteArg-${name}`] = value;
    }
  }
  return overrides;
}

// src/error.tsx
var Errors = {
  INITIALIZE_ERROR: "INITIALIZE_ERROR",
  FETCH_ROUTES_ERROR: "FETCH_ROUTES_ERROR",
  NO_ROUTES_ERROR: "NO_ROUTES_ERROR",
  TOKEN_NOT_TRADABLE: "TOKEN_NOT_TRADABLE",
  COULD_NOT_FIND_ANY_ROUTE: "COULD_NOT_FIND_ANY_ROUTE",
  INVALID_REQUEST: "INVALID_REQUEST",
  ROUTE_PLAN_DOES_NOT_CONSUME_ALL_THE_AMOUNT: "ROUTE_PLAN_DOES_NOT_CONSUME_ALL_THE_AMOUNT"
};

// src/utils/api.ts
import { array, coerce, instance, integer, number, string, type, enums, optional } from "superstruct";
import { PublicKey } from "@solana/web3.js";
import JSBI from "jsbi";
var PublicKeyFromString = coerce(instance(PublicKey), string(), (value) => new PublicKey(value));
var AmountFromString = coerce(instance(JSBI), string(), (value) => JSBI.BigInt(value));
var NumberFromString = coerce(string(), number(), (value) => new Number(value));
var SwapInfo = type({
  ammKey: PublicKeyFromString,
  label: string(),
  inputMint: string(),
  outputMint: string(),
  inAmount: AmountFromString,
  outAmount: AmountFromString,
  feeAmount: AmountFromString,
  feeMint: PublicKeyFromString
});
var RoutePlanStep = type({
  swapInfo: SwapInfo,
  percent: integer()
});
var RoutePlanWithMetadata = array(RoutePlanStep);
var QuoteResponse = type({
  inputMint: PublicKeyFromString,
  inAmount: AmountFromString,
  outputMint: PublicKeyFromString,
  outAmount: AmountFromString,
  otherAmountThreshold: AmountFromString,
  priceImpactPct: NumberFromString,
  swapMode: enums(["ExactIn", "ExactOut"]),
  routePlan: RoutePlanWithMetadata,
  slippageBps: number(),
  contextSlot: number(),
  computedAutoSlippage: optional(number())
});

// src/utils/ata.ts
import { ASSOCIATED_TOKEN_PROGRAM_ID, TOKEN_PROGRAM_ID } from "@solana/spl-token";
import { PublicKey as PublicKey2 } from "@solana/web3.js";
function getAssociatedTokenAddressSync(mint, owner, tokenProgramId = TOKEN_PROGRAM_ID) {
  const [ata] = PublicKey2.findProgramAddressSync(
    [owner.toBuffer(), tokenProgramId.toBuffer(), mint.toBuffer()],
    ASSOCIATED_TOKEN_PROGRAM_ID
  );
  return ata;
}

// src/utils/useDebounce.ts
import { useEffect, useState } from "react";
function useDebounce(value, delay) {
  const [debouncedValue, setDebouncedValue] = useState(value);
  useEffect(
    () => {
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, delay);
      return () => {
        clearTimeout(handler);
      };
    },
    [value, delay]
    // Only re-call effect if value or delay changes
  );
  return debouncedValue;
}

// src/index.tsx
import { JUPITER_PROGRAM_V4_ID, JUPITER_PROGRAM_V6_ID, OPENBOOK_PROGRAM_ID } from "@jup-ag/common";
var JupiterContext = createContext(null);
var JupiterProvider = ({
  onlyDirectRoutes,
  userPublicKey,
  jupiterQuoteApiUrl,
  children,
  ...jupiterLoadProps
}) => {
  const [programIdToLabelMap, setProgramIdToLabelMap] = useState2(/* @__PURE__ */ new Map());
  const [error, setError] = useState2();
  const [programIdsExcluded, setProgramIdsExcluded] = useState2(/* @__PURE__ */ new Set());
  const platformFeeAndAccountsRef = useRef();
  useEffect2(() => {
    platformFeeAndAccountsRef.current = jupiterLoadProps.platformFeeAndAccounts;
  }, [jupiterLoadProps.platformFeeAndAccounts]);
  const setPlatformFeeAccount = useCallback((mint, platformFeeAccount) => {
    if (!platformFeeAndAccountsRef.current) return;
    if (!platformFeeAndAccountsRef.current.feeAccounts) platformFeeAndAccountsRef.current.feeAccounts = /* @__PURE__ */ new Map();
    platformFeeAndAccountsRef.current.feeAccounts.set(mint, platformFeeAccount);
  }, []);
  const getPlatformFeeAccount = useCallback(async (platformFeeMint) => {
    var _a;
    const platformFeeAndAccounts = platformFeeAndAccountsRef.current;
    const platformFeeAccount = (_a = platformFeeAndAccounts == null ? void 0 : platformFeeAndAccounts.feeAccounts) == null ? void 0 : _a.get(platformFeeMint);
    const platformFeeBps = platformFeeAndAccounts == null ? void 0 : platformFeeAndAccounts.feeBps;
    if (platformFeeAccount) return platformFeeAccount;
    if (platformFeeAccount === null || !platformFeeBps) return;
    if (platformFeeAndAccounts == null ? void 0 : platformFeeAndAccounts.referralAccount) {
      const referralTokenAccount = await getReferralTokenAccount({
        connection: jupiterLoadProps.connection,
        referralAccount: platformFeeAndAccounts.referralAccount,
        mint: new PublicKey3(platformFeeMint)
      });
      setPlatformFeeAccount(platformFeeMint, referralTokenAccount);
      if (referralTokenAccount) return referralTokenAccount;
    }
    return;
  }, []);
  const jupiterApiClient = useMemo(() => {
    if (jupiterQuoteApiUrl) {
      return createJupiterApiClient({
        basePath: jupiterQuoteApiUrl
      });
    }
    return createJupiterApiClient();
  }, [jupiterQuoteApiUrl]);
  useEffect2(() => {
    setProgramIdsExcluded((set) => {
      if (set.size) {
        return /* @__PURE__ */ new Set();
      }
      return set;
    });
  }, [jupiterLoadProps.asLegacyTransaction]);
  useEffect2(() => {
    async function fetchProgramIdToLabelMap() {
      const programIdToLabel = await jupiterApiClient.programIdToLabelGet();
      const programIdToLabelMap2 = new Map(Object.entries(programIdToLabel));
      setProgramIdToLabelMap(programIdToLabelMap2);
    }
    fetchProgramIdToLabelMap();
  }, [jupiterApiClient]);
  return /* @__PURE__ */ React.createElement(
    JupiterContext.Provider,
    {
      value: {
        ...jupiterLoadProps,
        getPlatformFeeAccount,
        userPublicKey,
        programIdToLabelMap,
        jupiterApiClient,
        error,
        setError,
        onlyDirectRoutes,
        programIdsExcluded,
        setProgramIdsExcluded,
        whitelistedFeeMints: jupiterLoadProps.whitelistedFeeMints
      }
    },
    children
  );
};
var useJupiterClient = () => {
  const context = useContext(JupiterContext);
  if (!context) {
    throw new Error("JupiterProvider is required");
  }
  return context.jupiterApiClient;
};
var useJupiter = ({
  amount,
  inputMint,
  outputMint,
  autoSlippage,
  maxAutoSlippageBps,
  autoSlippageCollisionUsdValue,
  slippageBps,
  swapMode,
  debounceTime = 250,
  maxAccounts,
  disabledProgramIds,
  computeAutoSlippage,
  minimizeSlippage,
  preferLiquidDexes
}) => {
  const context = useContext(JupiterContext);
  const [loading, setLoading] = useState2(false);
  const [quoteResponseMeta, setQuoteResponseMeta] = useState2();
  const [refreshCount, setRefreshCount] = useState2(0);
  const lastRefreshCount = useRef(refreshCount);
  const [debouncedAmount, debouncedInputMint, debouncedOutputMint, debouncedSwapMode] = useDebounce(
    React.useMemo(() => {
      if (JSBI2.greaterThan(amount, JSBI2.BigInt(0))) {
        setLoading(true);
      }
      return [amount, inputMint, outputMint, swapMode];
    }, [amount.toString(), inputMint == null ? void 0 : inputMint.toBase58(), outputMint == null ? void 0 : outputMint.toBase58(), swapMode]),
    debounceTime
  );
  const lastRefreshTimestamp = useRef(0);
  const lastQueryTimestamp = useRef(0);
  if (!context) {
    throw new Error("JupiterProvider is required");
  }
  const {
    error,
    setError,
    routeCacheDuration = 0,
    onlyDirectRoutes,
    jupiterApiClient,
    userPublicKey,
    wrapUnwrapSOL = true,
    asLegacyTransaction,
    platformFeeAndAccounts,
    programIdToLabelMap,
    setProgramIdsExcluded,
    programIdsExcluded,
    whitelistedFeeMints
  } = context;
  useEffect2(() => {
    lastRefreshCount.current = -1;
  }, [
    [debouncedInputMint == null ? void 0 : debouncedInputMint.toString(), debouncedOutputMint == null ? void 0 : debouncedOutputMint.toString()].sort().join("-"),
    autoSlippage,
    maxAutoSlippageBps,
    autoSlippageCollisionUsdValue,
    slippageBps
  ]);
  useEffect2(() => {
    if (lastRefreshTimestamp.current && (/* @__PURE__ */ new Date()).getTime() - lastRefreshTimestamp.current >= routeCacheDuration) {
      lastRefreshCount.current = -1;
    }
    if (JSBI2.greaterThan(debouncedAmount, JSBI2.BigInt(0))) {
      setLoading(true);
    } else {
      setLoading(false);
    }
  }, [
    refreshCount,
    debouncedAmount,
    slippageBps,
    debouncedInputMint,
    debouncedOutputMint,
    onlyDirectRoutes,
    asLegacyTransaction
  ]);
  const fetchQuote = useCallback(async () => {
    try {
      if (JSBI2.equal(debouncedAmount, JSBI2.BigInt(0)) || error === Errors.INITIALIZE_ERROR) {
        return null;
      } else if (debouncedAmount) {
        if (!debouncedInputMint || !debouncedOutputMint) return null;
        if (slippageBps && autoSlippage) throw new Error("Cannot set both slippageBps and autoSlippage");
        if (!autoSlippage && maxAutoSlippageBps) throw new Error("Cannot set maxAutoSlippageBps without autoSlippage");
        const inputMint2 = debouncedInputMint.toBase58();
        const outputMint2 = debouncedOutputMint.toBase58();
        if (swapMode === SwapMode.ExactOut && maxAccounts) {
          console.warn("maxAccounts is not supported for ExactOut swap mode");
        }
        let quoteResponse;
        let original;
        const excludeDexes = Array.from([...programIdsExcluded, ...disabledProgramIds || []]).map((id) => programIdToLabelMap.get(id)).filter((id) => Boolean(id));
        const quoteParamsOverrides = getQuoteParamsOverrides();
        const apiClient = new Proxy(jupiterApiClient, {
          get: function(target, prop) {
            if (prop === "request") {
              return (context2, initOverrides) => {
                if (Object.keys(quoteParamsOverrides).length) {
                  console.log("Quoting with overrides:", quoteParamsOverrides);
                  Object.assign(context2.query, quoteParamsOverrides);
                }
                return target.request(context2, initOverrides);
              };
            }
            return Reflect.get(target, prop);
          }
        });
        const supportedMint = (swapMode === SwapMode.ExactOut ? [inputMint2] : [inputMint2, outputMint2]).filter(
          (mint) => {
            if (whitelistedFeeMints) {
              return whitelistedFeeMints == null ? void 0 : whitelistedFeeMints.includes(mint);
            }
            return true;
          }
        );
        const platformFeeBps = supportedMint.length > 0 ? platformFeeAndAccounts == null ? void 0 : platformFeeAndAccounts.feeBps : void 0;
        const response = await promiseRetry(
          (retry) => apiClient.quoteGet({
            inputMint: inputMint2,
            outputMint: outputMint2,
            amount: debouncedAmount.toString(),
            // making it number for u64.MAX would cause precision err
            autoSlippage,
            maxAutoSlippageBps,
            autoSlippageCollisionUsdValue,
            slippageBps,
            swapMode,
            excludeDexes: excludeDexes.length ? [excludeDexes.filter(Boolean).join(",")] : void 0,
            onlyDirectRoutes,
            asLegacyTransaction,
            platformFeeBps,
            maxAccounts,
            computeAutoSlippage,
            minimizeSlippage,
            preferLiquidDexes
          }).catch((e) => {
            retry(e);
          }),
          {
            retries: 1,
            minTimeout: 100
          }
        );
        original = Object.assign({}, response);
        quoteResponse = create(response, QuoteResponse);
        return { quoteResponse, original, timestamp: (/* @__PURE__ */ new Date()).getTime() };
      }
      return null;
    } catch (error2) {
      throw error2;
    }
  }, [
    debouncedAmount,
    debouncedInputMint,
    debouncedOutputMint,
    debouncedSwapMode,
    slippageBps,
    maxAutoSlippageBps,
    swapMode,
    maxAccounts,
    userPublicKey,
    wrapUnwrapSOL,
    refreshCount,
    onlyDirectRoutes,
    asLegacyTransaction,
    programIdsExcluded,
    preferLiquidDexes
  ]);
  useEffect2(() => {
    let requestAborted = false;
    (async () => {
      if (JSBI2.equal(debouncedAmount, JSBI2.BigInt(0)) || error === Errors.INITIALIZE_ERROR) {
        setLoading(false);
        setQuoteResponseMeta(void 0);
      } else if (debouncedAmount) {
        if (!debouncedInputMint || !debouncedOutputMint) return;
        const lastUpdatedTime = (/* @__PURE__ */ new Date()).getTime();
        lastQueryTimestamp.current = lastUpdatedTime;
        lastRefreshTimestamp.current = lastUpdatedTime;
        try {
          const fetchedQuote = await fetchQuote();
          if (!(fetchedQuote == null ? void 0 : fetchedQuote.quoteResponse) || !(fetchedQuote == null ? void 0 : fetchedQuote.original) || !(fetchedQuote == null ? void 0 : fetchedQuote.timestamp) || requestAborted)
            return;
          setQuoteResponseMeta({
            quoteResponse: fetchedQuote.quoteResponse,
            original: fetchedQuote.original
          });
          setError(void 0);
        } catch (e) {
          console.error(e);
          if (lastQueryTimestamp.current !== lastUpdatedTime) {
            return;
          }
          const haveErrorCode = await (async () => {
            try {
              if ("response" in e && e.response instanceof Response) {
                const errorObj = await e.response.json();
                return errorObj.errorCode || null;
              }
            } catch (error2) {
              console.log("Error parsing error response", error2);
              return null;
            }
          })();
          if (haveErrorCode) {
            setError(haveErrorCode);
          } else {
            setError(Errors.FETCH_ROUTES_ERROR);
          }
          setQuoteResponseMeta(void 0);
        } finally {
          if (lastQueryTimestamp.current !== lastUpdatedTime) {
            return;
          }
          lastRefreshCount.current = refreshCount;
          setLoading(false);
        }
      }
    })();
    return () => {
      requestAborted = true;
    };
  }, [fetchQuote]);
  const { exchange, fetchSwapTransaction } = useJupiterExchange(swapMode);
  return {
    exchange,
    fetchQuote,
    fetchSwapTransaction,
    refresh: React.useCallback(() => {
      if (!loading && lastRefreshTimestamp.current) {
        setRefreshCount((refreshCount2) => refreshCount2 + 1);
      }
    }, [loading]),
    lastRefreshTimestamp: lastRefreshTimestamp.current,
    loading,
    quoteResponseMeta,
    error,
    programIdToLabelMap,
    setProgramIdsExcluded,
    programIdsExcluded
  };
};
var useJupiterExchange = (swapMode) => {
  const context = useContext(JupiterContext);
  if (!context) {
    throw new Error("JupiterProvider is required");
  }
  const {
    jupiterApiClient,
    userPublicKey,
    connection,
    wrapUnwrapSOL = true,
    asLegacyTransaction,
    platformFeeAndAccounts,
    useSharedAccounts,
    getPlatformFeeAccount,
    whitelistedFeeMints
  } = context;
  const fetchSwapTransaction = useCallback(
    async ({
      quoteResponseMeta,
      userPublicKey: userPublicKey2,
      wrapUnwrapSOL: propsWrapUnwrapSOL,
      prioritizationFeeLamports,
      allowOptimizedWrappedSolTokenAccount,
      dynamicSlippage
    }) => {
      if (!userPublicKey2) throw new Error("User public key is required");
      const wrapAndUnwrapSol = Boolean(propsWrapUnwrapSOL ?? wrapUnwrapSOL ?? true);
      const { inputMint, outputMint } = quoteResponseMeta.quoteResponse;
      const supportedMints = (swapMode === SwapMode.ExactOut ? [inputMint] : [inputMint, outputMint]).filter((mint) => {
        if (whitelistedFeeMints) {
          return whitelistedFeeMints == null ? void 0 : whitelistedFeeMints.includes(mint.toString());
        }
        return true;
      });
      const feeAccount = supportedMints.length !== 0 ? await (async () => {
        for (const mint of supportedMints) {
          const account = await getPlatformFeeAccount(mint.toBase58());
          if (account) {
            return account;
          }
        }
        return void 0;
      })() : void 0;
      const result = await promiseRetry(
        (retry) => jupiterApiClient.swapPost({
          swapRequest: {
            quoteResponse: quoteResponseMeta.original,
            userPublicKey: userPublicKey2.toBase58(),
            wrapAndUnwrapSol,
            feeAccount: feeAccount == null ? void 0 : feeAccount.toBase58(),
            asLegacyTransaction,
            useSharedAccounts,
            dynamicComputeUnitLimit: true,
            allowOptimizedWrappedSolTokenAccount,
            prioritizationFeeLamports,
            dynamicSlippage,
            correctLastValidBlockHeight: true
          }
        }).catch(async (res) => {
          try {
            const { error } = await res.json();
            return { error: new TransactionError(error) };
          } catch (e) {
            return retry({
              error: new TransactionError("Unknown error")
            });
          }
        }),
        {
          retries: 1,
          minTimeout: 100
        }
      );
      if ("error" in result) {
        return result;
      }
      const { swapTransaction: swapTransactionSerialized, lastValidBlockHeight } = result;
      const swapTransactionBuffer = Buffer.from(swapTransactionSerialized, "base64");
      const swapTransaction = asLegacyTransaction ? Transaction.from(swapTransactionBuffer) : VersionedTransaction.deserialize(swapTransactionBuffer);
      checkTxSize(swapTransaction);
      return {
        swapTransaction,
        blockhash: getTransactionBlockhash(swapTransaction),
        lastValidBlockHeight,
        dynamicSlippageReport: result.dynamicSlippageReport
      };
    },
    [wrapUnwrapSOL, asLegacyTransaction, useSharedAccounts, getPlatformFeeAccount, jupiterApiClient]
  );
  const exchange = useCallback(
    async ({
      wallet,
      quoteResponseMeta,
      onTransaction,
      prioritizationFeeLamports,
      ...restExchangeProps
    }) => {
      if (!userPublicKey) throw new Error("User public key is required");
      if (!wallet) throw new Error("Wallet is required");
      const { inputMint, outputMint } = quoteResponseMeta.quoteResponse;
      const wrapAndUnwrapSol = Boolean(restExchangeProps.wrapUnwrapSOL ?? wrapUnwrapSOL ?? true);
      const [result, accountInfos] = await Promise.all([
        fetchSwapTransaction({
          quoteResponseMeta,
          userPublicKey,
          wrapUnwrapSOL: wrapAndUnwrapSol,
          prioritizationFeeLamports,
          ...restExchangeProps
        }),
        connection.getMultipleAccountsInfo([inputMint, outputMint])
      ]);
      if ("error" in result) {
        return result;
      }
      const { blockhash, lastValidBlockHeight, swapTransaction } = result;
      const [sourceAddress, destinationAddress] = [inputMint, outputMint].map(
        (mint, idx) => {
          var _a;
          return getAssociatedTokenAddressSync(mint, userPublicKey, ((_a = accountInfos[idx]) == null ? void 0 : _a.owner) || TOKEN_PROGRAM_ID2);
        }
      );
      return executeTransaction({
        connection,
        wallet,
        onTransaction,
        inputMint,
        outputMint,
        sourceAddress,
        destinationAddress,
        swapTransaction,
        blockhashWithBlockHeight: { lastValidBlockHeight, blockhash },
        owner: new Owner(userPublicKey),
        wrapUnwrapSOL: wrapAndUnwrapSol
      });
    },
    [userPublicKey, wrapUnwrapSOL, asLegacyTransaction, connection, platformFeeAndAccounts, jupiterApiClient, swapMode]
  );
  return { exchange, fetchSwapTransaction };
};
var calculateFeeForSwap = (quoteResponse, mintToAtaMap, openbookMarketToOpenOrderMap, wrapUnwrapSOL, useSharedAccounts) => {
  const atasToBeCreated = /* @__PURE__ */ new Set();
  const openOrdersToBeCreated = /* @__PURE__ */ new Set();
  const isInputMintSOL = Boolean(wrapUnwrapSOL && (quoteResponse == null ? void 0 : quoteResponse.inputMint.equals(WRAPPED_SOL_MINT)));
  const walkRoutePlan = (routePlan) => {
    if (!useSharedAccounts) {
      routePlan.forEach((routeStep) => {
        if (!mintToAtaMap.has(routeStep.swapInfo.outputMint)) {
          atasToBeCreated.add(routeStep.swapInfo.outputMint);
        }
        if (routeStep.swapInfo.label === "Openbook") {
          if (!openbookMarketToOpenOrderMap.has(routeStep.swapInfo.ammKey.toBase58())) {
            openOrdersToBeCreated.add(routeStep.swapInfo.outputMint);
          }
        }
      });
    }
  };
  if (quoteResponse) {
    if (useSharedAccounts) {
      let numOfAtasToBeCreated = 0;
      if (quoteResponse.inputMint.equals(WRAPPED_SOL_MINT) || quoteResponse.outputMint.equals(WRAPPED_SOL_MINT) && wrapUnwrapSOL && !mintToAtaMap.has(WRAPPED_SOL_MINT.toBase58())) {
        numOfAtasToBeCreated = 1;
      }
      return calculateTransactionDepositAndFeeV2({
        numOfAtasToBeCreated,
        numOfOpenOrdersToBeCreated: 0,
        isInputMintSOL
      });
    }
    if (quoteResponse.inputMint.equals(WRAPPED_SOL_MINT) && wrapUnwrapSOL && !mintToAtaMap.has(WRAPPED_SOL_MINT.toBase58())) {
      atasToBeCreated.add(WRAPPED_SOL_MINT.toBase58());
    }
    walkRoutePlan(quoteResponse.routePlan);
  }
  return calculateTransactionDepositAndFeeV2({
    numOfAtasToBeCreated: atasToBeCreated.size,
    numOfOpenOrdersToBeCreated: openOrdersToBeCreated.size,
    isInputMintSOL
  });
};
export {
  Errors,
  JUPITER_FEE_OWNER,
  JUPITER_PROGRAM_V4_ID,
  JUPITER_PROGRAM_V6_ID,
  JUPITER_V6_ERRORS,
  JupiterProvider,
  LAMPORTS_PER_SIGNATURE,
  MARKETS_URL,
  OPENBOOK_PROGRAM_ID,
  Owner,
  SwapMode,
  TOKEN_LIST_URL,
  calculateFeeForSwap,
  useJupiter,
  useJupiterClient,
  useJupiterExchange
};
