// @jup-ag/react-hook/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } }// src/index.tsx


var _api = require('@jup-ag/api');












var _common = require('@jup-ag/common');
var _optimist = require('@mercurial-finance/optimist');
var _spltoken = require('@solana/spl-token');
var _web3js = require('@solana/web3.js');
var _jsbi = require('jsbi'); var _jsbi2 = _interopRequireDefault(_jsbi);
var _promiseretry = require('promise-retry'); var _promiseretry2 = _interopRequireDefault(_promiseretry);
var _react = require('react'); var _react2 = _interopRequireDefault(_react);
var _superstruct = require('superstruct');

// src/debug.ts
function getQuoteParamsOverrides() {
  const overrides = {};
  const quoteParamsJson = localStorage.getItem("quoteParams");
  if (quoteParamsJson) {
    const quoteArgs = JSON.parse(quoteParamsJson);
    Object.assign(overrides, quoteArgs);
  }
  const quoteArgsJson = localStorage.getItem("quoteArgs");
  if (quoteArgsJson) {
    const quoteArgs = JSON.parse(quoteArgsJson);
    for (const [name, value] of Object.entries(quoteArgs)) {
      overrides[`quoteArg-${name}`] = value;
    }
  }
  return overrides;
}

// src/error.tsx
var Errors = {
  INITIALIZE_ERROR: "INITIALIZE_ERROR",
  FETCH_ROUTES_ERROR: "FETCH_ROUTES_ERROR",
  NO_ROUTES_ERROR: "NO_ROUTES_ERROR",
  TOKEN_NOT_TRADABLE: "TOKEN_NOT_TRADABLE",
  COULD_NOT_FIND_ANY_ROUTE: "COULD_NOT_FIND_ANY_ROUTE",
  INVALID_REQUEST: "INVALID_REQUEST",
  ROUTE_PLAN_DOES_NOT_CONSUME_ALL_THE_AMOUNT: "ROUTE_PLAN_DOES_NOT_CONSUME_ALL_THE_AMOUNT"
};

// src/utils/api.ts



var PublicKeyFromString = _superstruct.coerce.call(void 0, _superstruct.instance.call(void 0, _web3js.PublicKey), _superstruct.string.call(void 0, ), (value) => new (0, _web3js.PublicKey)(value));
var AmountFromString = _superstruct.coerce.call(void 0, _superstruct.instance.call(void 0, _jsbi2.default), _superstruct.string.call(void 0, ), (value) => _jsbi2.default.BigInt(value));
var NumberFromString = _superstruct.coerce.call(void 0, _superstruct.string.call(void 0, ), _superstruct.number.call(void 0, ), (value) => new Number(value));
var SwapInfo = _superstruct.type.call(void 0, {
  ammKey: PublicKeyFromString,
  label: _superstruct.string.call(void 0, ),
  inputMint: _superstruct.string.call(void 0, ),
  outputMint: _superstruct.string.call(void 0, ),
  inAmount: AmountFromString,
  outAmount: AmountFromString,
  feeAmount: AmountFromString,
  feeMint: PublicKeyFromString
});
var RoutePlanStep = _superstruct.type.call(void 0, {
  swapInfo: SwapInfo,
  percent: _superstruct.integer.call(void 0, )
});
var RoutePlanWithMetadata = _superstruct.array.call(void 0, RoutePlanStep);
var QuoteResponse = _superstruct.type.call(void 0, {
  inputMint: PublicKeyFromString,
  inAmount: AmountFromString,
  outputMint: PublicKeyFromString,
  outAmount: AmountFromString,
  otherAmountThreshold: AmountFromString,
  priceImpactPct: NumberFromString,
  swapMode: _superstruct.enums.call(void 0, ["ExactIn", "ExactOut"]),
  routePlan: RoutePlanWithMetadata,
  slippageBps: _superstruct.number.call(void 0, ),
  contextSlot: _superstruct.number.call(void 0, ),
  computedAutoSlippage: _superstruct.optional.call(void 0, _superstruct.number.call(void 0, ))
});

// src/utils/ata.ts


function getAssociatedTokenAddressSync(mint, owner, tokenProgramId = _spltoken.TOKEN_PROGRAM_ID) {
  const [ata] = _web3js.PublicKey.findProgramAddressSync(
    [owner.toBuffer(), tokenProgramId.toBuffer(), mint.toBuffer()],
    _spltoken.ASSOCIATED_TOKEN_PROGRAM_ID
  );
  return ata;
}

// src/utils/useDebounce.ts

function useDebounce(value, delay) {
  const [debouncedValue, setDebouncedValue] = _react.useState.call(void 0, value);
  _react.useEffect.call(void 0, 
    () => {
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, delay);
      return () => {
        clearTimeout(handler);
      };
    },
    [value, delay]
    // Only re-call effect if value or delay changes
  );
  return debouncedValue;
}

// src/index.tsx

var JupiterContext = _react.createContext.call(void 0, null);
var JupiterProvider = ({
  onlyDirectRoutes,
  userPublicKey,
  jupiterQuoteApiUrl,
  children,
  ...jupiterLoadProps
}) => {
  const [programIdToLabelMap, setProgramIdToLabelMap] = _react.useState.call(void 0, /* @__PURE__ */ new Map());
  const [error, setError] = _react.useState.call(void 0, );
  const [programIdsExcluded, setProgramIdsExcluded] = _react.useState.call(void 0, /* @__PURE__ */ new Set());
  const platformFeeAndAccountsRef = _react.useRef.call(void 0, );
  _react.useEffect.call(void 0, () => {
    platformFeeAndAccountsRef.current = jupiterLoadProps.platformFeeAndAccounts;
  }, [jupiterLoadProps.platformFeeAndAccounts]);
  const setPlatformFeeAccount = _react.useCallback.call(void 0, (mint, platformFeeAccount) => {
    if (!platformFeeAndAccountsRef.current) return;
    if (!platformFeeAndAccountsRef.current.feeAccounts) platformFeeAndAccountsRef.current.feeAccounts = /* @__PURE__ */ new Map();
    platformFeeAndAccountsRef.current.feeAccounts.set(mint, platformFeeAccount);
  }, []);
  const getPlatformFeeAccount = _react.useCallback.call(void 0, async (platformFeeMint) => {
    var _a;
    const platformFeeAndAccounts = platformFeeAndAccountsRef.current;
    const platformFeeAccount = (_a = platformFeeAndAccounts == null ? void 0 : platformFeeAndAccounts.feeAccounts) == null ? void 0 : _a.get(platformFeeMint);
    const platformFeeBps = platformFeeAndAccounts == null ? void 0 : platformFeeAndAccounts.feeBps;
    if (platformFeeAccount) return platformFeeAccount;
    if (platformFeeAccount === null || !platformFeeBps) return;
    if (platformFeeAndAccounts == null ? void 0 : platformFeeAndAccounts.referralAccount) {
      const referralTokenAccount = await _common.getReferralTokenAccount.call(void 0, {
        connection: jupiterLoadProps.connection,
        referralAccount: platformFeeAndAccounts.referralAccount,
        mint: new (0, _web3js.PublicKey)(platformFeeMint)
      });
      setPlatformFeeAccount(platformFeeMint, referralTokenAccount);
      if (referralTokenAccount) return referralTokenAccount;
    }
    return;
  }, []);
  const jupiterApiClient = _react.useMemo.call(void 0, () => {
    if (jupiterQuoteApiUrl) {
      return _api.createJupiterApiClient.call(void 0, {
        basePath: jupiterQuoteApiUrl
      });
    }
    return _api.createJupiterApiClient.call(void 0, );
  }, [jupiterQuoteApiUrl]);
  _react.useEffect.call(void 0, () => {
    setProgramIdsExcluded((set) => {
      if (set.size) {
        return /* @__PURE__ */ new Set();
      }
      return set;
    });
  }, [jupiterLoadProps.asLegacyTransaction]);
  _react.useEffect.call(void 0, () => {
    async function fetchProgramIdToLabelMap() {
      const programIdToLabel = await jupiterApiClient.programIdToLabelGet();
      const programIdToLabelMap2 = new Map(Object.entries(programIdToLabel));
      setProgramIdToLabelMap(programIdToLabelMap2);
    }
    fetchProgramIdToLabelMap();
  }, [jupiterApiClient]);
  return /* @__PURE__ */ _react2.default.createElement(
    JupiterContext.Provider,
    {
      value: {
        ...jupiterLoadProps,
        getPlatformFeeAccount,
        userPublicKey,
        programIdToLabelMap,
        jupiterApiClient,
        error,
        setError,
        onlyDirectRoutes,
        programIdsExcluded,
        setProgramIdsExcluded,
        whitelistedFeeMints: jupiterLoadProps.whitelistedFeeMints
      }
    },
    children
  );
};
var useJupiterClient = () => {
  const context = _react.useContext.call(void 0, JupiterContext);
  if (!context) {
    throw new Error("JupiterProvider is required");
  }
  return context.jupiterApiClient;
};
var useJupiter = ({
  amount,
  inputMint,
  outputMint,
  autoSlippage,
  maxAutoSlippageBps,
  autoSlippageCollisionUsdValue,
  slippageBps,
  swapMode,
  debounceTime = 250,
  maxAccounts,
  disabledProgramIds,
  computeAutoSlippage,
  minimizeSlippage,
  preferLiquidDexes
}) => {
  const context = _react.useContext.call(void 0, JupiterContext);
  const [loading, setLoading] = _react.useState.call(void 0, false);
  const [quoteResponseMeta, setQuoteResponseMeta] = _react.useState.call(void 0, );
  const [refreshCount, setRefreshCount] = _react.useState.call(void 0, 0);
  const lastRefreshCount = _react.useRef.call(void 0, refreshCount);
  const [debouncedAmount, debouncedInputMint, debouncedOutputMint, debouncedSwapMode] = useDebounce(
    _react2.default.useMemo(() => {
      if (_jsbi2.default.greaterThan(amount, _jsbi2.default.BigInt(0))) {
        setLoading(true);
      }
      return [amount, inputMint, outputMint, swapMode];
    }, [amount.toString(), inputMint == null ? void 0 : inputMint.toBase58(), outputMint == null ? void 0 : outputMint.toBase58(), swapMode]),
    debounceTime
  );
  const lastRefreshTimestamp = _react.useRef.call(void 0, 0);
  const lastQueryTimestamp = _react.useRef.call(void 0, 0);
  if (!context) {
    throw new Error("JupiterProvider is required");
  }
  const {
    error,
    setError,
    routeCacheDuration = 0,
    onlyDirectRoutes,
    jupiterApiClient,
    userPublicKey,
    wrapUnwrapSOL = true,
    asLegacyTransaction,
    platformFeeAndAccounts,
    programIdToLabelMap,
    setProgramIdsExcluded,
    programIdsExcluded,
    whitelistedFeeMints
  } = context;
  _react.useEffect.call(void 0, () => {
    lastRefreshCount.current = -1;
  }, [
    [debouncedInputMint == null ? void 0 : debouncedInputMint.toString(), debouncedOutputMint == null ? void 0 : debouncedOutputMint.toString()].sort().join("-"),
    autoSlippage,
    maxAutoSlippageBps,
    autoSlippageCollisionUsdValue,
    slippageBps
  ]);
  _react.useEffect.call(void 0, () => {
    if (lastRefreshTimestamp.current && (/* @__PURE__ */ new Date()).getTime() - lastRefreshTimestamp.current >= routeCacheDuration) {
      lastRefreshCount.current = -1;
    }
    if (_jsbi2.default.greaterThan(debouncedAmount, _jsbi2.default.BigInt(0))) {
      setLoading(true);
    } else {
      setLoading(false);
    }
  }, [
    refreshCount,
    debouncedAmount,
    slippageBps,
    debouncedInputMint,
    debouncedOutputMint,
    onlyDirectRoutes,
    asLegacyTransaction
  ]);
  const fetchQuote = _react.useCallback.call(void 0, async () => {
    try {
      if (_jsbi2.default.equal(debouncedAmount, _jsbi2.default.BigInt(0)) || error === Errors.INITIALIZE_ERROR) {
        return null;
      } else if (debouncedAmount) {
        if (!debouncedInputMint || !debouncedOutputMint) return null;
        if (slippageBps && autoSlippage) throw new Error("Cannot set both slippageBps and autoSlippage");
        if (!autoSlippage && maxAutoSlippageBps) throw new Error("Cannot set maxAutoSlippageBps without autoSlippage");
        const inputMint2 = debouncedInputMint.toBase58();
        const outputMint2 = debouncedOutputMint.toBase58();
        if (swapMode === _common.SwapMode.ExactOut && maxAccounts) {
          console.warn("maxAccounts is not supported for ExactOut swap mode");
        }
        let quoteResponse;
        let original;
        const excludeDexes = Array.from([...programIdsExcluded, ...disabledProgramIds || []]).map((id) => programIdToLabelMap.get(id)).filter((id) => Boolean(id));
        const quoteParamsOverrides = getQuoteParamsOverrides();
        const apiClient = new Proxy(jupiterApiClient, {
          get: function(target, prop) {
            if (prop === "request") {
              return (context2, initOverrides) => {
                if (Object.keys(quoteParamsOverrides).length) {
                  console.log("Quoting with overrides:", quoteParamsOverrides);
                  Object.assign(context2.query, quoteParamsOverrides);
                }
                return target.request(context2, initOverrides);
              };
            }
            return Reflect.get(target, prop);
          }
        });
        const supportedMint = (swapMode === _common.SwapMode.ExactOut ? [inputMint2] : [inputMint2, outputMint2]).filter(
          (mint) => {
            if (whitelistedFeeMints) {
              return whitelistedFeeMints == null ? void 0 : whitelistedFeeMints.includes(mint);
            }
            return true;
          }
        );
        const platformFeeBps = supportedMint.length > 0 ? platformFeeAndAccounts == null ? void 0 : platformFeeAndAccounts.feeBps : void 0;
        const response = await _promiseretry2.default.call(void 0, 
          (retry) => apiClient.quoteGet({
            inputMint: inputMint2,
            outputMint: outputMint2,
            amount: debouncedAmount.toString(),
            // making it number for u64.MAX would cause precision err
            autoSlippage,
            maxAutoSlippageBps,
            autoSlippageCollisionUsdValue,
            slippageBps,
            swapMode,
            excludeDexes: excludeDexes.length ? [excludeDexes.filter(Boolean).join(",")] : void 0,
            onlyDirectRoutes,
            asLegacyTransaction,
            platformFeeBps,
            maxAccounts,
            computeAutoSlippage,
            minimizeSlippage,
            preferLiquidDexes
          }).catch((e) => {
            retry(e);
          }),
          {
            retries: 1,
            minTimeout: 100
          }
        );
        original = Object.assign({}, response);
        quoteResponse = _superstruct.create.call(void 0, response, QuoteResponse);
        return { quoteResponse, original, timestamp: (/* @__PURE__ */ new Date()).getTime() };
      }
      return null;
    } catch (error2) {
      throw error2;
    }
  }, [
    debouncedAmount,
    debouncedInputMint,
    debouncedOutputMint,
    debouncedSwapMode,
    slippageBps,
    maxAutoSlippageBps,
    swapMode,
    maxAccounts,
    userPublicKey,
    wrapUnwrapSOL,
    refreshCount,
    onlyDirectRoutes,
    asLegacyTransaction,
    programIdsExcluded,
    preferLiquidDexes
  ]);
  _react.useEffect.call(void 0, () => {
    let requestAborted = false;
    (async () => {
      if (_jsbi2.default.equal(debouncedAmount, _jsbi2.default.BigInt(0)) || error === Errors.INITIALIZE_ERROR) {
        setLoading(false);
        setQuoteResponseMeta(void 0);
      } else if (debouncedAmount) {
        if (!debouncedInputMint || !debouncedOutputMint) return;
        const lastUpdatedTime = (/* @__PURE__ */ new Date()).getTime();
        lastQueryTimestamp.current = lastUpdatedTime;
        lastRefreshTimestamp.current = lastUpdatedTime;
        try {
          const fetchedQuote = await fetchQuote();
          if (!(fetchedQuote == null ? void 0 : fetchedQuote.quoteResponse) || !(fetchedQuote == null ? void 0 : fetchedQuote.original) || !(fetchedQuote == null ? void 0 : fetchedQuote.timestamp) || requestAborted)
            return;
          setQuoteResponseMeta({
            quoteResponse: fetchedQuote.quoteResponse,
            original: fetchedQuote.original
          });
          setError(void 0);
        } catch (e) {
          console.error(e);
          if (lastQueryTimestamp.current !== lastUpdatedTime) {
            return;
          }
          const haveErrorCode = await (async () => {
            try {
              if ("response" in e && e.response instanceof Response) {
                const errorObj = await e.response.json();
                return errorObj.errorCode || null;
              }
            } catch (error2) {
              console.log("Error parsing error response", error2);
              return null;
            }
          })();
          if (haveErrorCode) {
            setError(haveErrorCode);
          } else {
            setError(Errors.FETCH_ROUTES_ERROR);
          }
          setQuoteResponseMeta(void 0);
        } finally {
          if (lastQueryTimestamp.current !== lastUpdatedTime) {
            return;
          }
          lastRefreshCount.current = refreshCount;
          setLoading(false);
        }
      }
    })();
    return () => {
      requestAborted = true;
    };
  }, [fetchQuote]);
  const { exchange, fetchSwapTransaction } = useJupiterExchange(swapMode);
  return {
    exchange,
    fetchQuote,
    fetchSwapTransaction,
    refresh: _react2.default.useCallback(() => {
      if (!loading && lastRefreshTimestamp.current) {
        setRefreshCount((refreshCount2) => refreshCount2 + 1);
      }
    }, [loading]),
    lastRefreshTimestamp: lastRefreshTimestamp.current,
    loading,
    quoteResponseMeta,
    error,
    programIdToLabelMap,
    setProgramIdsExcluded,
    programIdsExcluded
  };
};
var useJupiterExchange = (swapMode) => {
  const context = _react.useContext.call(void 0, JupiterContext);
  if (!context) {
    throw new Error("JupiterProvider is required");
  }
  const {
    jupiterApiClient,
    userPublicKey,
    connection,
    wrapUnwrapSOL = true,
    asLegacyTransaction,
    platformFeeAndAccounts,
    useSharedAccounts,
    getPlatformFeeAccount,
    whitelistedFeeMints
  } = context;
  const fetchSwapTransaction = _react.useCallback.call(void 0, 
    async ({
      quoteResponseMeta,
      userPublicKey: userPublicKey2,
      wrapUnwrapSOL: propsWrapUnwrapSOL,
      prioritizationFeeLamports,
      allowOptimizedWrappedSolTokenAccount,
      dynamicSlippage
    }) => {
      if (!userPublicKey2) throw new Error("User public key is required");
      const wrapAndUnwrapSol = Boolean(_nullishCoalesce(_nullishCoalesce(propsWrapUnwrapSOL, () => ( wrapUnwrapSOL)), () => ( true)));
      const { inputMint, outputMint } = quoteResponseMeta.quoteResponse;
      const supportedMints = (swapMode === _common.SwapMode.ExactOut ? [inputMint] : [inputMint, outputMint]).filter((mint) => {
        if (whitelistedFeeMints) {
          return whitelistedFeeMints == null ? void 0 : whitelistedFeeMints.includes(mint.toString());
        }
        return true;
      });
      const feeAccount = supportedMints.length !== 0 ? await (async () => {
        for (const mint of supportedMints) {
          const account = await getPlatformFeeAccount(mint.toBase58());
          if (account) {
            return account;
          }
        }
        return void 0;
      })() : void 0;
      const result = await _promiseretry2.default.call(void 0, 
        (retry) => jupiterApiClient.swapPost({
          swapRequest: {
            quoteResponse: quoteResponseMeta.original,
            userPublicKey: userPublicKey2.toBase58(),
            wrapAndUnwrapSol,
            feeAccount: feeAccount == null ? void 0 : feeAccount.toBase58(),
            asLegacyTransaction,
            useSharedAccounts,
            dynamicComputeUnitLimit: true,
            allowOptimizedWrappedSolTokenAccount,
            prioritizationFeeLamports,
            dynamicSlippage,
            correctLastValidBlockHeight: true
          }
        }).catch(async (res) => {
          try {
            const { error } = await res.json();
            return { error: new (0, _optimist.TransactionError)(error) };
          } catch (e) {
            return retry({
              error: new (0, _optimist.TransactionError)("Unknown error")
            });
          }
        }),
        {
          retries: 1,
          minTimeout: 100
        }
      );
      if ("error" in result) {
        return result;
      }
      const { swapTransaction: swapTransactionSerialized, lastValidBlockHeight } = result;
      const swapTransactionBuffer = Buffer.from(swapTransactionSerialized, "base64");
      const swapTransaction = asLegacyTransaction ? _web3js.Transaction.from(swapTransactionBuffer) : _web3js.VersionedTransaction.deserialize(swapTransactionBuffer);
      _optimist.checkTxSize.call(void 0, swapTransaction);
      return {
        swapTransaction,
        blockhash: _common.getTransactionBlockhash.call(void 0, swapTransaction),
        lastValidBlockHeight,
        dynamicSlippageReport: result.dynamicSlippageReport
      };
    },
    [wrapUnwrapSOL, asLegacyTransaction, useSharedAccounts, getPlatformFeeAccount, jupiterApiClient]
  );
  const exchange = _react.useCallback.call(void 0, 
    async ({
      wallet,
      quoteResponseMeta,
      onTransaction,
      prioritizationFeeLamports,
      ...restExchangeProps
    }) => {
      if (!userPublicKey) throw new Error("User public key is required");
      if (!wallet) throw new Error("Wallet is required");
      const { inputMint, outputMint } = quoteResponseMeta.quoteResponse;
      const wrapAndUnwrapSol = Boolean(_nullishCoalesce(_nullishCoalesce(restExchangeProps.wrapUnwrapSOL, () => ( wrapUnwrapSOL)), () => ( true)));
      const [result, accountInfos] = await Promise.all([
        fetchSwapTransaction({
          quoteResponseMeta,
          userPublicKey,
          wrapUnwrapSOL: wrapAndUnwrapSol,
          prioritizationFeeLamports,
          ...restExchangeProps
        }),
        connection.getMultipleAccountsInfo([inputMint, outputMint])
      ]);
      if ("error" in result) {
        return result;
      }
      const { blockhash, lastValidBlockHeight, swapTransaction } = result;
      const [sourceAddress, destinationAddress] = [inputMint, outputMint].map(
        (mint, idx) => {
          var _a;
          return getAssociatedTokenAddressSync(mint, userPublicKey, ((_a = accountInfos[idx]) == null ? void 0 : _a.owner) || _spltoken.TOKEN_PROGRAM_ID);
        }
      );
      return _common.executeTransaction.call(void 0, {
        connection,
        wallet,
        onTransaction,
        inputMint,
        outputMint,
        sourceAddress,
        destinationAddress,
        swapTransaction,
        blockhashWithBlockHeight: { lastValidBlockHeight, blockhash },
        owner: new (0, _common.Owner)(userPublicKey),
        wrapUnwrapSOL: wrapAndUnwrapSol
      });
    },
    [userPublicKey, wrapUnwrapSOL, asLegacyTransaction, connection, platformFeeAndAccounts, jupiterApiClient, swapMode]
  );
  return { exchange, fetchSwapTransaction };
};
var calculateFeeForSwap = (quoteResponse, mintToAtaMap, openbookMarketToOpenOrderMap, wrapUnwrapSOL, useSharedAccounts) => {
  const atasToBeCreated = /* @__PURE__ */ new Set();
  const openOrdersToBeCreated = /* @__PURE__ */ new Set();
  const isInputMintSOL = Boolean(wrapUnwrapSOL && (quoteResponse == null ? void 0 : quoteResponse.inputMint.equals(_spltoken.NATIVE_MINT)));
  const walkRoutePlan = (routePlan) => {
    if (!useSharedAccounts) {
      routePlan.forEach((routeStep) => {
        if (!mintToAtaMap.has(routeStep.swapInfo.outputMint)) {
          atasToBeCreated.add(routeStep.swapInfo.outputMint);
        }
        if (routeStep.swapInfo.label === "Openbook") {
          if (!openbookMarketToOpenOrderMap.has(routeStep.swapInfo.ammKey.toBase58())) {
            openOrdersToBeCreated.add(routeStep.swapInfo.outputMint);
          }
        }
      });
    }
  };
  if (quoteResponse) {
    if (useSharedAccounts) {
      let numOfAtasToBeCreated = 0;
      if (quoteResponse.inputMint.equals(_spltoken.NATIVE_MINT) || quoteResponse.outputMint.equals(_spltoken.NATIVE_MINT) && wrapUnwrapSOL && !mintToAtaMap.has(_spltoken.NATIVE_MINT.toBase58())) {
        numOfAtasToBeCreated = 1;
      }
      return _common.calculateTransactionDepositAndFeeV2.call(void 0, {
        numOfAtasToBeCreated,
        numOfOpenOrdersToBeCreated: 0,
        isInputMintSOL
      });
    }
    if (quoteResponse.inputMint.equals(_spltoken.NATIVE_MINT) && wrapUnwrapSOL && !mintToAtaMap.has(_spltoken.NATIVE_MINT.toBase58())) {
      atasToBeCreated.add(_spltoken.NATIVE_MINT.toBase58());
    }
    walkRoutePlan(quoteResponse.routePlan);
  }
  return _common.calculateTransactionDepositAndFeeV2.call(void 0, {
    numOfAtasToBeCreated: atasToBeCreated.size,
    numOfOpenOrdersToBeCreated: openOrdersToBeCreated.size,
    isInputMintSOL
  });
};

















exports.Errors = Errors; exports.JUPITER_FEE_OWNER = _common.JUPITER_FEE_OWNER; exports.JUPITER_PROGRAM_V4_ID = _common.JUPITER_PROGRAM_V4_ID; exports.JUPITER_PROGRAM_V6_ID = _common.JUPITER_PROGRAM_V6_ID; exports.JUPITER_V6_ERRORS = _common.JUPITER_V6_ERRORS; exports.JupiterProvider = JupiterProvider; exports.LAMPORTS_PER_SIGNATURE = _common.LAMPORTS_PER_SIGNATURE; exports.MARKETS_URL = _common.MARKETS_URL; exports.OPENBOOK_PROGRAM_ID = _common.OPENBOOK_PROGRAM_ID; exports.Owner = _common.Owner; exports.SwapMode = _common.SwapMode; exports.TOKEN_LIST_URL = _common.TOKEN_LIST_URL; exports.calculateFeeForSwap = calculateFeeForSwap; exports.useJupiter = useJupiter; exports.useJupiterClient = useJupiterClient; exports.useJupiterExchange = useJupiterExchange;
