// jupiverse-kit/dist/index.esm.js
'use client';

import * as React from 'react';
import React__default, { forwardRef, useState, useEffect, useMemo, useCallback, useRef, Suspense } from 'react';
import { RotateCcw, Settings, Coins, ChevronDown, Search, SquareArrowOutUpRightIcon, ArrowDown, Loader2, WalletIcon } from 'lucide-react';
import { useWallet, useConnection } from '@solana/wallet-adapter-react';
import { UnifiedWalletButton, ConnectionProvider, UnifiedWalletProvider } from '@jup-ag/wallet-adapter';
import * as Dialog from '@radix-ui/react-dialog';
import { create } from 'zustand';
import { Connection, VersionedTransaction, PublicKey, LAMPORTS_PER_SOL } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import { toast } from 'sonner';
import * as AllWalletAdapters from '@solana/wallet-adapter-wallets';
import { useWrappedReownAdapter } from '@jup-ag/jup-mobile-adapter';

function _arrayLikeToArray(r, a) {
  (null == a || a > r.length) && (a = r.length);
  for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e];
  return n;
}
function _arrayWithHoles(r) {
  if (Array.isArray(r)) return r;
}
function _arrayWithoutHoles(r) {
  if (Array.isArray(r)) return _arrayLikeToArray(r);
}
function asyncGeneratorStep(n, t, e, r, o, a, c) {
  try {
    var i = n[a](c),
      u = i.value;
  } catch (n) {
    return void e(n);
  }
  i.done ? t(u) : Promise.resolve(u).then(r, o);
}
function _asyncToGenerator(n) {
  return function () {
    var t = this,
      e = arguments;
    return new Promise(function (r, o) {
      var a = n.apply(t, e);
      function _next(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "next", n);
      }
      function _throw(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "throw", n);
      }
      _next(void 0);
    });
  };
}
function _defineProperty(e, r, t) {
  return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
    value: t,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : e[r] = t, e;
}
function _iterableToArray(r) {
  if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r);
}
function _iterableToArrayLimit(r, l) {
  var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"];
  if (null != t) {
    var e,
      n,
      i,
      u,
      a = [],
      f = !0,
      o = !1;
    try {
      if (i = (t = t.call(r)).next, 0 === l) {
        if (Object(t) !== t) return;
        f = !1;
      } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0);
    } catch (r) {
      o = !0, n = r;
    } finally {
      try {
        if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return;
      } finally {
        if (o) throw n;
      }
    }
    return a;
  }
}
function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _nonIterableSpread() {
  throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function ownKeys(e, r) {
  var t = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var o = Object.getOwnPropertySymbols(e);
    r && (o = o.filter(function (r) {
      return Object.getOwnPropertyDescriptor(e, r).enumerable;
    })), t.push.apply(t, o);
  }
  return t;
}
function _objectSpread2(e) {
  for (var r = 1; r < arguments.length; r++) {
    var t = null != arguments[r] ? arguments[r] : {};
    r % 2 ? ownKeys(Object(t), !0).forEach(function (r) {
      _defineProperty(e, r, t[r]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) {
      Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
    });
  }
  return e;
}
function _objectWithoutProperties(e, t) {
  if (null == e) return {};
  var o,
    r,
    i = _objectWithoutPropertiesLoose(e, t);
  if (Object.getOwnPropertySymbols) {
    var s = Object.getOwnPropertySymbols(e);
    for (r = 0; r < s.length; r++) o = s[r], t.includes(o) || {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]);
  }
  return i;
}
function _objectWithoutPropertiesLoose(r, e) {
  if (null == r) return {};
  var t = {};
  for (var n in r) if ({}.hasOwnProperty.call(r, n)) {
    if (e.includes(n)) continue;
    t[n] = r[n];
  }
  return t;
}
function _regeneratorRuntime() {
  _regeneratorRuntime = function () {
    return e;
  };
  var t,
    e = {},
    r = Object.prototype,
    n = r.hasOwnProperty,
    o = Object.defineProperty || function (t, e, r) {
      t[e] = r.value;
    },
    i = "function" == typeof Symbol ? Symbol : {},
    a = i.iterator || "@@iterator",
    c = i.asyncIterator || "@@asyncIterator",
    u = i.toStringTag || "@@toStringTag";
  function define(t, e, r) {
    return Object.defineProperty(t, e, {
      value: r,
      enumerable: !0,
      configurable: !0,
      writable: !0
    }), t[e];
  }
  try {
    define({}, "");
  } catch (t) {
    define = function (t, e, r) {
      return t[e] = r;
    };
  }
  function wrap(t, e, r, n) {
    var i = e && e.prototype instanceof Generator ? e : Generator,
      a = Object.create(i.prototype),
      c = new Context(n || []);
    return o(a, "_invoke", {
      value: makeInvokeMethod(t, r, c)
    }), a;
  }
  function tryCatch(t, e, r) {
    try {
      return {
        type: "normal",
        arg: t.call(e, r)
      };
    } catch (t) {
      return {
        type: "throw",
        arg: t
      };
    }
  }
  e.wrap = wrap;
  var h = "suspendedStart",
    l = "suspendedYield",
    f = "executing",
    s = "completed",
    y = {};
  function Generator() {}
  function GeneratorFunction() {}
  function GeneratorFunctionPrototype() {}
  var p = {};
  define(p, a, function () {
    return this;
  });
  var d = Object.getPrototypeOf,
    v = d && d(d(values([])));
  v && v !== r && n.call(v, a) && (p = v);
  var g = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(p);
  function defineIteratorMethods(t) {
    ["next", "throw", "return"].forEach(function (e) {
      define(t, e, function (t) {
        return this._invoke(e, t);
      });
    });
  }
  function AsyncIterator(t, e) {
    function invoke(r, o, i, a) {
      var c = tryCatch(t[r], t, o);
      if ("throw" !== c.type) {
        var u = c.arg,
          h = u.value;
        return h && "object" == typeof h && n.call(h, "__await") ? e.resolve(h.__await).then(function (t) {
          invoke("next", t, i, a);
        }, function (t) {
          invoke("throw", t, i, a);
        }) : e.resolve(h).then(function (t) {
          u.value = t, i(u);
        }, function (t) {
          return invoke("throw", t, i, a);
        });
      }
      a(c.arg);
    }
    var r;
    o(this, "_invoke", {
      value: function (t, n) {
        function callInvokeWithMethodAndArg() {
          return new e(function (e, r) {
            invoke(t, n, e, r);
          });
        }
        return r = r ? r.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg();
      }
    });
  }
  function makeInvokeMethod(e, r, n) {
    var o = h;
    return function (i, a) {
      if (o === f) throw Error("Generator is already running");
      if (o === s) {
        if ("throw" === i) throw a;
        return {
          value: t,
          done: !0
        };
      }
      for (n.method = i, n.arg = a;;) {
        var c = n.delegate;
        if (c) {
          var u = maybeInvokeDelegate(c, n);
          if (u) {
            if (u === y) continue;
            return u;
          }
        }
        if ("next" === n.method) n.sent = n._sent = n.arg;else if ("throw" === n.method) {
          if (o === h) throw o = s, n.arg;
          n.dispatchException(n.arg);
        } else "return" === n.method && n.abrupt("return", n.arg);
        o = f;
        var p = tryCatch(e, r, n);
        if ("normal" === p.type) {
          if (o = n.done ? s : l, p.arg === y) continue;
          return {
            value: p.arg,
            done: n.done
          };
        }
        "throw" === p.type && (o = s, n.method = "throw", n.arg = p.arg);
      }
    };
  }
  function maybeInvokeDelegate(e, r) {
    var n = r.method,
      o = e.iterator[n];
    if (o === t) return r.delegate = null, "throw" === n && e.iterator.return && (r.method = "return", r.arg = t, maybeInvokeDelegate(e, r), "throw" === r.method) || "return" !== n && (r.method = "throw", r.arg = new TypeError("The iterator does not provide a '" + n + "' method")), y;
    var i = tryCatch(o, e.iterator, r.arg);
    if ("throw" === i.type) return r.method = "throw", r.arg = i.arg, r.delegate = null, y;
    var a = i.arg;
    return a ? a.done ? (r[e.resultName] = a.value, r.next = e.nextLoc, "return" !== r.method && (r.method = "next", r.arg = t), r.delegate = null, y) : a : (r.method = "throw", r.arg = new TypeError("iterator result is not an object"), r.delegate = null, y);
  }
  function pushTryEntry(t) {
    var e = {
      tryLoc: t[0]
    };
    1 in t && (e.catchLoc = t[1]), 2 in t && (e.finallyLoc = t[2], e.afterLoc = t[3]), this.tryEntries.push(e);
  }
  function resetTryEntry(t) {
    var e = t.completion || {};
    e.type = "normal", delete e.arg, t.completion = e;
  }
  function Context(t) {
    this.tryEntries = [{
      tryLoc: "root"
    }], t.forEach(pushTryEntry, this), this.reset(!0);
  }
  function values(e) {
    if (e || "" === e) {
      var r = e[a];
      if (r) return r.call(e);
      if ("function" == typeof e.next) return e;
      if (!isNaN(e.length)) {
        var o = -1,
          i = function next() {
            for (; ++o < e.length;) if (n.call(e, o)) return next.value = e[o], next.done = !1, next;
            return next.value = t, next.done = !0, next;
          };
        return i.next = i;
      }
    }
    throw new TypeError(typeof e + " is not iterable");
  }
  return GeneratorFunction.prototype = GeneratorFunctionPrototype, o(g, "constructor", {
    value: GeneratorFunctionPrototype,
    configurable: !0
  }), o(GeneratorFunctionPrototype, "constructor", {
    value: GeneratorFunction,
    configurable: !0
  }), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, u, "GeneratorFunction"), e.isGeneratorFunction = function (t) {
    var e = "function" == typeof t && t.constructor;
    return !!e && (e === GeneratorFunction || "GeneratorFunction" === (e.displayName || e.name));
  }, e.mark = function (t) {
    return Object.setPrototypeOf ? Object.setPrototypeOf(t, GeneratorFunctionPrototype) : (t.__proto__ = GeneratorFunctionPrototype, define(t, u, "GeneratorFunction")), t.prototype = Object.create(g), t;
  }, e.awrap = function (t) {
    return {
      __await: t
    };
  }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, c, function () {
    return this;
  }), e.AsyncIterator = AsyncIterator, e.async = function (t, r, n, o, i) {
    void 0 === i && (i = Promise);
    var a = new AsyncIterator(wrap(t, r, n, o), i);
    return e.isGeneratorFunction(r) ? a : a.next().then(function (t) {
      return t.done ? t.value : a.next();
    });
  }, defineIteratorMethods(g), define(g, u, "Generator"), define(g, a, function () {
    return this;
  }), define(g, "toString", function () {
    return "[object Generator]";
  }), e.keys = function (t) {
    var e = Object(t),
      r = [];
    for (var n in e) r.push(n);
    return r.reverse(), function next() {
      for (; r.length;) {
        var t = r.pop();
        if (t in e) return next.value = t, next.done = !1, next;
      }
      return next.done = !0, next;
    };
  }, e.values = values, Context.prototype = {
    constructor: Context,
    reset: function (e) {
      if (this.prev = 0, this.next = 0, this.sent = this._sent = t, this.done = !1, this.delegate = null, this.method = "next", this.arg = t, this.tryEntries.forEach(resetTryEntry), !e) for (var r in this) "t" === r.charAt(0) && n.call(this, r) && !isNaN(+r.slice(1)) && (this[r] = t);
    },
    stop: function () {
      this.done = !0;
      var t = this.tryEntries[0].completion;
      if ("throw" === t.type) throw t.arg;
      return this.rval;
    },
    dispatchException: function (e) {
      if (this.done) throw e;
      var r = this;
      function handle(n, o) {
        return a.type = "throw", a.arg = e, r.next = n, o && (r.method = "next", r.arg = t), !!o;
      }
      for (var o = this.tryEntries.length - 1; o >= 0; --o) {
        var i = this.tryEntries[o],
          a = i.completion;
        if ("root" === i.tryLoc) return handle("end");
        if (i.tryLoc <= this.prev) {
          var c = n.call(i, "catchLoc"),
            u = n.call(i, "finallyLoc");
          if (c && u) {
            if (this.prev < i.catchLoc) return handle(i.catchLoc, !0);
            if (this.prev < i.finallyLoc) return handle(i.finallyLoc);
          } else if (c) {
            if (this.prev < i.catchLoc) return handle(i.catchLoc, !0);
          } else {
            if (!u) throw Error("try statement without catch or finally");
            if (this.prev < i.finallyLoc) return handle(i.finallyLoc);
          }
        }
      }
    },
    abrupt: function (t, e) {
      for (var r = this.tryEntries.length - 1; r >= 0; --r) {
        var o = this.tryEntries[r];
        if (o.tryLoc <= this.prev && n.call(o, "finallyLoc") && this.prev < o.finallyLoc) {
          var i = o;
          break;
        }
      }
      i && ("break" === t || "continue" === t) && i.tryLoc <= e && e <= i.finallyLoc && (i = null);
      var a = i ? i.completion : {};
      return a.type = t, a.arg = e, i ? (this.method = "next", this.next = i.finallyLoc, y) : this.complete(a);
    },
    complete: function (t, e) {
      if ("throw" === t.type) throw t.arg;
      return "break" === t.type || "continue" === t.type ? this.next = t.arg : "return" === t.type ? (this.rval = this.arg = t.arg, this.method = "return", this.next = "end") : "normal" === t.type && e && (this.next = e), y;
    },
    finish: function (t) {
      for (var e = this.tryEntries.length - 1; e >= 0; --e) {
        var r = this.tryEntries[e];
        if (r.finallyLoc === t) return this.complete(r.completion, r.afterLoc), resetTryEntry(r), y;
      }
    },
    catch: function (t) {
      for (var e = this.tryEntries.length - 1; e >= 0; --e) {
        var r = this.tryEntries[e];
        if (r.tryLoc === t) {
          var n = r.completion;
          if ("throw" === n.type) {
            var o = n.arg;
            resetTryEntry(r);
          }
          return o;
        }
      }
      throw Error("illegal catch attempt");
    },
    delegateYield: function (e, r, n) {
      return this.delegate = {
        iterator: values(e),
        resultName: r,
        nextLoc: n
      }, "next" === this.method && (this.arg = t), y;
    }
  }, e;
}
function _slicedToArray(r, e) {
  return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest();
}
function _toConsumableArray(r) {
  return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread();
}
function _toPrimitive(t, r) {
  if ("object" != typeof t || !t) return t;
  var e = t[Symbol.toPrimitive];
  if (void 0 !== e) {
    var i = e.call(t, r || "default");
    if ("object" != typeof i) return i;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return ("string" === r ? String : Number)(t);
}
function _toPropertyKey(t) {
  var i = _toPrimitive(t, "string");
  return "symbol" == typeof i ? i : i + "";
}
function _unsupportedIterableToArray(r, a) {
  if (r) {
    if ("string" == typeof r) return _arrayLikeToArray(r, a);
    var t = {}.toString.call(r).slice(8, -1);
    return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0;
  }
}

var _path$1, _path2$1, _path3$1, _path4$1, _path5$1, _path6$1, _path7$1, _path8$1, _defs$1;
function _extends$1() { return _extends$1 = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends$1.apply(null, arguments); }
var SvgPoweredbyjupiterBright = function SvgPoweredbyjupiterBright(props, ref) {
  return /*#__PURE__*/React.createElement("svg", _extends$1({
    width: 152,
    height: 24,
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    ref: ref
  }, props), _path$1 || (_path$1 = /*#__PURE__*/React.createElement("path", {
    d: "M0 15.1364H1.05682V11.9489H2.96591C4.99858 11.9489 5.91477 10.7131 5.91477 9.17052C5.91477 7.62792 4.99858 6.40912 2.94886 6.40912H0V15.1364ZM1.05682 11.0114V7.34662H2.91477C4.3338 7.34662 4.875 8.12222 4.875 9.17052C4.875 10.2188 4.3338 11.0114 2.93182 11.0114H1.05682ZM10.0696 15.2728C11.8423 15.2728 13.0355 13.9262 13.0355 11.8978C13.0355 9.85232 11.8423 8.50572 10.0696 8.50572C8.29687 8.50572 7.10369 9.85232 7.10369 11.8978C7.10369 13.9262 8.29687 15.2728 10.0696 15.2728ZM10.0696 14.3694C8.72301 14.3694 8.10937 13.2103 8.10937 11.8978C8.10937 10.5853 8.72301 9.40912 10.0696 9.40912C11.4162 9.40912 12.0298 10.5853 12.0298 11.8978C12.0298 13.2103 11.4162 14.3694 10.0696 14.3694ZM15.831 15.1364H16.8196L18.2514 10.108H18.3537L19.7855 15.1364H20.7742L22.7685 8.59092H21.7117L20.2969 13.5853H20.2287L18.848 8.59092H17.7742L16.3764 13.6023H16.3083L14.8935 8.59092H13.8367L15.831 15.1364ZM26.6197 15.2728C27.9492 15.2728 28.9208 14.608 29.2276 13.6194L28.2561 13.3466C28.0004 14.0284 27.408 14.3694 26.6197 14.3694C25.4393 14.3694 24.6254 13.6066 24.5785 12.2046H29.3299V11.7784C29.3299 9.34092 27.8811 8.50572 26.5174 8.50572C24.7447 8.50572 23.5686 9.90342 23.5686 11.9148C23.5686 13.9262 24.7276 15.2728 26.6197 15.2728ZM24.5785 11.3353C24.6467 10.3168 25.3669 9.40912 26.5174 9.40912C27.6083 9.40912 28.3072 10.2273 28.3072 11.3353H24.5785ZM30.8597 15.1364H31.8654V10.9944C31.8654 10.108 32.5643 9.46032 33.5188 9.46032C33.7873 9.46032 34.0643 9.51142 34.1325 9.52842V8.50572C34.0174 8.49722 33.7532 8.48872 33.6041 8.48872C32.82 8.48872 32.1382 8.93192 31.8995 9.57962H31.8313V8.59092H30.8597V15.1364ZM37.8697 15.2728C39.1992 15.2728 40.1708 14.608 40.4776 13.6194L39.5061 13.3466C39.2504 14.0284 38.658 14.3694 37.8697 14.3694C36.6893 14.3694 35.8754 13.6066 35.8285 12.2046H40.5799V11.7784C40.5799 9.34092 39.1311 8.50572 37.7674 8.50572C35.9947 8.50572 34.8186 9.90342 34.8186 11.9148C34.8186 13.9262 35.9776 15.2728 37.8697 15.2728ZM35.8285 11.3353C35.8967 10.3168 36.6169 9.40912 37.7674 9.40912C38.8583 9.40912 39.5572 10.2273 39.5572 11.3353H35.8285ZM44.5813 15.2728C45.8427 15.2728 46.2688 14.4887 46.4904 14.1307H46.6097V15.1364H47.5813V6.40912H46.5757V9.63072H46.4904C46.2688 9.28982 45.8768 8.50572 44.5984 8.50572C42.945 8.50572 41.8029 9.81822 41.8029 11.8807C41.8029 13.9603 42.945 15.2728 44.5813 15.2728ZM44.7177 14.3694C43.4563 14.3694 42.8086 13.2614 42.8086 11.8637C42.8086 10.483 43.4393 9.40912 44.7177 9.40912C45.945 9.40912 46.5927 10.3978 46.5927 11.8637C46.5927 13.3466 45.9279 14.3694 44.7177 14.3694ZM53.0742 15.1364H54.0458V14.1307H54.1651C54.3867 14.4887 54.8129 15.2728 56.0742 15.2728C57.7106 15.2728 58.8526 13.9603 58.8526 11.8807C58.8526 9.81822 57.7106 8.50572 56.0572 8.50572C54.7788 8.50572 54.3867 9.28982 54.1651 9.63072H54.0799V6.40912H53.0742V15.1364ZM54.0629 11.8637C54.0629 10.3978 54.7106 9.40912 55.9379 9.40912C57.2163 9.40912 57.847 10.483 57.847 11.8637C57.847 13.2614 57.1992 14.3694 55.9379 14.3694C54.7276 14.3694 54.0629 13.3466 54.0629 11.8637ZM60.8044 17.5909C61.712 17.5909 62.3512 17.1094 62.7305 16.0909L65.5174 8.60802L64.418 8.59092L62.6112 13.8069H62.543L60.7362 8.59092H59.6453L62.0657 15.1705L61.8782 15.6819C61.5117 16.679 61.06 16.7728 60.3271 16.5853L60.0714 17.4716C60.1737 17.5228 60.4634 17.5909 60.8044 17.5909Z",
    fill: "black",
    fillOpacity: 0.45
  })), _path2$1 || (_path2$1 = /*#__PURE__*/React.createElement("path", {
    d: "M75.8355 18.875C76.8494 20.2853 78.15 21.4654 79.652 22.3378C81.154 23.2103 82.8235 23.7554 84.5509 23.9374C83.6622 22.6 82.3703 21.3696 80.7575 20.4328C79.1448 19.4961 77.4368 18.9844 75.8355 18.875Z",
    fill: "url(#paint0_linear_13278_146917)"
  })), _path3$1 || (_path3$1 = /*#__PURE__*/React.createElement("path", {
    d: "M82.9247 16.7029C79.8175 14.8977 76.4545 14.4377 74.0481 15.2504C74.2804 16.0181 74.5868 16.7615 74.9629 17.47C77.0536 17.4216 79.3363 17.9894 81.4604 19.2231C83.5844 20.4568 85.209 22.1594 86.2031 24C87.0053 23.9754 87.8033 23.8732 88.5858 23.6949C88.0994 21.2024 86.031 18.5085 82.9247 16.7029Z",
    fill: "url(#paint1_linear_13278_146917)"
  })), _path4$1 || (_path4$1 = /*#__PURE__*/React.createElement("path", {
    d: "M97.7313 9.37567C97.3363 7.7688 96.6215 6.25802 95.6296 4.93353C94.6377 3.60904 93.3891 2.498 91.9583 1.66676C90.5276 0.835525 88.944 0.301128 87.3021 0.0954825C85.6602 -0.110163 83.9937 0.01716 82.4021 0.469849C85.061 0.79525 88.0124 1.79342 90.9069 3.47493C93.8013 5.15644 96.1324 7.22558 97.7313 9.37567Z",
    fill: "url(#paint2_linear_13278_146917)"
  })), _path5$1 || (_path5$1 = /*#__PURE__*/React.createElement("path", {
    d: "M93.8627 15.2687C92.5017 13.0084 90.1702 10.8441 87.2982 9.17561C84.4261 7.50712 81.3917 6.5541 78.7564 6.49105C76.4379 6.43614 74.6978 7.11013 73.9836 8.33974C73.9795 8.34706 73.9738 8.35398 73.9693 8.3613C73.9051 8.59193 73.8497 8.82296 73.7993 9.05481C74.7967 8.66107 75.9522 8.44184 77.2372 8.41743C80.0946 8.36374 83.2925 9.27771 86.2434 10.9922C89.1944 12.7066 91.5743 15.0328 92.9426 17.5409C93.556 18.6708 93.9388 19.7833 94.0905 20.8461C94.267 20.6883 94.4403 20.5252 94.6083 20.3556C94.6128 20.3479 94.6156 20.3397 94.6201 20.3316C95.3343 19.1008 95.0582 17.2557 93.8627 15.2687Z",
    fill: "url(#paint3_linear_13278_146917)"
  })), _path6$1 || (_path6$1 = /*#__PURE__*/React.createElement("path", {
    d: "M85.1131 12.9364C80.7149 10.3812 75.8546 9.98094 73.5174 11.7641C73.522 12.3223 73.5647 12.8795 73.6451 13.4318C74.3326 13.2234 75.0412 13.0922 75.7578 13.0405C78.3695 12.8441 81.2489 13.5717 83.8619 15.0905C86.4749 16.6094 88.5347 18.7513 89.6581 21.1158C89.9686 21.7635 90.2055 22.4441 90.3642 23.1446C90.884 22.9409 91.3894 22.7021 91.8769 22.43C92.2682 19.516 89.5125 15.492 85.1131 12.9364Z",
    fill: "url(#paint4_linear_13278_146917)"
  })), _path7$1 || (_path7$1 = /*#__PURE__*/React.createElement("path", {
    d: "M96.125 11.4856C94.7485 9.22769 92.3967 7.05727 89.5043 5.37779C86.6119 3.69831 83.5657 2.73025 80.9218 2.65256C78.9064 2.59439 77.3453 3.0829 76.5358 4.00582C79.8968 3.43637 84.33 4.39345 88.6269 6.88969C92.9238 9.38592 95.9525 12.7636 97.1215 15.9651C97.5222 14.8055 97.1736 13.2073 96.125 11.4856Z",
    fill: "url(#paint5_linear_13278_146917)"
  })), _path8$1 || (_path8$1 = /*#__PURE__*/React.createElement("path", {
    d: "M108.516 6.68182L110.568 6.68182L110.568 13.5277C110.568 14.1605 110.425 14.7102 110.141 15.1768C109.86 15.6435 109.468 16.003 108.966 16.2555C108.465 16.508 107.881 16.6342 107.217 16.6342C106.625 16.6342 106.088 16.5304 105.606 16.3226C105.126 16.1117 104.746 15.7921 104.465 15.3638C104.184 14.9324 104.045 14.3906 104.048 13.7386L106.114 13.7386C106.12 13.9975 106.173 14.2196 106.272 14.405C106.374 14.5872 106.513 14.7278 106.689 14.8269C106.868 14.9228 107.079 14.9707 107.322 14.9707C107.578 14.9707 107.793 14.9164 107.969 14.8077C108.148 14.6958 108.284 14.5328 108.377 14.3187C108.469 14.1046 108.516 13.8409 108.516 13.5277L108.516 6.68182ZM116.968 13.3647L116.968 9.13636L119.01 9.13636L119.01 16.5L117.049 16.5L117.049 15.1625L116.972 15.1625C116.806 15.5939 116.53 15.9407 116.143 16.2028C115.76 16.4648 115.291 16.5959 114.738 16.5959C114.246 16.5959 113.813 16.484 113.439 16.2603C113.065 16.0366 112.773 15.7186 112.562 15.3063C112.354 14.894 112.249 14.4002 112.246 13.8249L112.246 9.13636L114.288 9.13636L114.288 13.4606C114.291 13.8952 114.408 14.2388 114.638 14.4913C114.868 14.7438 115.176 14.87 115.563 14.87C115.809 14.87 116.039 14.8141 116.253 14.7022C116.467 14.5872 116.64 14.4178 116.771 14.1941C116.905 13.9703 116.971 13.6939 116.968 13.3647ZM120.643 19.2614L120.643 9.13636L122.657 9.13636L122.657 10.3732L122.748 10.3732C122.838 10.1751 122.967 9.97372 123.136 9.76918C123.309 9.56144 123.533 9.38885 123.808 9.25142C124.086 9.1108 124.431 9.04048 124.843 9.04048C125.38 9.04048 125.875 9.18111 126.329 9.46236C126.783 9.74041 127.146 10.1607 127.417 10.7232C127.689 11.2825 127.825 11.984 127.825 12.8278C127.825 13.6491 127.692 14.3427 127.427 14.9084C127.165 15.4709 126.807 15.8976 126.353 16.1884C125.903 16.476 125.398 16.6199 124.838 16.6199C124.442 16.6199 124.105 16.5543 123.827 16.4233C123.552 16.2923 123.327 16.1277 123.151 15.9295C122.975 15.7282 122.841 15.5252 122.748 15.3207L122.686 15.3207L122.686 19.2614L120.643 19.2614ZM122.643 12.8182C122.643 13.256 122.703 13.638 122.825 13.964C122.946 14.29 123.122 14.544 123.352 14.7262C123.582 14.9052 123.862 14.9947 124.191 14.9947C124.523 14.9947 124.805 14.9036 125.035 14.7214C125.265 14.536 125.439 14.2804 125.557 13.9544C125.679 13.6252 125.74 13.2464 125.74 12.8182C125.74 12.3931 125.68 12.0192 125.562 11.6964C125.444 11.3736 125.27 11.1211 125.04 10.9389C124.809 10.7567 124.527 10.6657 124.191 10.6657C123.859 10.6657 123.577 10.7536 123.347 10.9293C123.12 11.1051 122.946 11.3544 122.825 11.6772C122.703 12 122.643 12.3803 122.643 12.8182ZM129.186 16.5L129.186 9.13636L131.229 9.13636L131.229 16.5L129.186 16.5ZM130.212 8.18715C129.909 8.18715 129.648 8.08647 129.431 7.88512C129.217 7.68058 129.11 7.43608 129.11 7.15163C129.11 6.87038 129.217 6.62909 129.431 6.42774C129.648 6.22319 129.909 6.12092 130.212 6.12092C130.516 6.12092 130.775 6.22319 130.989 6.42774C131.206 6.62909 131.315 6.87038 131.315 7.15163C131.315 7.43608 131.206 7.68058 130.989 7.88512C130.775 8.08647 130.516 8.18715 130.212 8.18715ZM136.781 9.13637L136.781 10.6705L132.347 10.6705L132.347 9.13636L136.781 9.13637ZM133.354 7.37216L135.396 7.37216L135.396 14.2372C135.396 14.4258 135.425 14.5728 135.482 14.6783C135.54 14.7805 135.62 14.8525 135.722 14.894C135.827 14.9355 135.949 14.9563 136.086 14.9563C136.182 14.9563 136.278 14.9483 136.374 14.9324C136.47 14.9132 136.543 14.8988 136.594 14.8892L136.916 16.4089C136.813 16.4409 136.67 16.4776 136.484 16.5192C136.299 16.5639 136.073 16.5911 135.808 16.6007C135.316 16.6199 134.885 16.5543 134.514 16.4041C134.146 16.2539 133.86 16.0206 133.656 15.7042C133.451 15.3878 133.35 14.9883 133.354 14.5057L133.354 7.37216ZM141.388 16.6438C140.631 16.6438 139.979 16.4904 139.432 16.1836C138.889 15.8736 138.47 15.4357 138.176 14.87C137.882 14.3011 137.735 13.6284 137.735 12.8517C137.735 12.0943 137.882 11.4295 138.176 10.8574C138.47 10.2853 138.884 9.83949 139.418 9.51989C139.955 9.20029 140.585 9.04048 141.307 9.04048C141.793 9.04048 142.245 9.11879 142.664 9.27539C143.086 9.4288 143.453 9.66051 143.766 9.97053C144.083 10.2805 144.329 10.6705 144.505 11.1403C144.68 11.6069 144.768 12.1534 144.768 12.7798L144.768 13.3407L138.55 13.3407L138.55 12.0751L142.846 12.0751C142.846 11.7811 142.782 11.5206 142.654 11.2937C142.526 11.0668 142.349 10.8894 142.122 10.7615C141.898 10.6305 141.638 10.565 141.341 10.565C141.031 10.565 140.756 10.6369 140.516 10.7807C140.279 10.9213 140.094 11.1115 139.96 11.3512C139.826 11.5877 139.757 11.8514 139.754 12.1422L139.754 13.3455C139.754 13.7099 139.821 14.0247 139.955 14.29C140.092 14.5552 140.286 14.7598 140.535 14.9036C140.784 15.0474 141.08 15.1193 141.422 15.1193C141.649 15.1193 141.857 15.0874 142.045 15.0234C142.234 14.9595 142.395 14.8636 142.529 14.7358C142.664 14.608 142.766 14.4514 142.836 14.266L144.725 14.3906C144.629 14.8445 144.433 15.2408 144.135 15.5795C143.841 15.9151 143.461 16.1772 142.994 16.3658C142.531 16.5511 141.996 16.6438 141.388 16.6438ZM146.101 16.5L146.101 9.13637L148.081 9.13637L148.081 10.4212L148.158 10.4212C148.292 9.96414 148.517 9.61896 148.834 9.38566C149.15 9.14915 149.514 9.0309 149.927 9.0309C150.029 9.0309 150.139 9.03729 150.257 9.05007C150.376 9.06286 150.48 9.08044 150.569 9.10281L150.569 10.915C150.473 10.8862 150.341 10.8606 150.171 10.8382C150.002 10.8159 149.847 10.8047 149.706 10.8047C149.406 10.8047 149.137 10.8702 148.901 11.0012C148.667 11.1291 148.482 11.3081 148.345 11.5382C148.21 11.7683 148.143 12.0336 148.143 12.334L148.143 16.5L146.101 16.5Z",
    fill: "#111726"
  })), _defs$1 || (_defs$1 = /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear_13278_146917",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#A4D756"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00B6E7"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint1_linear_13278_146917",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#A4D756"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00B6E7"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint2_linear_13278_146917",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#A4D756"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00B6E7"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint3_linear_13278_146917",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#A4D756"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00B6E7"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint4_linear_13278_146917",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#A4D756"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00B6E7"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint5_linear_13278_146917",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#A4D756"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00B6E7"
  })))));
};
var ForwardRef$1 = /*#__PURE__*/forwardRef(SvgPoweredbyjupiterBright);

var _path, _path2, _path3, _path4, _path5, _path6, _path7, _path8, _defs;
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
var SvgPoweredbyjupiterDark = function SvgPoweredbyjupiterDark(props, ref) {
  return /*#__PURE__*/React.createElement("svg", _extends({
    width: 151,
    height: 24,
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    ref: ref
  }, props), _path || (_path = /*#__PURE__*/React.createElement("path", {
    d: "M0 15.1364H1.05682V11.9489H2.96591C4.99858 11.9489 5.91477 10.7131 5.91477 9.17052C5.91477 7.62792 4.99858 6.40912 2.94886 6.40912H0V15.1364ZM1.05682 11.0114V7.34662H2.91477C4.3338 7.34662 4.875 8.12222 4.875 9.17052C4.875 10.2188 4.3338 11.0114 2.93182 11.0114H1.05682ZM10.0696 15.2728C11.8423 15.2728 13.0355 13.9262 13.0355 11.8978C13.0355 9.85232 11.8423 8.50572 10.0696 8.50572C8.29687 8.50572 7.10369 9.85232 7.10369 11.8978C7.10369 13.9262 8.29687 15.2728 10.0696 15.2728ZM10.0696 14.3694C8.72301 14.3694 8.10937 13.2103 8.10937 11.8978C8.10937 10.5853 8.72301 9.40912 10.0696 9.40912C11.4162 9.40912 12.0298 10.5853 12.0298 11.8978C12.0298 13.2103 11.4162 14.3694 10.0696 14.3694ZM15.831 15.1364H16.8196L18.2514 10.108H18.3537L19.7855 15.1364H20.7742L22.7685 8.59092H21.7117L20.2969 13.5853H20.2287L18.848 8.59092H17.7742L16.3764 13.6023H16.3083L14.8935 8.59092H13.8367L15.831 15.1364ZM26.6197 15.2728C27.9492 15.2728 28.9208 14.608 29.2276 13.6194L28.2561 13.3466C28.0004 14.0284 27.408 14.3694 26.6197 14.3694C25.4393 14.3694 24.6254 13.6066 24.5785 12.2046H29.3299V11.7784C29.3299 9.34092 27.8811 8.50572 26.5174 8.50572C24.7447 8.50572 23.5686 9.90342 23.5686 11.9148C23.5686 13.9262 24.7276 15.2728 26.6197 15.2728ZM24.5785 11.3353C24.6467 10.3168 25.3669 9.40912 26.5174 9.40912C27.6083 9.40912 28.3072 10.2273 28.3072 11.3353H24.5785ZM30.8597 15.1364H31.8654V10.9944C31.8654 10.108 32.5643 9.46032 33.5188 9.46032C33.7873 9.46032 34.0643 9.51142 34.1325 9.52842V8.50572C34.0174 8.49722 33.7532 8.48872 33.6041 8.48872C32.82 8.48872 32.1382 8.93192 31.8995 9.57962H31.8313V8.59092H30.8597V15.1364ZM37.8697 15.2728C39.1992 15.2728 40.1708 14.608 40.4776 13.6194L39.5061 13.3466C39.2504 14.0284 38.658 14.3694 37.8697 14.3694C36.6893 14.3694 35.8754 13.6066 35.8285 12.2046H40.5799V11.7784C40.5799 9.34092 39.1311 8.50572 37.7674 8.50572C35.9947 8.50572 34.8186 9.90342 34.8186 11.9148C34.8186 13.9262 35.9776 15.2728 37.8697 15.2728ZM35.8285 11.3353C35.8967 10.3168 36.6169 9.40912 37.7674 9.40912C38.8583 9.40912 39.5572 10.2273 39.5572 11.3353H35.8285ZM44.5813 15.2728C45.8427 15.2728 46.2688 14.4887 46.4904 14.1307H46.6097V15.1364H47.5813V6.40912H46.5757V9.63072H46.4904C46.2688 9.28982 45.8768 8.50572 44.5984 8.50572C42.945 8.50572 41.8029 9.81822 41.8029 11.8807C41.8029 13.9603 42.945 15.2728 44.5813 15.2728ZM44.7177 14.3694C43.4563 14.3694 42.8086 13.2614 42.8086 11.8637C42.8086 10.483 43.4393 9.40912 44.7177 9.40912C45.945 9.40912 46.5927 10.3978 46.5927 11.8637C46.5927 13.3466 45.9279 14.3694 44.7177 14.3694ZM53.0742 15.1364H54.0458V14.1307H54.1651C54.3867 14.4887 54.8129 15.2728 56.0742 15.2728C57.7106 15.2728 58.8526 13.9603 58.8526 11.8807C58.8526 9.81822 57.7106 8.50572 56.0572 8.50572C54.7788 8.50572 54.3867 9.28982 54.1651 9.63072H54.0799V6.40912H53.0742V15.1364ZM54.0629 11.8637C54.0629 10.3978 54.7106 9.40912 55.9379 9.40912C57.2163 9.40912 57.847 10.483 57.847 11.8637C57.847 13.2614 57.1992 14.3694 55.9379 14.3694C54.7276 14.3694 54.0629 13.3466 54.0629 11.8637ZM60.8044 17.5909C61.712 17.5909 62.3512 17.1094 62.7305 16.0909L65.5174 8.60802L64.418 8.59092L62.6112 13.8069H62.543L60.7362 8.59092H59.6453L62.0657 15.1705L61.8782 15.6819C61.5117 16.679 61.06 16.7728 60.3271 16.5853L60.0714 17.4716C60.1737 17.5228 60.4634 17.5909 60.8044 17.5909Z",
    fill: "white",
    fillOpacity: 0.35
  })), _path2 || (_path2 = /*#__PURE__*/React.createElement("path", {
    d: "M75.8354 18.875C76.8494 20.2853 78.15 21.4654 79.652 22.3378C81.154 23.2103 82.8235 23.7554 84.5509 23.9374C83.6621 22.6 82.3703 21.3696 80.7575 20.4328C79.1448 19.4961 77.4368 18.9844 75.8354 18.875Z",
    fill: "url(#paint0_linear_13278_146910)"
  })), _path3 || (_path3 = /*#__PURE__*/React.createElement("path", {
    d: "M82.9247 16.7029C79.8175 14.8977 76.4545 14.4377 74.0481 15.2504C74.2804 16.0181 74.5868 16.7615 74.9629 17.47C77.0536 17.4216 79.3363 17.9894 81.4604 19.2231C83.5844 20.4568 85.209 22.1594 86.2031 24C87.0053 23.9754 87.8033 23.8732 88.5858 23.6949C88.0994 21.2024 86.031 18.5085 82.9247 16.7029Z",
    fill: "url(#paint1_linear_13278_146910)"
  })), _path4 || (_path4 = /*#__PURE__*/React.createElement("path", {
    d: "M97.7313 9.37567C97.3363 7.7688 96.6215 6.25802 95.6296 4.93353C94.6377 3.60904 93.3891 2.498 91.9583 1.66676C90.5275 0.835525 88.9439 0.301128 87.302 0.0954825C85.6602 -0.110163 83.9937 0.01716 82.4021 0.469849C85.061 0.79525 88.0124 1.79342 90.9068 3.47493C93.8013 5.15644 96.1324 7.22558 97.7313 9.37567Z",
    fill: "url(#paint2_linear_13278_146910)"
  })), _path5 || (_path5 = /*#__PURE__*/React.createElement("path", {
    d: "M93.8627 15.2687C92.5017 13.0084 90.1702 10.8441 87.2981 9.17561C84.4261 7.50712 81.3917 6.5541 78.7564 6.49105C76.4379 6.43614 74.6978 7.11013 73.9835 8.33974C73.9795 8.34706 73.9738 8.35398 73.9693 8.3613C73.905 8.59193 73.8497 8.82296 73.7993 9.05481C74.7966 8.66107 75.9522 8.44184 77.2371 8.41743C80.0946 8.36374 83.2925 9.27771 86.2434 10.9922C89.1944 12.7066 91.5743 15.0328 92.9426 17.5409C93.556 18.6708 93.9388 19.7833 94.0905 20.8461C94.267 20.6883 94.4403 20.5252 94.6083 20.3556C94.6128 20.3479 94.6156 20.3397 94.6201 20.3316C95.3343 19.1008 95.0581 17.2557 93.8627 15.2687Z",
    fill: "url(#paint3_linear_13278_146910)"
  })), _path6 || (_path6 = /*#__PURE__*/React.createElement("path", {
    d: "M85.113 12.9364C80.7148 10.3812 75.8546 9.98094 73.5174 11.7641C73.522 12.3223 73.5646 12.8795 73.6451 13.4318C74.3326 13.2234 75.0412 13.0922 75.7578 13.0405C78.3695 12.8441 81.2489 13.5717 83.8619 15.0905C86.4749 16.6094 88.5346 18.7513 89.6581 21.1158C89.9686 21.7635 90.2055 22.4441 90.3642 23.1446C90.884 22.9409 91.3894 22.7021 91.8769 22.43C92.2682 19.516 89.5125 15.492 85.113 12.9364Z",
    fill: "url(#paint4_linear_13278_146910)"
  })), _path7 || (_path7 = /*#__PURE__*/React.createElement("path", {
    d: "M96.125 11.4856C94.7485 9.22769 92.3967 7.05727 89.5043 5.37779C86.6119 3.69831 83.5657 2.73025 80.9218 2.65256C78.9064 2.59439 77.3452 3.0829 76.5358 4.00582C79.8968 3.43637 84.33 4.39345 88.6269 6.88969C92.9238 9.38592 95.9525 12.7636 97.1215 15.9651C97.5221 14.8055 97.1736 13.2073 96.125 11.4856Z",
    fill: "url(#paint5_linear_13278_146910)"
  })), _path8 || (_path8 = /*#__PURE__*/React.createElement("path", {
    d: "M108.516 7.43182L110.568 7.43182L110.568 14.2777C110.568 14.9105 110.425 15.4602 110.141 15.9268C109.86 16.3935 109.468 16.753 108.966 17.0055C108.465 17.258 107.881 17.3842 107.217 17.3842C106.625 17.3842 106.088 17.2804 105.606 17.0726C105.126 16.8617 104.746 16.5421 104.465 16.1138C104.184 15.6824 104.045 15.1406 104.048 14.4886L106.114 14.4886C106.12 14.7475 106.173 14.9696 106.272 15.155C106.374 15.3372 106.513 15.4778 106.689 15.5769C106.868 15.6728 107.079 15.7207 107.322 15.7207C107.578 15.7207 107.793 15.6664 107.969 15.5577C108.148 15.4458 108.284 15.2828 108.377 15.0687C108.469 14.8546 108.516 14.5909 108.516 14.2777L108.516 7.43182ZM116.968 14.1147L116.968 9.88636L119.01 9.88636L119.01 17.25L117.049 17.25L117.049 15.9125L116.972 15.9125C116.806 16.3439 116.53 16.6907 116.143 16.9528C115.76 17.2148 115.291 17.3459 114.738 17.3459C114.246 17.3459 113.813 17.234 113.439 17.0103C113.065 16.7866 112.773 16.4686 112.562 16.0563C112.354 15.644 112.249 15.1502 112.246 14.5749L112.246 9.88636L114.288 9.88636L114.288 14.2106C114.291 14.6452 114.408 14.9888 114.638 15.2413C114.868 15.4938 115.176 15.62 115.563 15.62C115.809 15.62 116.039 15.5641 116.253 15.4522C116.467 15.3372 116.64 15.1678 116.771 14.9441C116.905 14.7203 116.971 14.4439 116.968 14.1147ZM120.643 20.0114L120.643 9.88636L122.657 9.88636L122.657 11.1232L122.748 11.1232C122.838 10.9251 122.967 10.7237 123.136 10.5192C123.309 10.3114 123.533 10.1389 123.808 10.0014C124.086 9.8608 124.431 9.79048 124.843 9.79048C125.38 9.79048 125.875 9.93111 126.329 10.2124C126.783 10.4904 127.146 10.9107 127.417 11.4732C127.689 12.0325 127.825 12.734 127.825 13.5778C127.825 14.3991 127.692 15.0927 127.427 15.6584C127.165 16.2209 126.807 16.6476 126.353 16.9384C125.903 17.226 125.398 17.3699 124.838 17.3699C124.442 17.3699 124.105 17.3043 123.827 17.1733C123.552 17.0423 123.327 16.8777 123.151 16.6795C122.975 16.4782 122.841 16.2752 122.748 16.0707L122.686 16.0707L122.686 20.0114L120.643 20.0114ZM122.643 13.5682C122.643 14.006 122.703 14.388 122.825 14.714C122.946 15.04 123.122 15.294 123.352 15.4762C123.582 15.6552 123.862 15.7447 124.191 15.7447C124.523 15.7447 124.805 15.6536 125.035 15.4714C125.265 15.286 125.439 15.0304 125.557 14.7044C125.679 14.3752 125.74 13.9964 125.74 13.5682C125.74 13.1431 125.68 12.7692 125.562 12.4464C125.444 12.1236 125.27 11.8711 125.04 11.6889C124.809 11.5067 124.527 11.4157 124.191 11.4157C123.859 11.4157 123.577 11.5036 123.347 11.6793C123.12 11.8551 122.946 12.1044 122.825 12.4272C122.703 12.75 122.643 13.1303 122.643 13.5682ZM129.186 17.25L129.186 9.88636L131.229 9.88636L131.229 17.25L129.186 17.25ZM130.212 8.93715C129.909 8.93715 129.648 8.83647 129.431 8.63512C129.217 8.43058 129.11 8.18608 129.11 7.90163C129.11 7.62038 129.217 7.37908 129.431 7.17774C129.648 6.97319 129.909 6.87092 130.212 6.87092C130.516 6.87092 130.775 6.97319 130.989 7.17774C131.206 7.37908 131.315 7.62038 131.315 7.90163C131.315 8.18608 131.206 8.43058 130.989 8.63512C130.775 8.83647 130.516 8.93715 130.212 8.93715ZM136.781 9.88636L136.781 11.4205L132.347 11.4205L132.347 9.88636L136.781 9.88636ZM133.354 8.12216L135.396 8.12216L135.396 14.9872C135.396 15.1758 135.425 15.3228 135.482 15.4283C135.54 15.5305 135.62 15.6025 135.722 15.644C135.827 15.6855 135.949 15.7063 136.086 15.7063C136.182 15.7063 136.278 15.6983 136.374 15.6824C136.47 15.6632 136.543 15.6488 136.594 15.6392L136.916 17.1589C136.813 17.1909 136.67 17.2276 136.484 17.2692C136.299 17.3139 136.073 17.3411 135.808 17.3507C135.316 17.3699 134.885 17.3043 134.514 17.1541C134.146 17.0039 133.86 16.7706 133.656 16.4542C133.451 16.1378 133.35 15.7383 133.354 15.2557L133.354 8.12216ZM141.388 17.3938C140.631 17.3938 139.979 17.2404 139.432 16.9336C138.889 16.6236 138.47 16.1857 138.176 15.62C137.882 15.0511 137.735 14.3784 137.735 13.6017C137.735 12.8443 137.882 12.1795 138.176 11.6074C138.47 11.0353 138.884 10.5895 139.418 10.2699C139.955 9.95029 140.585 9.79048 141.307 9.79048C141.793 9.79048 142.245 9.86879 142.664 10.0254C143.086 10.1788 143.453 10.4105 143.766 10.7205C144.083 11.0305 144.329 11.4205 144.505 11.8903C144.68 12.3569 144.768 12.9034 144.768 13.5298L144.768 14.0907L138.55 14.0907L138.55 12.8251L142.846 12.8251C142.846 12.5311 142.782 12.2706 142.654 12.0437C142.526 11.8168 142.349 11.6394 142.122 11.5115C141.898 11.3805 141.638 11.315 141.341 11.315C141.03 11.315 140.756 11.3869 140.516 11.5307C140.279 11.6713 140.094 11.8615 139.96 12.1012C139.826 12.3377 139.757 12.6014 139.754 12.8922L139.754 14.0955C139.754 14.4599 139.821 14.7747 139.955 15.04C140.092 15.3052 140.286 15.5098 140.535 15.6536C140.784 15.7974 141.08 15.8693 141.422 15.8693C141.649 15.8693 141.857 15.8374 142.045 15.7734C142.234 15.7095 142.395 15.6136 142.529 15.4858C142.664 15.358 142.766 15.2014 142.836 15.016L144.725 15.1406C144.629 15.5945 144.433 15.9908 144.135 16.3295C143.841 16.6651 143.461 16.9272 142.994 17.1158C142.531 17.3011 141.996 17.3938 141.388 17.3938ZM146.101 17.25L146.101 9.88636L148.081 9.88636L148.081 11.1712L148.158 11.1712C148.292 10.7141 148.517 10.369 148.834 10.1357C149.15 9.89915 149.514 9.7809 149.927 9.7809C150.029 9.7809 150.139 9.78729 150.257 9.80007C150.376 9.81286 150.48 9.83043 150.569 9.85281L150.569 11.665C150.473 11.6362 150.341 11.6106 150.171 11.5882C150.002 11.5659 149.847 11.5547 149.706 11.5547C149.406 11.5547 149.137 11.6202 148.901 11.7512C148.667 11.8791 148.482 12.0581 148.345 12.2882C148.21 12.5183 148.143 12.7836 148.143 13.084L148.143 17.25L146.101 17.25Z",
    fill: "white"
  })), _defs || (_defs = /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear_13278_146910",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#C7F284"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00BEF0"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint1_linear_13278_146910",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#C7F284"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00BEF0"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint2_linear_13278_146910",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#C7F284"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00BEF0"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint3_linear_13278_146910",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#C7F284"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00BEF0"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint4_linear_13278_146910",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#C7F284"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00BEF0"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint5_linear_13278_146910",
    x1: 89.6424,
    y1: 4.875,
    x2: 78.5174,
    y2: 24,
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: 0.0001,
    stopColor: "#C7F284"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: 1,
    stopColor: "#00BEF0"
  })))));
};
var ForwardRef = /*#__PURE__*/forwardRef(SvgPoweredbyjupiterDark);

var SwapReset = function SwapReset(_ref) {
  var onRefresh = _ref.onRefresh,
    isCalculating = _ref.isCalculating,
    isSwapping = _ref.isSwapping;
  return /*#__PURE__*/React__default.createElement("button", {
    onClick: onRefresh,
    className: "inline-flex items-center justify-center rounded-full h-8 w-8 transition-colors text-muted dark:text-foreground-dark hover:bg-card dark:hover:bg-card-dark  hover:text-foreground dark:hover:text-foreground-dark",
    disabled: isCalculating || isSwapping,
    "aria-label": "Refresh quote"
  }, /*#__PURE__*/React__default.createElement(RotateCcw, {
    className: "h-4 w-4 ".concat(isCalculating ? "animate-spin [direction:reverse]" : "")
  }));
};

var SwapSettings = function SwapSettings(_ref) {
  var slippage = _ref.slippage,
    onSlippageChange = _ref.onSlippageChange,
    onSave = _ref.onSave;
  var _React$useState = React__default.useState(false),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    open = _React$useState2[0],
    setOpen = _React$useState2[1];
  var _React$useState3 = React__default.useState(slippage.toString()),
    _React$useState4 = _slicedToArray(_React$useState3, 2),
    tempSlippage = _React$useState4[0],
    setTempSlippage = _React$useState4[1];
  var _React$useState5 = React__default.useState(null),
    _React$useState6 = _slicedToArray(_React$useState5, 2),
    error = _React$useState6[0],
    setError = _React$useState6[1];
  var handleSave = function handleSave() {
    var slippageValue = parseFloat(tempSlippage);
    if (slippageValue < 0.01 || slippageValue > 100) {
      setError("Slippage must be between 0.01% and 100%");
      return;
    }
    onSlippageChange(slippageValue);
    onSave();
    setOpen(false);
  };
  var handleSlippageChange = function handleSlippageChange(value) {
    setError(null);
    setTempSlippage(value);
  };
  return /*#__PURE__*/React__default.createElement(Dialog.Root, {
    open: open,
    onOpenChange: setOpen
  }, /*#__PURE__*/React__default.createElement(Dialog.Trigger, {
    asChild: true
  }, /*#__PURE__*/React__default.createElement("button", {
    className: "inline-flex items-center justify-center rounded-full h-8 w-8 transition-colors text-muted dark:text-foreground-dark hover:bg-card dark:hover:bg-card-dark  hover:text-foreground dark:hover:text-foreground-dark",
    "aria-label": "Open swap settings"
  }, /*#__PURE__*/React__default.createElement(Settings, {
    className: "h-4 w-4"
  }))), /*#__PURE__*/React__default.createElement(Dialog.Portal, null, /*#__PURE__*/React__default.createElement(Dialog.Overlay, {
    className: " fixed inset-0 bg-black/50 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 z-40"
  }), /*#__PURE__*/React__default.createElement(Dialog.Content, {
    className: "font-sans fixed left-[50%] top-[50%] -translate-x-1/2 -translate-y-1/2 max-h-[85vh] w-[95vw] max-w-[400px] rounded-[20px] bg-card dark:bg-card-dark p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 z-50 overflow-hidden"
  }, /*#__PURE__*/React__default.createElement(Dialog.Title, {
    className: "text-xl font-semibold mb-6 text-foreground dark:text-foreground-dark"
  }, "Settings"), /*#__PURE__*/React__default.createElement("div", {
    className: "grid gap-6"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "space-y-4"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "flex items-center justify-between"
  }, /*#__PURE__*/React__default.createElement("label", {
    htmlFor: "slippage",
    className: "text-sm font-medium text-foreground dark:text-foreground-dark"
  }, "Max Slippage"), /*#__PURE__*/React__default.createElement("button", {
    className: "inline-flex items-center justify-center rounded-full px-3 text-sm h-8 bg-muted/20 hover:bg-muted/30 dark:bg-muted-dark/20 dark:hover:bg-muted-dark/30 text-foreground dark:text-foreground-dark transition-colors",
    onClick: function onClick() {
      setTempSlippage("3");
      setError(null);
    }
  }, "Reset")), /*#__PURE__*/React__default.createElement("div", {
    className: "flex items-center space-x-2"
  }, /*#__PURE__*/React__default.createElement("input", {
    id: "slippage",
    type: "number",
    value: tempSlippage,
    onChange: function onChange(e) {
      return handleSlippageChange(e.target.value);
    },
    className: "[appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none flex h-12 w-full rounded-2xl border border-input dark:border-input-dark bg-background dark:bg-background-dark px-4 py-2 text-base text-foreground dark:text-foreground-dark ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring dark:focus-visible:ring-ring-dark focus-visible:ring-offset-2",
    step: "0.1",
    min: "0.01",
    max: "100"
  }), /*#__PURE__*/React__default.createElement("span", {
    className: "text-base text-foreground dark:text-foreground-dark"
  }, "%")), error && (/*#__PURE__*/React__default.createElement("p", {
    className: "text-sm text-destructive dark:text-destructive-dark"
  }, error)), /*#__PURE__*/React__default.createElement("p", {
    className: "text-sm text-muted-foreground dark:text-muted-dark-foreground"
  }, "Your transaction will revert if the price changes unfavorably by more than this percentage.")), /*#__PURE__*/React__default.createElement("div", {
    className: "flex justify-end"
  }, /*#__PURE__*/React__default.createElement("button", {
    onClick: handleSave,
    className: "inline-flex items-center justify-center rounded-2xl text-sm font-medium h-8 px-6 bg-primary dark:bg-primary-dark text-primary-foreground dark:text-primary-dark-foreground hover:opacity-90 transition-opacity"
  }, "Confirm"))), /*#__PURE__*/React__default.createElement(Dialog.Close, {
    asChild: true
  }, /*#__PURE__*/React__default.createElement("button", {
    className: "absolute right-4 top-4 rounded-full p-2 opacity-70 hover:opacity-100 hover:bg-muted/20 dark:hover:bg-muted-dark/20 transition-all",
    "aria-label": "Close"
  }, /*#__PURE__*/React__default.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: "24",
    height: "24",
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: "2",
    strokeLinecap: "round",
    strokeLinejoin: "round",
    className: "h-5 w-5"
  }, /*#__PURE__*/React__default.createElement("line", {
    x1: "18",
    y1: "6",
    x2: "6",
    y2: "18"
  }), /*#__PURE__*/React__default.createElement("line", {
    x1: "6",
    y1: "6",
    x2: "18",
    y2: "18"
  })))))));
};

var SwapTokenImage = function SwapTokenImage(_ref) {
  var token = _ref.token;
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    imgError = _useState2[0],
    setImgError = _useState2[1];
  useEffect(function () {
    setImgError(false);
  }, [token.logoURI]);
  if (!token.logoURI || imgError) {
    return /*#__PURE__*/React__default.createElement(Coins, {
      className: "h-4 w-4 text-muted-foreground"
    });
  }
  return /*#__PURE__*/React__default.createElement("img", {
    src: token.logoURI,
    alt: token.symbol,
    className: "w-full h-full object-cover",
    onError: function onError() {
      return setImgError(true);
    },
    loading: "lazy",
    crossOrigin: "anonymous"
  });
};

var SwapTokenButton = function SwapTokenButton(_ref) {
  var token = _ref.token,
    onClick = _ref.onClick;
  return /*#__PURE__*/React__default.createElement("button", {
    className: "inline-flex items-center justify-center gap-2 h-10 px-4 rounded-2xl text-sm font-medium transition-all hover:bg-muted/20 dark:hover:bg-muted-dark/20 bg-background dark:bg-background-dark w-fit text-foreground dark:text-foreground-dark shadow-sm hover:shadow-md",
    onClick: onClick
  }, token ? (/*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement("div", {
    className: "w-5 h-5 relative rounded-full overflow-hidden bg-muted/10 dark:bg-muted-dark/10 flex items-center justify-center"
  }, /*#__PURE__*/React__default.createElement(SwapTokenImage, {
    token: token
  })), /*#__PURE__*/React__default.createElement("span", {
    className: "whitespace-nowrap font-semibold"
  }, token.symbol))) : (/*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(Coins, {
    className: "h-5 w-5 text-foreground dark:text-foreground-dark"
  }), /*#__PURE__*/React__default.createElement("span", {
    className: "whitespace-nowrap"
  }, "Select token"))), /*#__PURE__*/React__default.createElement(ChevronDown, {
    className: "h-5 w-5 text-foreground dark:text-foreground-dark ml-0.5"
  }));
};

var useTokens = function useTokens() {
  var _useState = useState([]),
    _useState2 = _slicedToArray(_useState, 2),
    tokens = _useState2[0],
    setTokens = _useState2[1];
  var _useState3 = useState(true),
    _useState4 = _slicedToArray(_useState3, 2),
    loading = _useState4[0],
    setLoading = _useState4[1];
  var _useState5 = useState(null),
    _useState6 = _slicedToArray(_useState5, 2),
    error = _useState6[0],
    setError = _useState6[1];
  useEffect(function () {
    var fetchTokens = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
        var response, data;
        return _regeneratorRuntime().wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;
              setLoading(true);
              _context.next = 4;
              return fetch("	https://api.jup.ag/tokens/v1/tagged/verified");
            case 4:
              response = _context.sent;
              if (response.ok) {
                _context.next = 7;
                break;
              }
              throw new Error("Failed to fetch tokens");
            case 7:
              _context.next = 9;
              return response.json();
            case 9:
              data = _context.sent;
              setTokens(data);
              _context.next = 16;
              break;
            case 13:
              _context.prev = 13;
              _context.t0 = _context["catch"](0);
              setError(_context.t0 instanceof Error ? _context.t0.message : "Failed to fetch tokens");
            case 16:
              _context.prev = 16;
              setLoading(false);
              return _context.finish(16);
            case 19:
            case "end":
              return _context.stop();
          }
        }, _callee, null, [[0, 13, 16, 19]]);
      }));
      return function fetchTokens() {
        return _ref.apply(this, arguments);
      };
    }();
    fetchTokens();
  }, []);
  return {
    tokens: tokens,
    loading: loading,
    error: error
  };
};

var SwapTokenDialog = function SwapTokenDialog(_ref) {
  var open = _ref.open,
    onClose = _ref.onClose,
    onSelect = _ref.onSelect;
  var _useTokens = useTokens(),
    tokens = _useTokens.tokens,
    loading = _useTokens.loading,
    error = _useTokens.error;
  var _useState = useState(""),
    _useState2 = _slicedToArray(_useState, 2),
    searchQuery = _useState2[0],
    setSearchQuery = _useState2[1];
  var filteredTokens = tokens.filter(function (token) {
    var query = searchQuery.toLowerCase();
    return token.symbol.toLowerCase().includes(query) || token.name.toLowerCase().includes(query) || token.address.toLowerCase().includes(query);
  });
  return /*#__PURE__*/React__default.createElement(Dialog.Root, {
    open: open,
    onOpenChange: function onOpenChange(isOpen) {
      return !isOpen && onClose();
    }
  }, /*#__PURE__*/React__default.createElement(Dialog.Portal, null, /*#__PURE__*/React__default.createElement(Dialog.Overlay, {
    className: "fixed inset-0 bg-black/50 backdrop-blur-sm data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 z-40"
  }), /*#__PURE__*/React__default.createElement(Dialog.Content, {
    className: "font-sans fixed left-[50%] top-[50%] -translate-x-1/2 -translate-y-1/2 max-h-[85vh] w-[95vw] max-w-[600px] rounded-[24px] bg-card/95 dark:bg-card-dark/95 backdrop-blur-md shadow-2xl border border-border/10 dark:border-border-dark/10 duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 z-50 flex flex-col overflow-hidden"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "flex items-center justify-between p-6 border-b border-border/10 dark:border-border-dark/10"
  }, /*#__PURE__*/React__default.createElement(Dialog.Title, {
    className: "text-xl font-bold bg-gradient-to-r from-nebula to-cosmic bg-clip-text text-transparent dark:from-helix dark:to-trifid"
  }, "Select a token"), /*#__PURE__*/React__default.createElement(Dialog.Close, {
    asChild: true
  }, /*#__PURE__*/React__default.createElement("button", {
    className: "rounded-full p-2 opacity-70 hover:opacity-100 hover:bg-muted/10 dark:hover:bg-muted-dark/10 transition-all",
    "aria-label": "Close"
  }, /*#__PURE__*/React__default.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: "24",
    height: "24",
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeWidth: "2",
    strokeLinecap: "round",
    strokeLinejoin: "round",
    className: "h-5 w-5"
  }, /*#__PURE__*/React__default.createElement("line", {
    x1: "18",
    y1: "6",
    x2: "6",
    y2: "18"
  }), /*#__PURE__*/React__default.createElement("line", {
    x1: "6",
    y1: "6",
    x2: "18",
    y2: "18"
  }))))), /*#__PURE__*/React__default.createElement("div", {
    className: "px-6 py-4"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "relative"
  }, /*#__PURE__*/React__default.createElement("input", {
    placeholder: "Search by name or paste address",
    value: searchQuery,
    onChange: function onChange(e) {
      return setSearchQuery(e.target.value);
    },
    className: "flex h-10 w-full rounded-lg border border-input/10 dark:border-input-dark/10 bg-background/50 dark:bg-background-dark/50 px-12 py-3 text-sm text-foreground dark:text-foreground-dark ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground/70 dark:placeholder:text-muted-dark-foreground/70 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-nebula/50 dark:focus-visible:ring-helix/50 focus-visible:ring-offset-1 transition-all"
  }), /*#__PURE__*/React__default.createElement(Search, {
    className: "absolute left-4 top-1/2 -translate-y-1/2 h-4 w-4 text-muted-foreground/70 dark:text-muted-dark-foreground/70"
  }))), /*#__PURE__*/React__default.createElement("div", {
    className: "flex-1 overflow-y-auto min-h-0 px-4"
  }, loading ? (/*#__PURE__*/React__default.createElement("div", {
    className: "flex items-center justify-center h-full text-muted-foreground dark:text-muted-dark-foreground"
  }, "Loading tokens...")) : error ? (/*#__PURE__*/React__default.createElement("div", {
    className: "flex items-center justify-center h-full text-destructive dark:text-destructive-dark"
  }, error)) : (/*#__PURE__*/React__default.createElement("div", {
    className: "space-y-1 py-2"
  }, filteredTokens.map(function (token) {
    return /*#__PURE__*/React__default.createElement("button", {
      key: token.address,
      className: "w-full flex items-center gap-4 p-3 rounded-2xl hover:bg-muted/5 dark:hover:bg-muted/5 active:bg-muted/10 dark:active:bg-muted-dark/10 transition-all duration-200 text-foreground dark:text-foreground-dark group",
      onClick: function onClick() {
        onSelect(token);
        onClose();
      }
    }, /*#__PURE__*/React__default.createElement("div", {
      className: "w-6 h-6 relative rounded-full overflow-hidden bg-gradient-to-br from-nebula/10 to-cosmic/10 dark:from-helix/10 dark:to-trifid/10 flex items-center justify-center group-hover:shadow-lg transition-shadow duration-200"
    }, /*#__PURE__*/React__default.createElement(SwapTokenImage, {
      token: token
    })), /*#__PURE__*/React__default.createElement("div", {
      className: "flex-1 text-left"
    }, /*#__PURE__*/React__default.createElement("div", {
      className: "font-semibold text-sm"
    }, token.symbol), /*#__PURE__*/React__default.createElement("div", {
      className: "text-xs text-muted-foreground/80 dark:text-muted-dark-foreground/80"
    }, token.name)));
  })))))));
};

function createJSONStorage(getStorage, options) {
  let storage;
  try {
    storage = getStorage();
  } catch (e) {
    return;
  }
  const persistStorage = {
    getItem: (name) => {
      var _a;
      const parse = (str2) => {
        if (str2 === null) {
          return null;
        }
        return JSON.parse(str2, void 0 );
      };
      const str = (_a = storage.getItem(name)) != null ? _a : null;
      if (str instanceof Promise) {
        return str.then(parse);
      }
      return parse(str);
    },
    setItem: (name, newValue) => storage.setItem(
      name,
      JSON.stringify(newValue, void 0 )
    ),
    removeItem: (name) => storage.removeItem(name)
  };
  return persistStorage;
}
const toThenable = (fn) => (input) => {
  try {
    const result = fn(input);
    if (result instanceof Promise) {
      return result;
    }
    return {
      then(onFulfilled) {
        return toThenable(onFulfilled)(result);
      },
      catch(_onRejected) {
        return this;
      }
    };
  } catch (e) {
    return {
      then(_onFulfilled) {
        return this;
      },
      catch(onRejected) {
        return toThenable(onRejected)(e);
      }
    };
  }
};
const persistImpl = (config, baseOptions) => (set, get, api) => {
  let options = {
    storage: createJSONStorage(() => localStorage),
    partialize: (state) => state,
    version: 0,
    merge: (persistedState, currentState) => ({
      ...currentState,
      ...persistedState
    }),
    ...baseOptions
  };
  let hasHydrated = false;
  const hydrationListeners = /* @__PURE__ */ new Set();
  const finishHydrationListeners = /* @__PURE__ */ new Set();
  let storage = options.storage;
  if (!storage) {
    return config(
      (...args) => {
        console.warn(
          `[zustand persist middleware] Unable to update item '${options.name}', the given storage is currently unavailable.`
        );
        set(...args);
      },
      get,
      api
    );
  }
  const setItem = () => {
    const state = options.partialize({ ...get() });
    return storage.setItem(options.name, {
      state,
      version: options.version
    });
  };
  const savedSetState = api.setState;
  api.setState = (state, replace) => {
    savedSetState(state, replace);
    void setItem();
  };
  const configResult = config(
    (...args) => {
      set(...args);
      void setItem();
    },
    get,
    api
  );
  api.getInitialState = () => configResult;
  let stateFromStorage;
  const hydrate = () => {
    var _a, _b;
    if (!storage) return;
    hasHydrated = false;
    hydrationListeners.forEach((cb) => {
      var _a2;
      return cb((_a2 = get()) != null ? _a2 : configResult);
    });
    const postRehydrationCallback = ((_b = options.onRehydrateStorage) == null ? void 0 : _b.call(options, (_a = get()) != null ? _a : configResult)) || void 0;
    return toThenable(storage.getItem.bind(storage))(options.name).then((deserializedStorageValue) => {
      if (deserializedStorageValue) {
        if (typeof deserializedStorageValue.version === "number" && deserializedStorageValue.version !== options.version) {
          if (options.migrate) {
            const migration = options.migrate(
              deserializedStorageValue.state,
              deserializedStorageValue.version
            );
            if (migration instanceof Promise) {
              return migration.then((result) => [true, result]);
            }
            return [true, migration];
          }
          console.error(
            `State loaded from storage couldn't be migrated since no migrate function was provided`
          );
        } else {
          return [false, deserializedStorageValue.state];
        }
      }
      return [false, void 0];
    }).then((migrationResult) => {
      var _a2;
      const [migrated, migratedState] = migrationResult;
      stateFromStorage = options.merge(
        migratedState,
        (_a2 = get()) != null ? _a2 : configResult
      );
      set(stateFromStorage, true);
      if (migrated) {
        return setItem();
      }
    }).then(() => {
      postRehydrationCallback == null ? void 0 : postRehydrationCallback(stateFromStorage, void 0);
      stateFromStorage = get();
      hasHydrated = true;
      finishHydrationListeners.forEach((cb) => cb(stateFromStorage));
    }).catch((e) => {
      postRehydrationCallback == null ? void 0 : postRehydrationCallback(void 0, e);
    });
  };
  api.persist = {
    setOptions: (newOptions) => {
      options = {
        ...options,
        ...newOptions
      };
      if (newOptions.storage) {
        storage = newOptions.storage;
      }
    },
    clearStorage: () => {
      storage == null ? void 0 : storage.removeItem(options.name);
    },
    getOptions: () => options,
    rehydrate: () => hydrate(),
    hasHydrated: () => hasHydrated,
    onHydrate: (cb) => {
      hydrationListeners.add(cb);
      return () => {
        hydrationListeners.delete(cb);
      };
    },
    onFinishHydration: (cb) => {
      finishHydrationListeners.add(cb);
      return () => {
        finishHydrationListeners.delete(cb);
      };
    }
  };
  if (!options.skipHydration) {
    hydrate();
  }
  return stateFromStorage || configResult;
};
const persist = persistImpl;

// Define the initial state of the swap store
var initialState = {
  tokenFrom: null,
  tokenTo: null,
  amountFrom: "",
  amountTo: "",
  slippage: 3,
  quoteResponse: null,
  isCalculating: false,
  isSwapping: false,
  isFromDialogOpen: false,
  isToDialogOpen: false,
  error: null
};
var useSwapStore = create()(persist(function (set, get) {
  return _objectSpread2(_objectSpread2({}, initialState), {}, {
    setTokenFrom: function setTokenFrom(token) {
      return set({
        tokenFrom: token
      });
    },
    setTokenTo: function setTokenTo(token) {
      return set({
        tokenTo: token
      });
    },
    setAmountFrom: function setAmountFrom(amount) {
      // Always set the input amount
      set({
        amountFrom: amount
      });
      // Reset 'to' amount if 'from' amount is invalid
      if (!amount || isNaN(parseFloat(amount))) {
        set({
          amountTo: "0",
          quoteResponse: null
        });
      }
    },
    setAmountTo: function setAmountTo(amount) {
      return set({
        amountTo: amount
      });
    },
    setSlippage: function setSlippage(slippage) {
      return set({
        slippage: slippage
      });
    },
    setQuoteResponse: function setQuoteResponse(quote) {
      return set({
        quoteResponse: quote
      });
    },
    setIsCalculating: function setIsCalculating(isCalculating) {
      return set({
        isCalculating: isCalculating
      });
    },
    setIsSwapping: function setIsSwapping(isSwapping) {
      return set({
        isSwapping: isSwapping
      });
    },
    setIsFromDialogOpen: function setIsFromDialogOpen(isOpen) {
      return set({
        isFromDialogOpen: isOpen
      });
    },
    setIsToDialogOpen: function setIsToDialogOpen(isOpen) {
      return set({
        isToDialogOpen: isOpen
      });
    },
    setError: function setError(error) {
      set({
        error: error
      });
      // Reset 'to' amount when there's an error to trigger recalculation
      if (error) {
        set({
          amountTo: "",
          quoteResponse: null
        });
      }
    },
    reset: function reset() {
      var _get = get(),
        tokenFrom = _get.tokenFrom,
        tokenTo = _get.tokenTo;
      set(_objectSpread2(_objectSpread2({}, initialState), {}, {
        // Preserve selected tokens
        tokenFrom: tokenFrom,
        tokenTo: tokenTo
      }));
    },
    resetAmounts: function resetAmounts() {
      return set({
        amountFrom: "",
        amountTo: "",
        quoteResponse: null,
        error: null
      });
    },
    swapTokens: function swapTokens() {
      var _get2 = get(),
        tokenFrom = _get2.tokenFrom,
        tokenTo = _get2.tokenTo,
        amountFrom = _get2.amountFrom,
        amountTo = _get2.amountTo;
      set({
        tokenFrom: tokenTo,
        tokenTo: tokenFrom,
        amountFrom: amountTo,
        amountTo: amountFrom
      });
    }
  });
}, {
  name: "swap-storage",
  // Only persist tokens, amounts, and slippage
  partialize: function partialize(state) {
    return {
      tokenFrom: state.tokenFrom,
      tokenTo: state.tokenTo,
      amountFrom: state.amountFrom,
      amountTo: state.amountTo,
      slippage: state.slippage
    };
  }
}));
// Selector hook for better performance
var useSwapSelector = function useSwapSelector(selector) {
  return useSwapStore(selector);
};

// Minimum SOL to keep for gas fees (0.01 SOL)
var MIN_SOL_FOR_GAS = 0.01 * LAMPORTS_PER_SOL;
var WSOL_MINT = "So11111111111111111111111111111111111111112";
var useSwap = function useSwap(_ref) {
  var rpcUrl = _ref.rpcUrl,
    referralKey = _ref.referralKey,
    platformFeeBps = _ref.platformFeeBps,
    apiKey = _ref.apiKey;
  var _useWallet = useWallet(),
    publicKey = _useWallet.publicKey,
    signTransaction = _useWallet.signTransaction;
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    loading = _useState2[0],
    setLoading = _useState2[1];
  var _useState3 = useState(null),
    _useState4 = _slicedToArray(_useState3, 2),
    error = _useState4[0],
    setError = _useState4[1];
  // Initialize connection with provided RPC URL
  var connection = useMemo(function () {
    return new Connection(rpcUrl);
  }, [rpcUrl]);
  // Get quote for swap
  var getQuote = useCallback(/*#__PURE__*/function () {
    var _ref2 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee(inputToken, outputToken, amount) {
      var slippageBps,
        url,
        response,
        data,
        errorMessage,
        _args = arguments;
      return _regeneratorRuntime().wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            slippageBps = _args.length > 3 && _args[3] !== undefined ? _args[3] : 300;
            _context.prev = 1;
            setLoading(true);
            setError(null);
            url = new URL("https://api.jup.ag/swap/v1/quote");
            url.searchParams.append("inputMint", inputToken.address);
            url.searchParams.append("outputMint", outputToken.address);
            url.searchParams.append("amount", (amount * Math.pow(10, inputToken.decimals)).toString());
            url.searchParams.append("slippageBps", slippageBps.toString());
            url.searchParams.append("restrictIntermediateTokens", "true");
            // Add platform fee if provided
            if (platformFeeBps) {
              url.searchParams.append("platformFeeBps", platformFeeBps.toString());
            }
            _context.next = 13;
            return fetch(url.toString());
          case 13:
            response = _context.sent;
            _context.next = 16;
            return response.json();
          case 16:
            data = _context.sent;
            if (response.ok) {
              _context.next = 19;
              break;
            }
            throw new Error(data.error || "Failed to get quote");
          case 19:
            return _context.abrupt("return", data);
          case 22:
            _context.prev = 22;
            _context.t0 = _context["catch"](1);
            errorMessage = _context.t0 instanceof Error ? _context.t0.message : "Failed to get quote";
            setError(errorMessage);
            throw new Error(errorMessage);
          case 27:
            _context.prev = 27;
            setLoading(false);
            return _context.finish(27);
          case 30:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[1, 22, 27, 30]]);
    }));
    return function (_x, _x2, _x3) {
      return _ref2.apply(this, arguments);
    };
  }(), [platformFeeBps]);
  // Execute swap
  var executeSwap = useCallback(/*#__PURE__*/function () {
    var _ref3 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee2(quoteResponse) {
      var feeAccount, _yield$PublicKey$find, _yield$PublicKey$find2, feeAccountAddress, headers, swapResponse, _yield$swapResponse$j, swapTransaction, swapTransactionBuf, transaction, signedTransaction, rawTransaction, txid, latestBlockhash, errorMessage;
      return _regeneratorRuntime().wrap(function _callee2$(_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            if (!(!publicKey || !signTransaction)) {
              _context2.next = 3;
              break;
            }
            setError("Wallet not connected");
            return _context2.abrupt("return", null);
          case 3:
            _context2.prev = 3;
            setLoading(true);
            setError(null);
            if (!referralKey) {
              _context2.next = 13;
              break;
            }
            _context2.next = 9;
            return PublicKey.findProgramAddressSync([Buffer.from("referral_ata"), new PublicKey(referralKey).toBuffer(), new PublicKey(quoteResponse.inputMint).toBuffer()], new PublicKey("REFER4ZgmyYx9c6He5XfaTMiGfdLwRnkV4RPp9t9iF3"));
          case 9:
            _yield$PublicKey$find = _context2.sent;
            _yield$PublicKey$find2 = _slicedToArray(_yield$PublicKey$find, 1);
            feeAccountAddress = _yield$PublicKey$find2[0];
            feeAccount = feeAccountAddress.toString();
          case 13:
            headers = {
              "Content-Type": "application/json"
            };
            if (apiKey) {
              headers["x-api-key"] = apiKey;
            }
            _context2.next = 17;
            return fetch("https://api.jup.ag/swap/v1/swap", {
              method: "POST",
              headers: headers,
              body: JSON.stringify({
                quoteResponse: quoteResponse,
                userPublicKey: publicKey.toString(),
                wrapAndUnwrapSol: true,
                dynamicComputeUnitLimit: true,
                feeAccount: feeAccount,
                dynamicSlippage: {
                  maxBps: quoteResponse.slippageBps
                },
                prioritizationFeeLamports: {
                  priorityLevelWithMaxLamports: {
                    maxLamports: 10000000,
                    priorityLevel: "veryHigh"
                  }
                }
              })
            });
          case 17:
            swapResponse = _context2.sent;
            if (swapResponse.ok) {
              _context2.next = 20;
              break;
            }
            throw new Error("Failed to prepare swap transaction");
          case 20:
            _context2.next = 22;
            return swapResponse.json();
          case 22:
            _yield$swapResponse$j = _context2.sent;
            swapTransaction = _yield$swapResponse$j.swapTransaction;
            // Deserialize the transaction
            swapTransactionBuf = Buffer.from(swapTransaction, "base64");
            transaction = VersionedTransaction.deserialize(new Uint8Array(swapTransactionBuf.toJSON().data)); // Sign the transaction
            _context2.next = 28;
            return signTransaction(transaction);
          case 28:
            signedTransaction = _context2.sent;
            // Execute the transaction
            rawTransaction = signedTransaction.serialize();
            _context2.next = 32;
            return connection.sendRawTransaction(rawTransaction, {
              skipPreflight: true,
              maxRetries: 2
            });
          case 32:
            txid = _context2.sent;
            _context2.next = 35;
            return connection.getLatestBlockhash();
          case 35:
            latestBlockhash = _context2.sent;
            _context2.next = 38;
            return connection.confirmTransaction({
              blockhash: latestBlockhash.blockhash,
              lastValidBlockHeight: latestBlockhash.lastValidBlockHeight,
              signature: txid
            });
          case 38:
            return _context2.abrupt("return", txid);
          case 41:
            _context2.prev = 41;
            _context2.t0 = _context2["catch"](3);
            errorMessage = _context2.t0 instanceof Error ? _context2.t0.message : "Failed to execute swap";
            setError(errorMessage);
            throw new Error(errorMessage);
          case 46:
            _context2.prev = 46;
            setLoading(false);
            return _context2.finish(46);
          case 49:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[3, 41, 46, 49]]);
    }));
    return function (_x4) {
      return _ref3.apply(this, arguments);
    };
  }(), [connection, publicKey, signTransaction, referralKey]);
  // Calculate output amount based on input
  var calculateOutputAmount = useCallback(/*#__PURE__*/function () {
    var _ref4 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee3(inputToken, outputToken, inputAmount) {
      var quote, outputAmount, errorMessage;
      return _regeneratorRuntime().wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            if (!(!inputAmount || inputAmount <= 0)) {
              _context3.next = 2;
              break;
            }
            return _context3.abrupt("return", "0");
          case 2:
            _context3.prev = 2;
            _context3.next = 5;
            return getQuote(inputToken, outputToken, inputAmount);
          case 5:
            quote = _context3.sent;
            if (quote) {
              _context3.next = 8;
              break;
            }
            return _context3.abrupt("return", "0");
          case 8:
            // Convert amount from base units to decimal
            outputAmount = (parseInt(quote.outAmount) / Math.pow(10, outputToken.decimals)).toString();
            return _context3.abrupt("return", outputAmount);
          case 12:
            _context3.prev = 12;
            _context3.t0 = _context3["catch"](2);
            errorMessage = _context3.t0 instanceof Error ? _context3.t0.message : "Failed to calculate output amount";
            console.error("Failed to calculate output amount:", _context3.t0);
            setError(errorMessage);
            throw new Error(errorMessage);
          case 18:
          case "end":
            return _context3.stop();
        }
      }, _callee3, null, [[2, 12]]);
    }));
    return function (_x5, _x6, _x7) {
      return _ref4.apply(this, arguments);
    };
  }(), [getQuote]);
  // Calculate max input amount accounting for gas fees if token is SOL
  var calculateMaxInput = useCallback(/*#__PURE__*/function () {
    var _ref5 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee4(inputToken, balance) {
      var maxAmount;
      return _regeneratorRuntime().wrap(function _callee4$(_context4) {
        while (1) switch (_context4.prev = _context4.next) {
          case 0:
            if (!(!balance || balance <= 0)) {
              _context4.next = 2;
              break;
            }
            return _context4.abrupt("return", 0);
          case 2:
            if (!(inputToken.address === WSOL_MINT)) {
              _context4.next = 5;
              break;
            }
            maxAmount = Math.max(0, (balance * LAMPORTS_PER_SOL - MIN_SOL_FOR_GAS) / LAMPORTS_PER_SOL);
            return _context4.abrupt("return", parseFloat(maxAmount.toFixed(inputToken.decimals)));
          case 5:
            return _context4.abrupt("return", balance);
          case 6:
          case "end":
            return _context4.stop();
        }
      }, _callee4);
    }));
    return function (_x8, _x9) {
      return _ref5.apply(this, arguments);
    };
  }(), []);
  return {
    getQuote: getQuote,
    executeSwap: executeSwap,
    calculateOutputAmount: calculateOutputAmount,
    calculateMaxInput: calculateMaxInput,
    loading: loading,
    error: error
  };
};

var useTokenBalance = function useTokenBalance(token) {
  var _useConnection = useConnection(),
    connection = _useConnection.connection;
  var _useWallet = useWallet(),
    publicKey = _useWallet.publicKey;
  var _useState = useState(null),
    _useState2 = _slicedToArray(_useState, 2),
    balance = _useState2[0],
    setBalance = _useState2[1];
  var _useState3 = useState(false),
    _useState4 = _slicedToArray(_useState3, 2),
    loading = _useState4[0],
    setLoading = _useState4[1];
  var fetchBalance = useCallback(/*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
    var _balance, solBalance, tokenAccount, tokenAmount, _balance2;
    return _regeneratorRuntime().wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          if (!(!publicKey || !token)) {
            _context.next = 3;
            break;
          }
          setBalance(null);
          return _context.abrupt("return");
        case 3:
          _context.prev = 3;
          setLoading(true);
          if (!(token.symbol === "SOL")) {
            _context.next = 13;
            break;
          }
          _context.next = 8;
          return connection.getBalance(publicKey, "finalized");
        case 8:
          _balance = _context.sent;
          solBalance = _balance / Math.pow(10, 9);
          setBalance(solBalance);
          _context.next = 17;
          break;
        case 13:
          _context.next = 15;
          return connection.getParsedTokenAccountsByOwner(publicKey, {
            programId: TOKEN_PROGRAM_ID,
            mint: new PublicKey(token.address)
          }, "finalized");
        case 15:
          tokenAccount = _context.sent;
          if (tokenAccount.value.length > 0) {
            tokenAmount = tokenAccount.value[0].account.data.parsed.info.tokenAmount;
            _balance2 = tokenAmount.uiAmount;
            setBalance(_balance2);
          } else {
            setBalance(0);
          }
        case 17:
          _context.next = 23;
          break;
        case 19:
          _context.prev = 19;
          _context.t0 = _context["catch"](3);
          console.error("Error fetching token balance:", _context.t0);
          setBalance(null);
        case 23:
          _context.prev = 23;
          setLoading(false);
          return _context.finish(23);
        case 26:
        case "end":
          return _context.stop();
      }
    }, _callee, null, [[3, 19, 23, 26]]);
  })), [connection, publicKey, token]);
  // Initial fetch when dependencies change
  useEffect(function () {
    fetchBalance();
  }, [fetchBalance]);
  return {
    balance: balance,
    loading: loading,
    refetch: fetchBalance
  };
};

var useSwapOperations = function useSwapOperations(config) {
  // State selectors
  var tokenFrom = useSwapSelector(function (state) {
    return state.tokenFrom;
  });
  var tokenTo = useSwapSelector(function (state) {
    return state.tokenTo;
  });
  var amountFrom = useSwapSelector(function (state) {
    return state.amountFrom;
  });
  var amountTo = useSwapSelector(function (state) {
    return state.amountTo;
  });
  var slippage = useSwapSelector(function (state) {
    return state.slippage;
  });
  var isCalculating = useSwapSelector(function (state) {
    return state.isCalculating;
  });
  var isSwapping = useSwapSelector(function (state) {
    return state.isSwapping;
  });
  // Actions
  var _useSwapStore = useSwapStore(),
    setAmountFrom = _useSwapStore.setAmountFrom,
    setAmountTo = _useSwapStore.setAmountTo,
    setQuoteResponse = _useSwapStore.setQuoteResponse,
    setIsCalculating = _useSwapStore.setIsCalculating,
    setIsSwapping = _useSwapStore.setIsSwapping,
    setError = _useSwapStore.setError;
  var _useSwap = useSwap(config),
    getQuote = _useSwap.getQuote,
    executeSwap = _useSwap.executeSwap,
    calculateMaxInput = _useSwap.calculateMaxInput;
  var _useTokenBalance = useTokenBalance(tokenFrom),
    fromBalance = _useTokenBalance.balance,
    refetchFromBalance = _useTokenBalance.refetch;
  var _useTokenBalance2 = useTokenBalance(tokenTo),
    toBalance = _useTokenBalance2.balance,
    refetchToBalance = _useTokenBalance2.refetch;
  var refreshTimerRef = useRef(null);
  // Memoized quote calculation with debounce
  var calculateQuote = useCallback(/*#__PURE__*/function () {
    var _ref = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee(amount) {
      var parsedAmount, quote, outputAmount, errorMessage;
      return _regeneratorRuntime().wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            if (!(!tokenFrom || !tokenTo)) {
              _context.next = 5;
              break;
            }
            setAmountTo("0");
            setQuoteResponse(null);
            setIsCalculating(false);
            return _context.abrupt("return");
          case 5:
            if (amount) {
              _context.next = 10;
              break;
            }
            setAmountTo("0");
            setQuoteResponse(null);
            setIsCalculating(false);
            return _context.abrupt("return");
          case 10:
            if (!isNaN(parseFloat(amount))) {
              _context.next = 16;
              break;
            }
            setAmountTo("0");
            setQuoteResponse(null);
            setIsCalculating(false);
            toast.error(/*#__PURE__*/React__default.createElement("div", {
              className: "flex flex-col bg-red-100 w-full p-4 rounded-lg"
            }, /*#__PURE__*/React__default.createElement("span", {
              className: "font-semibold"
            }, "Invalid Input: Please enter a valid number"), /*#__PURE__*/React__default.createElement("span", null, "The amount must be a valid numerical value")), {
              style: {
                padding: 0,
                margin: 0
              }
            });
            return _context.abrupt("return");
          case 16:
            parsedAmount = parseFloat(amount); // Show error for zero or negative amount after debounce
            if (!(parsedAmount <= 0)) {
              _context.next = 23;
              break;
            }
            setAmountTo("0");
            setQuoteResponse(null);
            setIsCalculating(false);
            toast.error(/*#__PURE__*/React__default.createElement("div", {
              className: "flex flex-col bg-red-100 w-full p-4 rounded-lg"
            }, /*#__PURE__*/React__default.createElement("span", {
              className: "font-semibold"
            }, "Invalid Amount"), /*#__PURE__*/React__default.createElement("span", null, "Amount must be greater than 0")), {
              style: {
                padding: 0,
                margin: 0
              }
            });
            return _context.abrupt("return");
          case 23:
            if (!(fromBalance !== null && parsedAmount > fromBalance)) {
              _context.next = 28;
              break;
            }
            setAmountTo("0");
            setQuoteResponse(null);
            toast.error(/*#__PURE__*/React__default.createElement("div", {
              className: "flex flex-col bg-red-100 w-full p-4 rounded-lg"
            }, /*#__PURE__*/React__default.createElement("span", {
              className: "font-semibold"
            }, "Insufficient Balance"), /*#__PURE__*/React__default.createElement("span", null, "You don't have enough ", tokenFrom.symbol, " in your wallet")), {
              style: {
                padding: 0,
                margin: 0
              }
            });
            return _context.abrupt("return");
          case 28:
            setIsCalculating(true);
            _context.prev = 29;
            _context.next = 32;
            return getQuote(tokenFrom, tokenTo, parsedAmount, slippage * 100);
          case 32:
            quote = _context.sent;
            if (quote) {
              setQuoteResponse(quote);
              outputAmount = (parseInt(quote.outAmount) / Math.pow(10, tokenTo.decimals)).toString();
              setAmountTo(outputAmount);
            }
            _context.next = 43;
            break;
          case 36:
            _context.prev = 36;
            _context.t0 = _context["catch"](29);
            errorMessage = _context.t0 instanceof Error ? _context.t0.message : "Failed to calculate output amount";
            console.error("Failed to calculate output amount:", _context.t0);
            setAmountTo("0");
            setQuoteResponse(null);
            toast.error(/*#__PURE__*/React__default.createElement("div", {
              className: "flex flex-col bg-red-100 w-full p-4 rounded-lg"
            }, /*#__PURE__*/React__default.createElement("span", {
              className: "font-semibold"
            }, "Quote Calculation Failed"), /*#__PURE__*/React__default.createElement("span", {
              className: "break-words text-wrap"
            }, errorMessage)), {
              style: {
                padding: 0,
                margin: 0
              }
            });
          case 43:
            _context.prev = 43;
            setIsCalculating(false);
            return _context.finish(43);
          case 46:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[29, 36, 43, 46]]);
    }));
    return function (_x) {
      return _ref.apply(this, arguments);
    };
  }(), [tokenFrom, tokenTo, slippage, fromBalance, getQuote, setAmountTo, setQuoteResponse, setIsCalculating]);
  // Add this new function to handle manual refresh
  var handleRefreshQuote = useCallback(function () {
    if (!amountFrom || !tokenFrom || !tokenTo) return;
    // Clear existing timer if any
    if (refreshTimerRef.current) {
      clearTimeout(refreshTimerRef.current);
    }
    // Trigger immediate quote calculation
    calculateQuote(amountFrom);
    // Start new refresh timer
    refreshTimerRef.current = setTimeout(function () {
      handleRefreshQuote();
    }, 15000);
  }, [amountFrom, tokenFrom, tokenTo, calculateQuote]);
  // Add auto-refresh effect
  useEffect(function () {
    if (amountFrom && tokenFrom && tokenTo && !isCalculating && !isSwapping) {
      refreshTimerRef.current = setTimeout(function () {
        handleRefreshQuote();
      }, 15000);
    }
    return function () {
      if (refreshTimerRef.current) {
        clearTimeout(refreshTimerRef.current);
      }
    };
  }, [amountFrom, tokenFrom, tokenTo, isCalculating, isSwapping, handleRefreshQuote]);
  // Debounced quote calculation
  var debouncedCalculateQuote = useCallback(function (amount) {
    var debounced = setTimeout(function () {
      return calculateQuote(amount);
    }, 500);
    return function () {
      return clearTimeout(debounced);
    };
  }, [calculateQuote]);
  // Effect to handle input validation and trigger quote calculation
  useEffect(function () {
    var validateAndCalculate = function validateAndCalculate() {
      // Reset 'to' amount for empty/invalid input without showing error
      if (!amountFrom || isNaN(parseFloat(amountFrom))) {
        setAmountTo("0");
        setQuoteResponse(null);
        return;
      }
      var parsedAmount = parseFloat(amountFrom);
      // Skip calculation for zero or negative amounts without showing error
      if (parsedAmount <= 0) {
        setAmountTo("0");
        setQuoteResponse(null);
        return;
      }
      // Proceed with quote calculation if input is valid and positive
      debouncedCalculateQuote(amountFrom);
    };
    validateAndCalculate();
  }, [amountFrom, debouncedCalculateQuote, setAmountTo, setQuoteResponse]);
  // Calculate maximum input amount
  var calculateMaximumInput = useCallback(/*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee2() {
    var maxAmount;
    return _regeneratorRuntime().wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          if (!(!tokenFrom || !fromBalance || fromBalance <= 0)) {
            _context2.next = 2;
            break;
          }
          return _context2.abrupt("return");
        case 2:
          _context2.next = 4;
          return calculateMaxInput(tokenFrom, fromBalance);
        case 4:
          maxAmount = _context2.sent;
          return _context2.abrupt("return", maxAmount.toFixed(tokenFrom.decimals));
        case 6:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  })), [tokenFrom, fromBalance, calculateMaxInput]);
  // Handle swap execution
  var handleSwap = useCallback(/*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee3() {
    var loadingToast, quote, txid, errorMessage;
    return _regeneratorRuntime().wrap(function _callee3$(_context3) {
      while (1) switch (_context3.prev = _context3.next) {
        case 0:
          if (!(!tokenFrom || !tokenTo || !amountFrom || !amountTo)) {
            _context3.next = 3;
            break;
          }
          toast.error(/*#__PURE__*/React__default.createElement("div", {
            className: "flex flex-col bg-red-100 w-full p-4 rounded-lg"
          }, /*#__PURE__*/React__default.createElement("span", {
            className: "font-semibold"
          }, "Invalid Swap"), /*#__PURE__*/React__default.createElement("span", null, "Please ensure you have selected tokens and entered valid amounts")), {
            style: {
              padding: 0,
              margin: 0
            }
          });
          return _context3.abrupt("return");
        case 3:
          setIsSwapping(true);
          loadingToast = toast.loading(/*#__PURE__*/React__default.createElement("div", {
            className: "flex flex-col p-4"
          }, /*#__PURE__*/React__default.createElement("span", {
            className: "font-semibold"
          }, "Processing Transaction"), /*#__PURE__*/React__default.createElement("span", {
            className: "text-xs text-black/50"
          }, "Please wait while your transaction is being processed")), {
            style: {
              padding: 0,
              margin: 0
            }
          });
          _context3.prev = 5;
          _context3.next = 8;
          return getQuote(tokenFrom, tokenTo, parseFloat(amountFrom), slippage * 100);
        case 8:
          quote = _context3.sent;
          if (quote) {
            _context3.next = 11;
            break;
          }
          throw new Error("Failed to get quote");
        case 11:
          _context3.next = 13;
          return executeSwap(quote);
        case 13:
          txid = _context3.sent;
          if (txid) {
            _context3.next = 16;
            break;
          }
          throw new Error("Failed to execute swap");
        case 16:
          // Reset amounts and states only on successful swap
          setAmountFrom("");
          setAmountTo("0");
          setQuoteResponse(null);
          setError(null);
          toast.dismiss(loadingToast);
          toast.success(/*#__PURE__*/React__default.createElement("div", {
            className: "flex flex-col bg-green-100 w-full p-4  rounded-lg"
          }, /*#__PURE__*/React__default.createElement("span", {
            className: "font-semibold"
          }, "Swap Successful!"), /*#__PURE__*/React__default.createElement("span", {
            className: "text-xs text-black/50"
          }, /*#__PURE__*/React__default.createElement("span", {
            className: "break-words"
          }, "Transaction ID:"), /*#__PURE__*/React__default.createElement("a", {
            href: "https://solana.fm/tx/".concat(txid),
            target: "_blank",
            rel: "noopener noreferrer",
            className: "text-blue-500 hover:underline break-words"
          }, txid, " ", /*#__PURE__*/React__default.createElement(SquareArrowOutUpRightIcon, {
            className: "h-3 w-3 inline align-text-bottom"
          })))), {
            style: {
              padding: 0,
              margin: 0
            }
          });
          _context3.next = 24;
          return Promise.all([refetchFromBalance(), refetchToBalance()]);
        case 24:
          _context3.next = 34;
          break;
        case 26:
          _context3.prev = 26;
          _context3.t0 = _context3["catch"](5);
          errorMessage = _context3.t0 instanceof Error ? _context3.t0.message : "Failed to execute swap";
          toast.dismiss(loadingToast);
          // Check if user rejected/cancelled the transaction
          if (errorMessage.toLowerCase().includes("user rejected") || errorMessage.toLowerCase().includes("cancelled")) {
            toast.error(/*#__PURE__*/React__default.createElement("div", {
              className: "flex flex-col bg-red-100 w-full p-4 rounded-lg"
            }, /*#__PURE__*/React__default.createElement("span", {
              className: "font-semibold"
            }, "Transaction Cancelled"), /*#__PURE__*/React__default.createElement("span", null, "You cancelled the transaction")), {
              style: {
                padding: 0,
                margin: 0
              }
            });
          } else {
            toast.error(/*#__PURE__*/React__default.createElement("div", {
              className: "flex flex-col bg-red-100 w-full p-4 rounded-lg"
            }, /*#__PURE__*/React__default.createElement("span", {
              className: "font-semibold"
            }, "Transaction Failed"), /*#__PURE__*/React__default.createElement("span", {
              className: "break-words text-wrap"
            }, errorMessage)), {
              style: {
                padding: 0,
                margin: 0
              }
            });
          }
          console.error("Swap failed:", _context3.t0);
          setError(errorMessage);
          // Refetch quote to ensure prices are up to date
          if (amountFrom && !isNaN(parseFloat(amountFrom))) {
            calculateQuote(amountFrom);
          }
        case 34:
          _context3.prev = 34;
          setIsSwapping(false);
          return _context3.finish(34);
        case 37:
        case "end":
          return _context3.stop();
      }
    }, _callee3, null, [[5, 26, 34, 37]]);
  })), [tokenFrom, tokenTo, amountFrom, amountTo, slippage, getQuote, executeSwap, setAmountFrom, setAmountTo, setQuoteResponse, setIsCalculating, setIsSwapping, setError, refetchFromBalance, refetchToBalance]);
  return {
    calculateMaximumInput: calculateMaximumInput,
    handleSwap: handleSwap,
    handleRefreshQuote: handleRefreshQuote,
    fromBalance: fromBalance,
    toBalance: toBalance
  };
};

var formatBalance = function formatBalance(balance) {
  if (balance === null) return "0";
  return balance.toFixed(5);
};

var Swap = function Swap(_ref) {
  var rpcUrl = _ref.rpcUrl,
    referralKey = _ref.referralKey,
    _ref$platformFeeBps = _ref.platformFeeBps,
    platformFeeBps = _ref$platformFeeBps === void 0 ? 0 : _ref$platformFeeBps,
    apiKey = _ref.apiKey;
  var _useTokens = useTokens(),
    tokens = _useTokens.tokens;
  var _useWallet = useWallet(),
    connected = _useWallet.connected;
  var _useSwapStore = useSwapStore(),
    tokenFrom = _useSwapStore.tokenFrom,
    tokenTo = _useSwapStore.tokenTo,
    amountFrom = _useSwapStore.amountFrom,
    amountTo = _useSwapStore.amountTo,
    slippage = _useSwapStore.slippage,
    isFromDialogOpen = _useSwapStore.isFromDialogOpen,
    isToDialogOpen = _useSwapStore.isToDialogOpen,
    isCalculating = _useSwapStore.isCalculating,
    isSwapping = _useSwapStore.isSwapping,
    setTokenFrom = _useSwapStore.setTokenFrom,
    setTokenTo = _useSwapStore.setTokenTo,
    setAmountFrom = _useSwapStore.setAmountFrom,
    setSlippage = _useSwapStore.setSlippage,
    setIsFromDialogOpen = _useSwapStore.setIsFromDialogOpen,
    setIsToDialogOpen = _useSwapStore.setIsToDialogOpen,
    swapTokens = _useSwapStore.swapTokens;
  var _useSwapOperations = useSwapOperations({
      rpcUrl: rpcUrl,
      referralKey: referralKey,
      platformFeeBps: platformFeeBps,
      apiKey: apiKey
    }),
    calculateMaximumInput = _useSwapOperations.calculateMaximumInput,
    handleSwap = _useSwapOperations.handleSwap,
    handleRefreshQuote = _useSwapOperations.handleRefreshQuote,
    fromBalance = _useSwapOperations.fromBalance,
    toBalance = _useSwapOperations.toBalance;
  // Initialize default tokens only if no tokens are selected or persisted
  useEffect(function () {
    // Skip if tokens haven't loaded yet
    if (tokens.length === 0) return;
    // Skip if we already have tokens selected
    if (tokenFrom || tokenTo) return;
    // Check if we have persisted tokens in localStorage
    var persistedState = localStorage.getItem("swap-storage");
    if (persistedState) {
      var _JSON$parse = JSON.parse(persistedState),
        state = _JSON$parse.state;
      // Skip if we have persisted tokens
      if (state.tokenFrom || state.tokenTo) return;
    }
    // Set default tokens only if no tokens are selected or persisted
    var solToken = tokens.find(function (token) {
      return token.symbol === "SOL";
    });
    var usdcToken = tokens.find(function (token) {
      return token.symbol === "USDC";
    });
    if (solToken && usdcToken) {
      setTokenFrom(solToken);
      setTokenTo(usdcToken);
    }
  }, [tokens, tokenFrom, tokenTo]); // Run when tokens load or selection changes
  var isSwapDisabled = !connected || !tokenFrom || !tokenTo || !amountFrom || !amountTo || isCalculating || isSwapping;
  return /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement("div", {
    className: "w-full max-w-[480px] p-4 font-sans"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "rounded-3xl shadow-xl bg-background dark:bg-background-dark"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "p-6"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "flex justify-between items-center mb-6"
  }, /*#__PURE__*/React__default.createElement("h2", {
    className: "text-xl font-bold text-foreground dark:text-foreground-dark"
  }, "Swap"), /*#__PURE__*/React__default.createElement("div", {
    className: "flex items-center"
  }, /*#__PURE__*/React__default.createElement(SwapSettings, {
    slippage: slippage,
    onSlippageChange: setSlippage,
    onSave: function onSave() {}
  }), /*#__PURE__*/React__default.createElement(SwapReset, {
    onRefresh: handleRefreshQuote,
    isCalculating: isCalculating,
    isSwapping: isSwapping
  }))), /*#__PURE__*/React__default.createElement("div", {
    className: "themed-card rounded-2xl p-4 mb-2"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "flex justify-between mb-3"
  }, /*#__PURE__*/React__default.createElement("input", {
    type: "number",
    placeholder: "0",
    value: amountFrom,
    onChange: function onChange(e) {
      return setAmountFrom(e.target.value);
    },
    className: "w-full bg-transparent text-xl font-medium p-0 h-auto focus-visible:outline-none focus-visible:ring-0 [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none",
    disabled: isSwapping
  }), /*#__PURE__*/React__default.createElement(SwapTokenButton, {
    token: tokenFrom,
    onClick: function onClick() {
      return setIsFromDialogOpen(true);
    }
  })), /*#__PURE__*/React__default.createElement("div", {
    className: "flex justify-between items-center"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "text-sm text-muted-foreground dark:text-muted-dark-foreground"
  }, tokenFrom && "Balance: ".concat(formatBalance(fromBalance), " ").concat(tokenFrom.symbol)), /*#__PURE__*/React__default.createElement("div", {
    className: "flex gap-1.5"
  }, /*#__PURE__*/React__default.createElement("button", {
    className: "h-7 px-2.5 text-xs font-medium rounded-lg transition-colors hover:bg-accent hover:text-accent-foreground",
    onClick: function () {
      var _onClick = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
        var amount;
        return _regeneratorRuntime().wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              if (fromBalance && tokenFrom) {
                amount = (fromBalance * 0.25).toFixed(tokenFrom.decimals);
                setAmountFrom(amount);
              }
            case 1:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      function onClick() {
        return _onClick.apply(this, arguments);
      }
      return onClick;
    }(),
    disabled: !fromBalance || fromBalance <= 0
  }, "25%"), /*#__PURE__*/React__default.createElement("button", {
    className: "h-7 px-2.5 text-xs font-medium rounded-lg transition-colors hover:bg-accent hover:text-accent-foreground",
    onClick: function () {
      var _onClick2 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee2() {
        var amount;
        return _regeneratorRuntime().wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              if (fromBalance && tokenFrom) {
                amount = (fromBalance * 0.5).toFixed(tokenFrom.decimals);
                setAmountFrom(amount);
              }
            case 1:
            case "end":
              return _context2.stop();
          }
        }, _callee2);
      }));
      function onClick() {
        return _onClick2.apply(this, arguments);
      }
      return onClick;
    }(),
    disabled: !fromBalance || fromBalance <= 0
  }, "50%"), /*#__PURE__*/React__default.createElement("button", {
    className: "h-7 px-2.5 text-xs font-medium rounded-lg transition-colors hover:bg-accent hover:text-accent-foreground",
    onClick: function () {
      var _onClick3 = _asyncToGenerator(/*#__PURE__*/_regeneratorRuntime().mark(function _callee3() {
        var maxAmount;
        return _regeneratorRuntime().wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return calculateMaximumInput();
            case 2:
              maxAmount = _context3.sent;
              if (maxAmount) setAmountFrom(maxAmount);
            case 4:
            case "end":
              return _context3.stop();
          }
        }, _callee3);
      }));
      function onClick() {
        return _onClick3.apply(this, arguments);
      }
      return onClick;
    }(),
    disabled: !fromBalance || fromBalance <= 0
  }, "MAX")))), /*#__PURE__*/React__default.createElement("div", {
    className: "flex justify-center -my-3 relative z-10 mt-3 mb-3"
  }, /*#__PURE__*/React__default.createElement("button", {
    className: "h-9 w-9 cursor-pointer rounded-xl bg-card dark:bg-card-dark hover:bg-accent dark:hover:bg-accent transition-colors flex items-center justify-center shadow-md disabled:opacity-50",
    onClick: swapTokens,
    disabled: isSwapping,
    "aria-label": "Swap token positions"
  }, /*#__PURE__*/React__default.createElement(ArrowDown, {
    className: "h-4 w-4 text-foreground dark:text-foreground-dark"
  }))), /*#__PURE__*/React__default.createElement("div", {
    className: "themed-card rounded-2xl p-4 mt-2 bg-muted/20 dark:bg-muted-dark/20"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "flex justify-between mb-3"
  }, /*#__PURE__*/React__default.createElement("input", {
    type: "number",
    placeholder: "0",
    value: amountTo,
    readOnly: true,
    className: "w-full bg-transparent text-xl font-medium p-0 h-auto focus-visible:outline-none focus-visible:ring-0 [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
  }), /*#__PURE__*/React__default.createElement(SwapTokenButton, {
    token: tokenTo,
    onClick: function onClick() {
      return setIsToDialogOpen(true);
    }
  })), /*#__PURE__*/React__default.createElement("div", {
    className: "text-sm text-muted-foreground dark:text-muted-dark-foreground"
  }, tokenTo && "Balance: ".concat(formatBalance(toBalance), " ").concat(tokenTo.symbol))), connected ? (/*#__PURE__*/React__default.createElement("button", {
    className: "w-full h-10 mt-6 py-4 text-base font-semibold rounded-2xl themed-button-primary disabled:opacity-50 disabled:cursor-not-allowed flex items-center justify-center",
    onClick: handleSwap,
    disabled: isSwapDisabled
  }, isCalculating ? (/*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(Loader2, {
    className: "mr-2 h-5 w-5 animate-spin"
  }), "Loading...")) : isSwapping ? (/*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(Loader2, {
    className: "mr-2 h-5 w-5 animate-spin"
  }), "Swapping...")) : "Swap")) : (/*#__PURE__*/React__default.createElement(UnifiedWalletButton, {
    buttonClassName: "font-sans font-semibold rounded-3xl mt-6 themed-button-primary w-full px-4 py-3 text-sm flex justify-center items-center text-center gap-2 hover:opacity-90 dark:hover:opacity-90 transition-colors text-white dark:text-white cursor-pointer",
    overrideContent: /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(WalletIcon, {
      className: "h-4 w-4 opacity-70"
    }), "Connect Wallet")
  })), /*#__PURE__*/React__default.createElement("div", {
    className: "flex justify-center mt-6"
  }, /*#__PURE__*/React__default.createElement("a", {
    href: "https://jup.ag",
    target: "_blank",
    rel: "noopener noreferrer"
  }, /*#__PURE__*/React__default.createElement(ForwardRef$1, {
    width: 150,
    height: 25,
    className: "block dark:hidden opacity-75 hover:opacity-100 transition-opacity",
    preserveAspectRatio: "xMidYMid meet"
  }), /*#__PURE__*/React__default.createElement(ForwardRef, {
    width: 150,
    height: 25,
    className: "hidden dark:block opacity-75 hover:opacity-100 transition-opacity",
    preserveAspectRatio: "xMidYMid meet"
  }))))), /*#__PURE__*/React__default.createElement(SwapTokenDialog, {
    open: isFromDialogOpen,
    onClose: function onClose() {
      return setIsFromDialogOpen(false);
    },
    onSelect: setTokenFrom
  }), /*#__PURE__*/React__default.createElement(SwapTokenDialog, {
    open: isToDialogOpen,
    onClose: function onClose() {
      return setIsToDialogOpen(false);
    },
    onSelect: setTokenTo
  })));
};

var WalletAdapterNetwork;
(function (WalletAdapterNetwork) {
    WalletAdapterNetwork["Mainnet"] = "mainnet-beta";
    WalletAdapterNetwork["Testnet"] = "testnet";
    WalletAdapterNetwork["Devnet"] = "devnet";
})(WalletAdapterNetwork || (WalletAdapterNetwork = {}));

var _excluded = ["UnsafeBurnerWalletAdapter", "WalletConnectWalletAdapter"];
var useAllWallets = function useAllWallets(metadata, walletConnectProjectId) {
  var _useWrappedReownAdapt = useWrappedReownAdapter({
      appKitOptions: {
        metadata: {
          name: (metadata === null || metadata === void 0 ? void 0 : metadata.name) || "",
          description: (metadata === null || metadata === void 0 ? void 0 : metadata.description) || "",
          url: (metadata === null || metadata === void 0 ? void 0 : metadata.url) || "",
          icons: (metadata === null || metadata === void 0 ? void 0 : metadata.iconUrls) || []
        },
        projectId: walletConnectProjectId,
        features: {
          analytics: false,
          socials: ["google", "x", "apple"],
          email: false
        },
        enableWallets: false
      }
    }),
    reownAdapter = _useWrappedReownAdapt.reownAdapter,
    jupiterAdapter = _useWrappedReownAdapt.jupiterAdapter;
  return useMemo(function () {
    if (typeof window === "undefined") {
      return [];
    }
    var WalletConnectWalletAdapter = AllWalletAdapters.WalletConnectWalletAdapter,
      allwalletAdapters = _objectWithoutProperties(AllWalletAdapters, _excluded);
    var walletAdapters = Object.keys(allwalletAdapters).filter(function (key) {
      return key.includes("Adapter");
    }).map(function (key) {
      return allwalletAdapters[key];
    }).map(function (WalletAdapter) {
      return new WalletAdapter();
    });
    var walletConnectWalletAdapter = function () {
      var adapter = new WalletConnectWalletAdapter({
        network: WalletAdapterNetwork.Mainnet,
        options: {
          relayUrl: "wss://relay.walletconnect.com",
          projectId: walletConnectProjectId,
          metadata: {
            name: metadata === null || metadata === void 0 ? void 0 : metadata.name,
            description: metadata === null || metadata === void 0 ? void 0 : metadata.description,
            url: metadata === null || metadata === void 0 ? void 0 : metadata.url,
            icons: metadata === null || metadata === void 0 ? void 0 : metadata.iconUrls
          }
        }
      });
      adapter.supportedTransactionVersions = new Set(["legacy"]);
      return adapter;
    }();
    return [].concat(_toConsumableArray(walletAdapters), [walletConnectWalletAdapter, reownAdapter, jupiterAdapter]).filter(function (item) {
      return item && item.name && item.icon;
    });
  }, [metadata, walletConnectProjectId, reownAdapter, jupiterAdapter]);
};

function JupiverseKitProvider(_ref) {
  var children = _ref.children,
    _ref$autoConnect = _ref.autoConnect,
    autoConnect = _ref$autoConnect === void 0 ? true : _ref$autoConnect,
    _ref$env = _ref.env,
    env = _ref$env === void 0 ? "mainnet-beta" : _ref$env,
    _ref$endpoint = _ref.endpoint,
    endpoint = _ref$endpoint === void 0 ? "" : _ref$endpoint,
    _ref$walletConnectPro = _ref.walletConnectProjectId,
    walletConnectProjectId = _ref$walletConnectPro === void 0 ? "" : _ref$walletConnectPro,
    _ref$metadata = _ref.metadata,
    metadata = _ref$metadata === void 0 ? {
      name: "Jupiverse Kit Unified Wallet",
      description: "Jupiverse Kit Unified Wallet",
      url: "https://jupiversekit.vercel.app",
      iconUrls: ["https://jupiversekit.vercel.app/favicon.ico"]
    } : _ref$metadata,
    _ref$theme = _ref.theme,
    theme = _ref$theme === void 0 ? "light" : _ref$theme,
    _ref$lang = _ref.lang,
    lang = _ref$lang === void 0 ? "en" : _ref$lang,
    customWallets = _ref.wallets;
  var defaultWallets = useAllWallets(metadata, walletConnectProjectId);
  var wallets = customWallets || defaultWallets;
  return /*#__PURE__*/React__default.createElement(ConnectionProvider, {
    endpoint: endpoint
  }, /*#__PURE__*/React__default.createElement(UnifiedWalletProvider, {
    wallets: wallets,
    config: {
      autoConnect: autoConnect,
      env: env,
      metadata: {
        name: metadata.name,
        description: metadata.description,
        url: metadata.url,
        iconUrls: metadata.iconUrls
      },
      notificationCallback: {
        onConnect: function onConnect(props) {
          return toast.success(/*#__PURE__*/React__default.createElement("div", {
            className: "flex flex-col bg-green-100 w-full p-4  rounded-lg"
          }, /*#__PURE__*/React__default.createElement("span", {
            className: "font-semibold"
          }, "Wallet Connected"), /*#__PURE__*/React__default.createElement("span", {
            className: "text-xs text-black/50"
          }, "Connected to wallet ".concat(props.shortAddress))), {
            style: {
              padding: 0,
              margin: 0
            }
          });
        },
        onConnecting: function onConnecting(props) {
          return toast.message(/*#__PURE__*/React__default.createElement("div", {
            className: "flex flex-col p-4"
          }, /*#__PURE__*/React__default.createElement("span", {
            className: "font-semibold"
          }, "Connecting to ", props.walletName)), {
            style: {
              padding: 0,
              margin: 0
            }
          });
        },
        onDisconnect: function onDisconnect(props) {
          return toast.message(/*#__PURE__*/React__default.createElement("div", {
            className: "flex flex-col p-4"
          }, /*#__PURE__*/React__default.createElement("span", {
            className: "font-semibold"
          }, "Disconnected from ", props.walletName), /*#__PURE__*/React__default.createElement("span", {
            className: "text-xs text-black/50"
          }, "Disconnected from wallet ".concat(props.shortAddress))), {
            style: {
              padding: 0,
              margin: 0
            }
          });
        },
        onNotInstalled: function onNotInstalled(props) {
          return toast.error(/*#__PURE__*/React__default.createElement("div", {
            className: "flex flex-col bg-red-100 w-full p-4 rounded-lg"
          }, /*#__PURE__*/React__default.createElement("span", {
            className: "font-semibold"
          }, props.walletName, " Wallet is not installed"), /*#__PURE__*/React__default.createElement("span", null, "Please go to the provider", " ", /*#__PURE__*/React__default.createElement("a", {
            target: "_blank",
            rel: "noopener noreferrer",
            className: "underline font-bold",
            href: props.metadata.url
          }, "website"), " ", "to download.")), {
            style: {
              padding: 0,
              margin: 0
            }
          });
        }
      },
      theme: theme,
      lang: lang
    }
  }, children));
}

var useScriptLoader = function useScriptLoader(isMounted, initTerminal) {
  var _useState = useState(null),
    _useState2 = _slicedToArray(_useState, 2),
    scriptElement = _useState2[0],
    setScriptElement = _useState2[1];
  useEffect(function () {
    if (!isMounted) return;
    // Use a ref to track if init has been called
    var hasInitialized = false;
    var loadScript = function loadScript() {
      var existingScript = document.querySelector('script[src="https://terminal.jup.ag/main-v4.js"]');
      if (existingScript) {
        if (!hasInitialized) {
          hasInitialized = true;
          initTerminal();
        }
        return;
      }
      var script = document.createElement("script");
      script.src = "https://terminal.jup.ag/main-v4.js";
      script.async = true;
      script.defer = true;
      script.setAttribute("data-preload", "");
      script.onload = function () {
        if (!hasInitialized) {
          hasInitialized = true;
          initTerminal();
        }
      };
      script.onerror = function (error) {
        console.error("Failed to load Jupiter Terminal script:", error);
      };
      document.head.appendChild(script);
      setScriptElement(script);
    };
    // Debounce the script loading
    var timeoutId = setTimeout(loadScript, 1000);
    return function () {
      clearTimeout(timeoutId);
    };
  }, [initTerminal, isMounted]);
  return {
    scriptElement: scriptElement
  };
};

var useMount = function useMount() {
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    isMounted = _useState2[0],
    setIsMounted = _useState2[1];
  useEffect(function () {
    setIsMounted(true);
  }, []);
  return isMounted;
};

var useInitTerminal = function useInitTerminal(_ref) {
  var displayMode = _ref.displayMode,
    integratedTargetId = _ref.integratedTargetId,
    endpoint = _ref.endpoint,
    formProps = _ref.formProps,
    containerClassName = _ref.containerClassName,
    containerStyles = _ref.containerStyles,
    strictTokenList = _ref.strictTokenList,
    defaultExplorer = _ref.defaultExplorer,
    enableWalletPassthrough = _ref.enableWalletPassthrough,
    refetchIntervalForTokenAccounts = _ref.refetchIntervalForTokenAccounts,
    widgetStyle = _ref.widgetStyle,
    _onSuccess = _ref.onSuccess,
    _onSwapError = _ref.onSwapError;
  var initTimeoutRef = useRef();
  var isInitializedRef = useRef(false);
  return useCallback(function () {
    if (typeof window === "undefined" || !window.Jupiter || isInitializedRef.current) {
      return;
    }
    // Clear any pending initialization
    if (initTimeoutRef.current) {
      clearTimeout(initTimeoutRef.current);
    }
    // Debounce initialization to prevent rapid re-inits
    initTimeoutRef.current = setTimeout(function () {
      if (!isInitializedRef.current) {
        isInitializedRef.current = true;
        window.Jupiter.init({
          displayMode: displayMode,
          integratedTargetId: integratedTargetId,
          endpoint: endpoint,
          formProps: formProps,
          containerClassName: containerClassName,
          containerStyles: containerStyles,
          strictTokenList: strictTokenList,
          defaultExplorer: defaultExplorer,
          enableWalletPassthrough: enableWalletPassthrough,
          refetchIntervalForTokenAccounts: refetchIntervalForTokenAccounts,
          widgetStyle: widgetStyle,
          // Pass through widgetStyle configuration
          onSuccess: function onSuccess(_ref2) {
            var txid = _ref2.txid,
              swapResult = _ref2.swapResult;
            if (_onSuccess) {
              _onSuccess({
                txid: txid,
                swapResult: swapResult,
                quoteResponseMeta: null
              });
            }
          },
          onSwapError: function onSwapError(_ref3) {
            var error = _ref3.error;
            if (_onSwapError) {
              _onSwapError({
                error: error,
                quoteResponseMeta: null
              });
            }
          }
        });
      }
    }, 1000);
    return function () {
      if (initTimeoutRef.current) {
        clearTimeout(initTimeoutRef.current);
      }
    };
  }, [displayMode, integratedTargetId, endpoint, formProps, containerClassName, containerStyles, strictTokenList, defaultExplorer, enableWalletPassthrough, refetchIntervalForTokenAccounts, widgetStyle, _onSuccess, _onSwapError]);
};

var useJupiterTerminal = function useJupiterTerminal(props) {
  var isMounted = useMount();
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    scriptLoaded = _useState2[0],
    setScriptLoaded = _useState2[1];
  var initTerminal = useInitTerminal(props);
  var handleScriptLoad = function handleScriptLoad() {
    setScriptLoaded(true);
    if (!props.skipInit) {
      initTerminal();
    }
  };
  var _useScriptLoader = useScriptLoader(isMounted, handleScriptLoad),
    scriptElement = _useScriptLoader.scriptElement;
  var isCleanedUp = useRef(false);
  // Cleanup effect
  useEffect(function () {
    return function () {
      if (!isCleanedUp.current) {
        isCleanedUp.current = true;
        if (window.Jupiter) {
          window.Jupiter.close();
        }
        if (scriptElement) {
          document.head.removeChild(scriptElement);
        }
      }
    };
  }, [scriptElement]);
  return {
    isMounted: isMounted,
    scriptLoaded: scriptLoaded
  };
};

var getTerminalPlaceholder = function getTerminalPlaceholder(displayMode, integratedTargetId) {
  var placeholder = /*#__PURE__*/React__default.createElement("div", {
    id: integratedTargetId ? "integrated-terminal" : undefined
  });
  return displayMode === "integrated" ? placeholder : null;
};

var IntegratedTerminal = function IntegratedTerminal(_ref) {
  var _ref$rpcUrl = _ref.rpcUrl,
    rpcUrl = _ref$rpcUrl === void 0 ? "https://api.mainnet-beta.solana.com" : _ref$rpcUrl,
    formProps = _ref.formProps,
    containerClassName = _ref.containerClassName,
    containerStyles = _ref.containerStyles,
    onSuccess = _ref.onSuccess,
    onSwapError = _ref.onSwapError,
    _ref$strictTokenList = _ref.strictTokenList,
    strictTokenList = _ref$strictTokenList === void 0 ? true : _ref$strictTokenList,
    defaultExplorer = _ref.defaultExplorer,
    simulateWalletpassthrough = _ref.simulateWalletpassthrough,
    refetchIntervalForTokenAccounts = _ref.refetchIntervalForTokenAccounts;
  var _useJupiterTerminal = useJupiterTerminal({
      displayMode: "integrated",
      integratedTargetId: "integrated-terminal",
      endpoint: rpcUrl,
      formProps: formProps,
      containerClassName: containerClassName,
      containerStyles: containerStyles,
      strictTokenList: strictTokenList,
      defaultExplorer: defaultExplorer,
      enableWalletPassthrough: simulateWalletpassthrough,
      refetchIntervalForTokenAccounts: refetchIntervalForTokenAccounts,
      onSuccess: onSuccess,
      onSwapError: onSwapError
    }),
    isMounted = _useJupiterTerminal.isMounted;
  var placeholder = getTerminalPlaceholder("integrated", "integrated-terminal");
  if (!isMounted) {
    return placeholder;
  }
  return /*#__PURE__*/React__default.createElement(Suspense, {
    fallback: placeholder
  }, placeholder);
};

var WidgetTerminal = function WidgetTerminal(_ref) {
  var _ref$rpcUrl = _ref.rpcUrl,
    rpcUrl = _ref$rpcUrl === void 0 ? "https://api.mainnet-beta.solana.com" : _ref$rpcUrl,
    formProps = _ref.formProps,
    simulateWalletpassthrough = _ref.simulateWalletpassthrough,
    _ref$strictTokenList = _ref.strictTokenList,
    strictTokenList = _ref$strictTokenList === void 0 ? true : _ref$strictTokenList,
    defaultExplorer = _ref.defaultExplorer,
    refetchIntervalForTokenAccounts = _ref.refetchIntervalForTokenAccounts,
    _ref$widgetPosition = _ref.widgetPosition,
    widgetPosition = _ref$widgetPosition === void 0 ? "bottom-right" : _ref$widgetPosition,
    _ref$widgetSize = _ref.widgetSize,
    widgetSize = _ref$widgetSize === void 0 ? "default" : _ref$widgetSize,
    onSuccess = _ref.onSuccess,
    onSwapError = _ref.onSwapError;
  var _useJupiterTerminal = useJupiterTerminal({
      displayMode: "widget",
      endpoint: rpcUrl,
      formProps: formProps,
      strictTokenList: strictTokenList,
      defaultExplorer: defaultExplorer,
      enableWalletPassthrough: simulateWalletpassthrough,
      refetchIntervalForTokenAccounts: refetchIntervalForTokenAccounts,
      widgetStyle: {
        position: widgetPosition,
        size: widgetSize
      },
      onSuccess: onSuccess,
      onSwapError: onSwapError
    }),
    isMounted = _useJupiterTerminal.isMounted;
  var placeholder = getTerminalPlaceholder("widget");
  if (!isMounted) {
    return placeholder;
  }
  return /*#__PURE__*/React__default.createElement(Suspense, {
    fallback: placeholder
  }, placeholder);
};

var ModalTerminal = function ModalTerminal(_ref) {
  var _ref$rpcUrl = _ref.rpcUrl,
    rpcUrl = _ref$rpcUrl === void 0 ? "https://api.mainnet-beta.solana.com" : _ref$rpcUrl,
    formProps = _ref.formProps,
    simulateWalletpassthrough = _ref.simulateWalletpassthrough,
    _ref$strictTokenList = _ref.strictTokenList,
    strictTokenList = _ref$strictTokenList === void 0 ? true : _ref$strictTokenList,
    defaultExplorer = _ref.defaultExplorer,
    refetchIntervalForTokenAccounts = _ref.refetchIntervalForTokenAccounts,
    onSuccess = _ref.onSuccess,
    onSwapError = _ref.onSwapError,
    _ref$buttonText = _ref.buttonText,
    buttonText = _ref$buttonText === void 0 ? "Launch Modal Terminal" : _ref$buttonText,
    _ref$buttonClassName = _ref.buttonClassName,
    buttonClassName = _ref$buttonClassName === void 0 ? "p-4 rounded-xl cursor-pointer flex flex-col items-center justify-center transition-colors duration-300 bg-gray-800 text-white hover:bg-gray-700 dark:bg-gray-200 dark:text-black dark:hover:bg-gray-300 font-sans font-semibold" : _ref$buttonClassName;
  var _useJupiterTerminal = useJupiterTerminal({
      displayMode: "modal",
      endpoint: rpcUrl,
      formProps: formProps,
      strictTokenList: strictTokenList,
      defaultExplorer: defaultExplorer,
      enableWalletPassthrough: simulateWalletpassthrough,
      refetchIntervalForTokenAccounts: refetchIntervalForTokenAccounts,
      onSuccess: onSuccess,
      onSwapError: onSwapError,
      skipInit: true
    }),
    scriptLoaded = _useJupiterTerminal.scriptLoaded;
  return /*#__PURE__*/React__default.createElement("button", {
    type: "button",
    onClick: function onClick() {
      if (scriptLoaded && window.Jupiter) {
        window.Jupiter.init({
          displayMode: "modal",
          endpoint: rpcUrl,
          formProps: formProps,
          strictTokenList: strictTokenList,
          defaultExplorer: defaultExplorer,
          enableWalletPassthrough: simulateWalletpassthrough,
          refetchIntervalForTokenAccounts: refetchIntervalForTokenAccounts,
          onSuccess: onSuccess,
          onSwapError: onSwapError
        });
      }
    },
    className: buttonClassName
  }, /*#__PURE__*/React__default.createElement("span", {
    className: "text-xs"
  }, buttonText));
};

export { IntegratedTerminal, JupiverseKitProvider, ModalTerminal, Swap, WidgetTerminal };
