// @jup-ag/terminal/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __pow = Math.pow;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// package.json
var require_package = __commonJS({
  "package.json"(exports2, module2) {
    module2.exports = {
      name: "@jup-ag/terminal",
      version: "4.5.7",
      private: false,
      license: "MIT",
      scripts: {
        dev: "next dev",
        build: "next build",
        tsup: "tsup && npx tailwindcss --yes -i ./src/styles/globals.css -o ./dist/global.css",
        start: "next start",
        lint: "next lint",
        "format:fix": "prettier --write src",
        "build-widget": "NODE_ENV=production MODE=widget webpack",
        analyse: "NODE_ENV=production MODE=widget ANALYSE=true webpack"
      },
      files: [
        "dist"
      ],
      exports: {
        ".": "./dist/index.js",
        "./tailwind": "./dist/tailwind.config.js",
        "./css": "./dist/global.css"
      },
      typesVersions: {
        "*": {
          index: [
            "dist/index.d.ts"
          ]
        }
      },
      peerDependencies: {
        "@solana/spl-token": "^0.1.8",
        "@solana/web3.js": "^1.87.6",
        react: "^18",
        "react-dom": "^18"
      },
      dependencies: {
        "@jup-ag/wallet-adapter": "0.2.3",
        "@popperjs/core": "^2.11.8",
        "@solana/wallet-adapter-wallets": "0.19.33",
        "@tanstack/react-query": "^4.36.1",
        "bn.js": "5.2.1",
        clsx: "^2.1.1",
        "decimal.js": "10.4.3",
        jotai: "^2.8.3",
        jsbi: "4.3.0",
        "lodash.debounce": "^4.0.8",
        next: "13.4.19",
        "next-seo": "5.15.0",
        "react-colorful": "^5.6.1",
        "react-hook-form": "7.42.1",
        "react-number-format": "5.1.3",
        "react-popper": "^2.3.0",
        "react-syntax-highlighter": "^15.5.0",
        "react-use": "^17.5.0",
        "react-virtualized-auto-sizer": "1.0.7",
        "react-window": "1.8.8",
        superstruct: "1.0.3",
        "tailwind-merge": "^2.5.2",
        "tailwindcss-animate": "^1.0.7"
      },
      devDependencies: {
        "@svgr/webpack": "^6.5.1",
        "@types/bn.js": "^5.1.5",
        "@types/bs58": "^4.0.4",
        "@types/lodash.debounce": "^4.0.9",
        "@types/node": "18.11.5",
        "@types/react": "18.0.23",
        "@types/react-dom": "18.0.7",
        "@types/react-syntax-highlighter": "^15.5.13",
        "@types/react-virtualized-auto-sizer": "~1.0.4",
        "@types/react-window": "~1.8.8",
        autoprefixer: "10.4.13",
        "css-loader": "^6.11.0",
        "css-minimizer-webpack-plugin": "^4.2.2",
        cssnano: "^5.1.15",
        "esbuild-plugins-node-modules-polyfill": "^1.6.6",
        eslint: "8.26.0",
        "eslint-config-next": "13.0.0",
        "eslint-config-prettier": "^8.10.0",
        "mini-css-extract-plugin": "^2.9.0",
        "next-transpile-modules": "10.0.0",
        "node-polyfill-webpack-plugin": "^2.0.1",
        postcss: "^8.4.38",
        "postcss-loader": "^7.3.4",
        "postcss-preset-env": "^7.8.3",
        prettier: "^3.3.2",
        "react-twitter-embed": "^4.0.4",
        sass: "^1.77.6",
        "sass-loader": "^13.3.3",
        "style-loader": "^3.3.4",
        stylus: "^0.59.0",
        "svg-inline-loader": "^0.8.2",
        tailwindcss: "3.3.3",
        "ts-loader": "^9.5.1",
        tsup: "8.2.4",
        typescript: "5.2.2",
        webpack: "^5.92.1",
        "webpack-bundle-analyzer": "^4.10.2",
        "webpack-cli": "^4.10.0"
      },
      pnpm: {
        overrides: {
          "@solana/web3.js": "1.87.6",
          "@solana/buffer-layout": "4.0.0"
        }
      }
    };
  }
});

// src/npmlibary.tsx
var npmlibary_exports = {};
__export(npmlibary_exports, {
  appProps: () => appProps,
  close: () => close,
  init: () => init2,
  resume: () => resume,
  syncProps: () => syncProps
});
module.exports = __toCommonJS(npmlibary_exports);

// src/library.tsx
var import_client = require("react-dom/client");
var import_jotai2 = require("jotai");
var import_react2 = require("react");

// src/icons/JupiterLogo.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var JupiterLogo = ({ width = 24, height = 24 }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("img", { src: "https://jup.ag/svg/jupiter-logo.svg", width, height, alt: "Jupiter aggregator" });
};
var JupiterLogo_default = JupiterLogo;

// src/icons/ChevronDownIcon.tsx
var import_jsx_runtime2 = require("react/jsx-runtime");
var ChevronDownIcon = ({ width = 10, height = 6 }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
    "svg",
    {
      width,
      height,
      viewBox: "0 0 10 6",
      fill: "inherit",
      xmlns: "http://www.w3.org/2000/svg",
      children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
        "path",
        {
          fillRule: "evenodd",
          clipRule: "evenodd",
          d: "M0.292893 0.292893C0.683416 -0.097631 1.31658 -0.097631 1.7071 0.292893L4.99999 3.58579L8.29288 0.292893C8.6834 -0.0976311 9.31657 -0.0976311 9.70709 0.292893C10.0976 0.683417 10.0976 1.31658 9.70709 1.70711L5.7071 5.70711C5.31657 6.09763 4.68341 6.09763 4.29289 5.70711L0.292893 1.70711C-0.0976309 1.31658 -0.0976309 0.683417 0.292893 0.292893Z",
          fill: "currentColor"
        }
      )
    }
  );
};
var ChevronDownIcon_default = ChevronDownIcon;

// src/stores/jotai-terminal-in-view.tsx
var import_jotai = require("jotai");
var terminalInViewAtom = (0, import_jotai.atom)(false);
var getTerminalInView = () => {
  var _a2;
  const store2 = (_a2 = window.Jupiter) == null ? void 0 : _a2.store;
  if (!store2) {
    console.warn("Jupiter store is not available.");
    return false;
  }
  return store2.get(terminalInViewAtom);
};
var setTerminalInView = (value) => {
  var _a2;
  const store2 = (_a2 = window.Jupiter) == null ? void 0 : _a2.store;
  if (!store2) {
    console.warn("Jupiter store is not available.");
    return;
  }
  store2.set(terminalInViewAtom, value);
};

// src/misc/cn.ts
var import_clsx = require("clsx");
var import_tailwind_merge = require("tailwind-merge");
function cn(...inputs) {
  return (0, import_tailwind_merge.twMerge)((0, import_clsx.clsx)(inputs));
}

// src/components/ShadowDomContainer.tsx
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_jsx_runtime3 = require("react/jsx-runtime");
var ShadowDomContainer = ({ children, stylesheetUrls }) => {
  const hostRef = (0, import_react.useRef)(null);
  const [shadowRoot, setShadowRoot] = (0, import_react.useState)(null);
  const [stylesLoaded, setStylesLoaded] = (0, import_react.useState)(false);
  (0, import_react.useEffect)(() => {
    if (hostRef.current && !shadowRoot) {
      const newShadowRoot = hostRef.current.attachShadow({ mode: "open" });
      const loadStyles = () => __async(void 0, null, function* () {
        try {
          const stylePromises = stylesheetUrls.map(
            (url) => fetch(url).then((res) => {
              if (!res.ok) {
                throw new Error(`Failed to fetch stylesheet: ${url}`);
              }
              return res.text();
            })
          );
          const allCssText = yield Promise.all(stylePromises);
          allCssText.forEach((cssText) => {
            const styleEl = document.createElement("style");
            styleEl.textContent = cssText;
            newShadowRoot.appendChild(styleEl);
          });
          setStylesLoaded(true);
        } catch (error) {
          console.error("Error fetching and applying styles:", error);
        }
      });
      loadStyles();
      const portalContainer = document.createElement("div");
      portalContainer.id = "portal-container";
      portalContainer.style.width = "100%";
      portalContainer.style.height = "100%";
      newShadowRoot.appendChild(portalContainer);
      setShadowRoot(newShadowRoot);
    }
  }, [hostRef, shadowRoot, stylesheetUrls]);
  const portalTarget = shadowRoot == null ? void 0 : shadowRoot.getElementById("portal-container");
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { ref: hostRef, children: portalTarget && stylesLoaded ? (0, import_react_dom.createPortal)(children, portalTarget) : null });
};

// src/library.tsx
var import_jsx_runtime4 = require("react/jsx-runtime");
var containerId = "jupiter-terminal-instance";
var packageJson = require_package();
var bundleName = `main-${packageJson.version}`;
var scriptDomain = (() => {
  var _a2;
  if (typeof window === "undefined" || typeof document === "undefined") return "";
  const url = (_a2 = document.currentScript) == null ? void 0 : _a2.src;
  if (url) {
    return new URL(url).origin;
  }
  return "";
})() || "https://terminal.jup.ag";
function loadRemote(id, href, type2) {
  return __async(this, null, function* () {
    return new Promise((res, rej) => {
      const existing = document.getElementById(id);
      if (existing) {
        res({});
      } else {
        let el = type2 === "text/javascript" ? document.createElement("script") : document.createElement("link");
        el.id = id;
        el.onload = res;
        el.onerror = rej;
        if (el instanceof HTMLScriptElement) {
          el.type = "text/javascript";
          el.src = href;
        } else if (el instanceof HTMLLinkElement) {
          el.rel = "stylesheet";
          el.href = href;
        }
        document.head.append(el);
      }
    });
  });
}
function loadJupiter() {
  return __async(this, null, function* () {
    if (process.env.NODE_ENV === "development") {
      return;
    }
    try {
      yield loadRemote("jupiter-load-script-app", `${scriptDomain}/${bundleName}-app.js`, "text/javascript");
    } catch (error) {
      console.error(`Error loading Jupiter Terminal: ${error}`);
      throw new Error(`Error loading Jupiter Terminal: ${error}`);
    }
  });
}
var defaultStyles = {
  zIndex: 50
};
var RenderLoadableJupiter = (props) => {
  const [loaded, setLoaded] = (0, import_react2.useState)(false);
  (0, import_react2.useEffect)(() => {
    loadJupiter();
    let intervalId;
    if (!loaded) {
      intervalId = setInterval(() => {
        var _a2;
        const instance2 = (_a2 = window.JupiterRenderer) == null ? void 0 : _a2.RenderJupiter;
        if (instance2) {
          setLoaded(true);
        }
      }, 50);
    }
    return () => {
      clearInterval(intervalId);
    };
  }, [loaded]);
  const RenderJupiter2 = (0, import_react2.useMemo)(() => {
    if (loaded) {
      return window.JupiterRenderer.RenderJupiter;
    }
    return EmptyJSX;
  }, [loaded]);
  return /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(RenderJupiter2, __spreadValues({}, props));
};
var EmptyJSX = () => /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(import_jsx_runtime4.Fragment, {});
var RenderShell = (props) => {
  const displayMode = props.displayMode;
  const containerStyles = props.containerStyles;
  const containerClassName = props.containerClassName;
  (0, import_react2.useEffect)(() => setTerminalInView(true), []);
  const displayClassName = (0, import_react2.useMemo)(() => {
    if (!displayMode || displayMode === "modal") {
      return "fixed top-0 w-screen h-screen flex items-center justify-center bg-black/50  z-10";
    } else if (displayMode === "integrated" || displayMode === "widget") {
      return "flex items-center justify-center w-full h-full  z-10";
    }
  }, [displayMode]);
  const contentClassName = (0, import_react2.useMemo)(() => {
    if (!displayMode || displayMode === "modal" || displayMode === "integrated") {
      return `flex flex-col h-[550px] w-screen max-w-[360px] overflow-auto text-black relative bg-black rounded-lg webkit-scrollbar ${containerClassName || ""}`;
    } else if (displayMode === "widget") {
      return "flex flex-col  h-[550px] w-full overflow-auto text-black relative webkit-scrollbar";
    }
  }, [containerClassName, displayMode]);
  const onClose = () => {
    if (window.Jupiter) {
      setTerminalInView(false);
      window.Jupiter.close();
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime4.jsxs)("div", { className: displayClassName, children: [
    /* @__PURE__ */ (0, import_jsx_runtime4.jsx)("div", { style: __spreadValues(__spreadValues({}, defaultStyles), containerStyles), className: contentClassName, children: /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(RenderLoadableJupiter, __spreadValues({}, props)) }),
    !displayMode || displayMode === "modal" ? /* @__PURE__ */ (0, import_jsx_runtime4.jsx)("div", { onClick: onClose, className: "absolute w-screen h-screen top-0 left-0 backdrop-blur-sm" }) : null
  ] });
};
var RenderWidgetShell = (props) => {
  var _a2, _b, _c, _d, _e;
  const [isOpen, setIsOpen] = (0, import_react2.useState)(false);
  const classes = (0, import_react2.useMemo)(() => {
    var _a3, _b2, _c2, _d2, _e2, _f, _g, _h, _i, _j, _k, _l;
    const size = ((_a3 = props.widgetStyle) == null ? void 0 : _a3.size) || "default";
    const offsetX = (_d2 = (_c2 = (_b2 = props.widgetStyle) == null ? void 0 : _b2.offset) == null ? void 0 : _c2.x) != null ? _d2 : 0;
    const offsetY = (_g = (_f = (_e2 = props.widgetStyle) == null ? void 0 : _e2.offset) == null ? void 0 : _f.y) != null ? _g : 0;
    let result = void 0;
    if (!((_h = props.widgetStyle) == null ? void 0 : _h.position) || ((_i = props.widgetStyle) == null ? void 0 : _i.position) === "bottom-right") {
      result = {
        containerClassName: "bottom-6 right-6",
        contentClassName: size === "default" ? "bottom-[60px] -right-3" : "bottom-[44px] -right-4",
        style: { transform: `translate(-${offsetX}px, -${offsetY}px)` }
      };
    }
    if (((_j = props.widgetStyle) == null ? void 0 : _j.position) === "bottom-left") {
      result = {
        containerClassName: "bottom-6 left-6",
        contentClassName: size === "default" ? "bottom-[60px] -left-3" : "bottom-[44px] -left-4",
        style: { transform: `translate(${offsetX}px, -${offsetY}px)` }
      };
    }
    if (((_k = props.widgetStyle) == null ? void 0 : _k.position) === "top-left") {
      result = {
        containerClassName: "top-6 left-6",
        contentClassName: size === "default" ? "top-[60px] -left-3" : "top-[44px] -left-4",
        style: { transform: `translate(${offsetX}px, ${offsetY}px)` }
      };
    }
    if (((_l = props.widgetStyle) == null ? void 0 : _l.position) === "top-right") {
      result = {
        containerClassName: "top-6 right-6",
        contentClassName: size === "default" ? "top-[60px] -right-3" : "top-[44px] -right-4",
        style: { transform: `translate(-${offsetX}px, ${offsetY}px)` }
      };
    }
    return __spreadProps(__spreadValues({}, result), {
      widgetContainerClassName: size === "default" ? "h-14 w-14" : "h-10 w-10",
      widgetLogoSize: size === "default" ? 42 : 32
    });
  }, [(_a2 = props.widgetStyle) == null ? void 0 : _a2.position, (_b = props.widgetStyle) == null ? void 0 : _b.size, (_c = props.widgetStyle) == null ? void 0 : _c.offset]);
  return /* @__PURE__ */ (0, import_jsx_runtime4.jsxs)("div", { className: `fixed ${classes.containerClassName}`, style: classes.style, children: [
    /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
      "div",
      {
        className: `${classes.widgetContainerClassName} rounded-full bg-black flex items-center justify-center cursor-pointer`,
        onClick: () => {
          if (isOpen) {
            setIsOpen(false);
            setTerminalInView(false);
          } else {
            setIsOpen(true);
            setTerminalInView(true);
          }
        },
        children: isOpen ? /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
          "div",
          {
            className: cn("text-white fill-current pt-1", {
              "rotate-180": ((_d = props.widgetStyle) == null ? void 0 : _d.position) === "top-left" || ((_e = props.widgetStyle) == null ? void 0 : _e.position) === "top-right"
            }),
            children: /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(ChevronDownIcon_default, { width: classes.widgetLogoSize * 0.4, height: classes.widgetLogoSize * 0.4 })
          }
        ) : /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(JupiterLogo_default, { width: classes.widgetLogoSize, height: classes.widgetLogoSize })
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
      "div",
      {
        id: "integrated-terminal",
        className: `absolute overflow-hidden ${classes.contentClassName} flex flex-col w-[90vw] h-[600px] max-w-[384px] max-h-[75vh] rounded-2xl bg-black transition-opacity duration-300 shadow-2xl ${!isOpen ? "!h-0 !w-0 opacity-0" : "opacity-100"}`,
        children: /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(RenderLoadableJupiter, __spreadValues({}, props))
      }
    )
  ] });
};
var store = (0, import_jotai2.createStore)();
var appProps = (0, import_jotai2.atom)(void 0);
function init(passedProps) {
  return __async(this, null, function* () {
    var _b;
    if (typeof window === "undefined" || typeof document === "undefined") return null;
    const props = __spreadValues({}, passedProps);
    const _a2 = props, {
      enableWalletPassthrough,
      passthroughWalletContextState,
      onRequestConnectWallet,
      onSwapError,
      onSuccess,
      onFormUpdate,
      onScreenUpdate,
      integratedTargetId
    } = _a2, restProps = __objRest(_a2, [
      "enableWalletPassthrough",
      "passthroughWalletContextState",
      "onRequestConnectWallet",
      "onSwapError",
      "onSuccess",
      "onFormUpdate",
      "onScreenUpdate",
      "integratedTargetId"
    ]);
    const targetDiv = document.createElement("div");
    const instanceExist = document.getElementById(containerId);
    window.Jupiter.store = store;
    store.set(appProps, __spreadProps(__spreadValues({}, props), { scriptDomain }));
    setTerminalInView(false);
    if (instanceExist) {
      window.Jupiter._instance = null;
      instanceExist.remove();
      (_b = window.Jupiter.root) == null ? void 0 : _b.unmount();
    }
    targetDiv.id = containerId;
    targetDiv.classList.add("w-full");
    targetDiv.classList.add("h-full");
    if (restProps.displayMode === "integrated") {
      const target = document.getElementById(integratedTargetId);
      if (!target) {
        throw new Error(`Jupiter Terminal: document.getElementById cannot find ${integratedTargetId}`);
      }
      target == null ? void 0 : target.appendChild(targetDiv);
    } else {
      document.body.appendChild(targetDiv);
    }
    if (enableWalletPassthrough) {
      window.Jupiter.enableWalletPassthrough = true;
      window.Jupiter.onRequestConnectWallet = onRequestConnectWallet;
    } else {
      window.Jupiter.enableWalletPassthrough = false;
    }
    let element;
    if (restProps.displayMode === "widget") {
      element = /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(RenderWidgetShell, __spreadValues({}, props));
    } else {
      element = /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(RenderShell, __spreadValues({}, props));
    }
    const stylesheetUrls = [
      `${scriptDomain}/${bundleName}-Tailwind.css`,
      `${scriptDomain}/scoped-preflight-v4.css`,
      `${scriptDomain}/${bundleName}-Jupiter.css`,
      "https://fonts.googleapis.com/css2?family=Inter:wght@100;200;300;400;500;600;700;800;900&family=Poppins&display=swap"
    ];
    const root = (0, import_client.createRoot)(targetDiv);
    root.render(/* @__PURE__ */ (0, import_jsx_runtime4.jsx)(ShadowDomContainer, { stylesheetUrls, children: element }));
    window.Jupiter.root = root;
    window.Jupiter._instance = element;
    window.Jupiter.onSwapError = onSwapError;
    window.Jupiter.onSuccess = onSuccess;
    window.Jupiter.onFormUpdate = onFormUpdate;
    window.Jupiter.onScreenUpdate = onScreenUpdate;
    window.Jupiter.localStoragePrefix = passedProps.localStoragePrefix || "jupiter-terminal";
  });
}
var _a;
var attributes = typeof document !== "undefined" ? (_a = document.currentScript) == null ? void 0 : _a.attributes : void 0;
if (typeof window !== "undefined" && typeof document !== "undefined" && attributes) {
  document.onreadystatechange = function() {
    const loadComplete = document.readyState === "complete";
    const shouldPreload = Boolean(attributes.getNamedItem("data-preload"));
    if (loadComplete && shouldPreload) {
      setTimeout(() => {
        loadJupiter().catch((error) => {
          console.error(`Error pre-loading Jupiter Terminal: ${error}`);
          throw new Error(`Error pre-loading Jupiter Terminal: ${error}`);
        });
      }, 2e3);
    }
  };
}
var resume = () => {
  const instanceExist = document.getElementById(containerId);
  if (instanceExist) {
    instanceExist.style.display = "block";
    return;
  }
};
var close = () => {
  const targetDiv = document.getElementById(containerId);
  if (targetDiv) {
    targetDiv.style.display = "none";
  }
};
var syncProps = (props) => {
  const currentProps = store.get(appProps);
  const newProps = __spreadProps(__spreadValues({}, currentProps), {
    passthroughWalletContextState: props.passthroughWalletContextState || (currentProps == null ? void 0 : currentProps.passthroughWalletContextState)
  });
  store.set(appProps, newProps);
};

// src/index.tsx
var import_jotai4 = require("jotai");

// src/contexts/ScreenProvider.tsx
var import_react3 = require("react");
var import_jsx_runtime5 = require("react/jsx-runtime");
var ScreenStateContext = (0, import_react3.createContext)({ screen: "Initial", setScreen() {
} });
function useScreenState() {
  return (0, import_react3.useContext)(ScreenStateContext);
}
var ScreenProvider = ({ children }) => {
  const [screen2, setScreen] = (0, import_react3.useState)("Initial");
  return /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(ScreenStateContext.Provider, { value: { screen: screen2, setScreen }, children });
};

// src/contexts/SwapContext.tsx
var import_decimal2 = __toESM(require("decimal.js"));
var import_jsbi2 = __toESM(require("jsbi"));
var import_react7 = require("react");

// src/constants/index.ts
var import_web3 = require("@solana/web3.js");
var USDC_MINT = new import_web3.PublicKey("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v");
var USDT_MINT = new import_web3.PublicKey("Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB");
var JLP_MINT = new import_web3.PublicKey("27G8MtK7VtTcCHkpASjSDdkWWYfoqT6ggEuKidVJidD4");
var WRAPPED_SOL_MINT = new import_web3.PublicKey("So11111111111111111111111111111111111111112");
var INITIAL_FORM_CONFIG = Object.freeze({
  simulateWalletPassthrough: false,
  defaultExplorer: "Solana Explorer",
  formProps: {
    fixedAmount: false,
    initialAmount: "",
    initialInputMint: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
    initialOutputMint: WRAPPED_SOL_MINT.toString()
  },
  colors: {
    primary: "199, 242, 132",
    background: "0, 0, 0",
    primaryText: "232, 249, 255",
    warning: "251, 191, 36",
    interactive: "33, 42, 54",
    module: "16, 23, 31"
  }
});

// src/misc/utils.ts
var import_web32 = require("@solana/web3.js");
var import_decimal = __toESM(require("decimal.js"));
var import_react4 = require("react");

// src/misc/tokenTags.ts
var checkIsStrictOrVerified = (tokenInfo) => {
  var _a2, _b, _c;
  return Boolean(
    ((_a2 = tokenInfo.tags) == null ? void 0 : _a2.includes("verified")) || ((_b = tokenInfo.tags) == null ? void 0 : _b.includes("strict")) || ((_c = tokenInfo.tags) == null ? void 0 : _c.includes("community"))
  );
};
var checkIsUnknownToken = (tokenInfo) => {
  return checkIsStrictOrVerified(tokenInfo) === false;
};
var checkIsToken2022 = (tokenInfo) => {
  var _a2;
  return (_a2 = tokenInfo.tags) == null ? void 0 : _a2.includes("token-2022");
};

// src/misc/utils.ts
var userLocale = typeof window !== "undefined" ? navigator.languages && navigator.languages.length ? navigator.languages[0] : navigator.language : "en-US";
var numberFormatter = new Intl.NumberFormat(userLocale, {
  style: "decimal",
  minimumFractionDigits: 0,
  maximumFractionDigits: 9
});
var getDecimalCount = (value) => {
  const parts = value.split(".");
  return parts.length > 1 ? parts[1].length : 0;
};
var formatNumber = {
  format: (val, precision) => {
    if (!val) {
      return "";
    }
    const defaultDecimals = getDecimalCount(val.toString());
    const numberFormatter2 = new Intl.NumberFormat(userLocale, {
      maximumFractionDigits: precision != null ? precision : defaultDecimals
    });
    return numberFormatter2.format(val.toString());
  }
};
function shortenAddress(address, chars = 4) {
  return `${address.slice(0, chars)}...${address.slice(-chars)}`;
}
function readableValue(lamportsAmount, decimals) {
  return new import_decimal.default(lamportsAmount.toString()).div(__pow(10, decimals)).toDP(decimals, import_decimal.default.ROUND_DOWN).toFixed();
}
function fromLamports(lamportsAmount, decimals) {
  return new import_decimal.default(lamportsAmount.toString()).div(__pow(10, decimals)).toDP(decimals, import_decimal.default.ROUND_DOWN).toNumber();
}
var detectedSeparator = formatNumber.format("1.1").substring(1, 2);
function useDebounce(value, delay) {
  const [debouncedValue, setDebouncedValue] = (0, import_react4.useState)(value);
  (0, import_react4.useEffect)(
    () => {
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, delay);
      return () => {
        clearTimeout(handler);
      };
    },
    [value, delay]
    // Only re-call effect if value or delay changes
  );
  return debouncedValue;
}
function useOutsideClick(ref, handler) {
  (0, import_react4.useEffect)(() => {
    const listener = (event) => {
      if (!ref.current || ref.current.contains(event.target)) {
        return;
      }
      handler(event);
    };
    document.addEventListener("mouseup", listener);
    return () => {
      document.removeEventListener("mouseup", listener);
    };
  }, [ref, handler]);
}
var hasNumericValue = (amount) => {
  if (amount && !Number.isNaN(Number(amount))) {
    return true;
  }
  return false;
};
var sortByUserBalance = (searchResults, balances) => {
  const userBalanceMap = /* @__PURE__ */ new Map();
  const userUsdValueMap = /* @__PURE__ */ new Map();
  Object.entries(balances).filter(([_, item]) => item.uiAmount > 0).forEach(([mint, item]) => {
    const tokenInfo = searchResults.find((token) => token.id === mint);
    if (!tokenInfo) return;
    const amount = item.uiAmount;
    userBalanceMap.set(mint, amount);
    const tokenPrice = tokenInfo.usdPrice || 0;
    if (tokenPrice) {
      const usdValue = new import_decimal.default(amount).mul(tokenPrice);
      if (usdValue.greaterThan(0)) {
        userUsdValueMap.set(mint, usdValue);
      }
    }
  });
  const deduplicatedResults = (() => {
    const map = /* @__PURE__ */ new Map();
    searchResults.forEach((item) => {
      if (!map.has(item.id)) {
        map.set(item.id, item);
      }
    });
    return Array.from(map.values());
  })();
  return deduplicatedResults.sort((t1, t2) => {
    var _a2, _b, _c, _d;
    const t1UsdValue = userUsdValueMap.get(t1.id);
    const t2UsdValue = userUsdValueMap.get(t2.id);
    if (t1UsdValue && t2UsdValue) {
      const usdComparison = t2UsdValue.cmp(t1UsdValue);
      if (usdComparison !== 0) return usdComparison;
    } else if (t1UsdValue && !t2UsdValue) {
      return -1;
    } else if (!t1UsdValue && t2UsdValue) {
      return 1;
    }
    const t1Balance = userBalanceMap.get(t1.id);
    const t2Balance = userBalanceMap.get(t2.id);
    if (t1Balance && t2Balance) {
      const balanceComparison = new import_decimal.default(t2Balance).cmp(t1Balance);
      if (balanceComparison !== 0) return balanceComparison;
    } else if (t1Balance && !t2Balance) {
      return -1;
    } else if (!t1Balance && t2Balance) {
      return 1;
    }
    const t1HasBalance = userBalanceMap.has(t1.id);
    const t2HasBalance = userBalanceMap.has(t2.id);
    if (t1HasBalance && !t2HasBalance) return -1;
    if (!t1HasBalance && t2HasBalance) return 1;
    let t1Score = 0;
    let t2Score = 0;
    const t1Volume = (((_a2 = t1.stats24h) == null ? void 0 : _a2.buyVolume) || 0) + (((_b = t1.stats24h) == null ? void 0 : _b.sellVolume) || 0);
    const t2Volume = (((_c = t2.stats24h) == null ? void 0 : _c.buyVolume) || 0) + (((_d = t2.stats24h) == null ? void 0 : _d.sellVolume) || 0);
    if (t1Volume > t2Volume) t1Score += 1;
    if (t2Volume > t1Volume) t2Score += 1;
    if (checkIsUnknownToken(t1)) t1Score -= 2;
    if (checkIsUnknownToken(t2)) t2Score -= 2;
    return t2Score - t1Score;
  });
};

// src/contexts/WalletPassthroughProvider.tsx
var import_wallet_adapter = require("@jup-ag/wallet-adapter");
var import_jotai3 = require("jotai");
var import_react5 = require("react");
var import_jsx_runtime6 = require("react/jsx-runtime");
var initialPassThrough = {
  publicKey: null,
  wallets: [],
  wallet: null,
  connect: () => __async(void 0, null, function* () {
  }),
  select: () => {
  },
  connecting: false,
  connected: false,
  disconnect: () => __async(void 0, null, function* () {
  }),
  autoConnect: false,
  disconnecting: false,
  sendTransaction: (transaction, connection, options) => __async(void 0, null, function* () {
    return "";
  }),
  signTransaction: void 0,
  signAllTransactions: void 0,
  signMessage: void 0,
  signIn: void 0
};
var WalletPassthroughContext = (0, import_react5.createContext)(initialPassThrough);
function useWalletPassThrough() {
  return (0, import_react5.useContext)(WalletPassthroughContext);
}
var FromWalletAdapter = ({ children }) => {
  const walletContextState = (0, import_wallet_adapter.useWallet)();
  return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(WalletPassthroughContext.Provider, { value: walletContextState, children });
};
var WalletPassthroughProvider = ({ children }) => {
  var _a2;
  const [atom3] = (0, import_jotai3.useAtom)(appProps);
  const wallet = (_a2 = atom3 == null ? void 0 : atom3.passthroughWalletContextState) == null ? void 0 : _a2.wallet;
  const walletPassthrough = (0, import_react5.useMemo)(() => {
    return __spreadProps(__spreadValues(__spreadValues({}, initialPassThrough), atom3 == null ? void 0 : atom3.passthroughWalletContextState), {
      disconnect: () => __async(void 0, null, function* () {
        var _a3, _b;
        try {
          if ((_a3 = wallet == null ? void 0 : wallet.adapter) == null ? void 0 : _a3.disconnect) {
            return (_b = wallet == null ? void 0 : wallet.adapter) == null ? void 0 : _b.disconnect();
          }
        } catch (error) {
          console.log(error);
        }
      })
    });
  }, [atom3 == null ? void 0 : atom3.passthroughWalletContextState, wallet == null ? void 0 : wallet.adapter]);
  if (!window.Jupiter.enableWalletPassthrough) {
    return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(FromWalletAdapter, { children });
  }
  if (walletPassthrough) {
    return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(WalletPassthroughContext.Provider, { value: walletPassthrough, children });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(import_jsx_runtime6.Fragment, { children });
};
var WalletPassthroughProvider_default = WalletPassthroughProvider;

// src/queries/useQuoteQuery.ts
var import_react_query = require("@tanstack/react-query");

// src/data/UltraSwapService.ts
var UltraSwapService = class {
  constructor() {
    this.BASE_URL = "https://ultra-api.jup.ag";
    this.ROUTE = {
      SWAP: `${this.BASE_URL}/execute`,
      ORDER: `${this.BASE_URL}/order`,
      ROUTERS: `${this.BASE_URL}/order/routers`,
      BALANCES: `${this.BASE_URL}/balances`,
      SHIELD: `${this.BASE_URL}/shield`
    };
  }
  getQuote(params, signal) {
    return __async(this, null, function* () {
      const queryParams = new URLSearchParams(
        Object.entries(params).filter(([_, value]) => value !== void 0).reduce(
          (acc, [key, value]) => __spreadProps(__spreadValues({}, acc), {
            [key]: value.toString()
          }),
          {}
        )
      );
      const response = yield fetch(`${this.ROUTE.ORDER}?${queryParams.toString()}`, { signal });
      if (!response.ok) {
        throw response;
      }
      const result = yield response.json();
      return result;
    });
  }
  submitSwap(signedTransaction, requestId) {
    return __async(this, null, function* () {
      const response = yield fetch(this.ROUTE.SWAP, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ signedTransaction, requestId })
      });
      if (!response.ok) {
        throw response;
      }
      const result = yield response.json();
      return result;
    });
  }
  getRouters() {
    return __async(this, null, function* () {
      const response = yield fetch(this.ROUTE.ROUTERS);
      if (!response.ok) {
        throw response;
      }
      const result = yield response.json();
      return result;
    });
  }
  getBalance(address, signal) {
    return __async(this, null, function* () {
      const response = yield fetch(`${this.ROUTE.BALANCES}/${address}`, { signal });
      if (!response.ok) {
        throw response;
      }
      const result = yield response.json();
      return result;
    });
  }
  getShield(mintAddress) {
    return __async(this, null, function* () {
      const response = yield fetch(`${this.ROUTE.SHIELD}?mints=${mintAddress.join(",")}`);
      if (!response.ok) {
        throw response;
      }
      const result = yield response.json();
      return result;
    });
  }
};
var ultraSwapService = new UltraSwapService();

// src/entity/FormattedUltraQuoteResponse.ts
var import_web33 = require("@solana/web3.js");
var import_jsbi = __toESM(require("jsbi"));
var import_superstruct = require("superstruct");
var PublicKeyFromString = (0, import_superstruct.coerce)((0, import_superstruct.instance)(import_web33.PublicKey), (0, import_superstruct.string)(), (value) => new import_web33.PublicKey(value));
var AmountFromString = (0, import_superstruct.coerce)((0, import_superstruct.instance)(import_jsbi.default), (0, import_superstruct.string)(), (value) => import_jsbi.default.BigInt(value));
var NumberFromString = (0, import_superstruct.coerce)((0, import_superstruct.string)(), (0, import_superstruct.number)(), (value) => Number(value));
var SwapInfo = (0, import_superstruct.type)({
  ammKey: PublicKeyFromString,
  label: (0, import_superstruct.string)(),
  inputMint: (0, import_superstruct.string)(),
  outputMint: (0, import_superstruct.string)(),
  inAmount: AmountFromString,
  outAmount: AmountFromString,
  feeAmount: AmountFromString,
  feeMint: PublicKeyFromString
});
var RoutePlanStep = (0, import_superstruct.type)({
  swapInfo: SwapInfo,
  percent: (0, import_superstruct.number)()
});
var RoutePlanWithMetadata = (0, import_superstruct.array)(RoutePlanStep);
var FormattedUltraQuoteResponse = (0, import_superstruct.type)({
  inputMint: PublicKeyFromString,
  inAmount: AmountFromString,
  outputMint: PublicKeyFromString,
  outAmount: AmountFromString,
  otherAmountThreshold: AmountFromString,
  priceImpactPct: NumberFromString,
  routePlan: RoutePlanWithMetadata,
  slippageBps: (0, import_superstruct.number)(),
  contextSlot: (0, import_superstruct.defaulted)((0, import_superstruct.number)(), 0),
  computedAutoSlippage: (0, import_superstruct.optional)((0, import_superstruct.number)()),
  transaction: (0, import_superstruct.nullable)((0, import_superstruct.string)()),
  swapType: (0, import_superstruct.string)(),
  gasless: (0, import_superstruct.boolean)(),
  requestId: (0, import_superstruct.string)(),
  prioritizationFeeLamports: (0, import_superstruct.optional)((0, import_superstruct.number)()),
  feeBps: (0, import_superstruct.number)(),
  router: (0, import_superstruct.string)()
});

// src/queries/useQuoteQuery.ts
var import_superstruct2 = require("superstruct");
var useQuoteQuery = (params, shouldRefetch = true) => {
  const { amount } = params;
  return (0, import_react_query.useQuery)({
    queryKey: ["quote", params],
    queryFn: (_0) => __async(void 0, [_0], function* ({ signal }) {
      if (Number(amount) === 0) {
        return null;
      }
      try {
        const response = yield ultraSwapService.getQuote(params, signal);
        const quoteResponse = (0, import_superstruct2.create)(response, FormattedUltraQuoteResponse, "conver FormattedUltraQuoteResponse Error");
        return {
          quoteResponse,
          original: response
        };
      } catch (e) {
        if (e instanceof Response) {
          const errorObj = yield e.json();
          throw errorObj.error;
        }
        throw e;
      }
    }),
    refetchInterval: shouldRefetch ? 5e3 : false,
    retry: 0,
    enabled: Number(amount) > 0,
    keepPreviousData: Number(amount) > 0,
    onError: (error) => {
      console.error("useQuoteQuery", error);
    }
  });
};

// src/queries/useUltraSwapMutation.ts
var import_wallet_adapter2 = require("@jup-ag/wallet-adapter");
var import_react_query2 = require("@tanstack/react-query");
var import_buffer = require("buffer");
var import_web34 = require("@solana/web3.js");

// src/entity/TransactionError.ts
var TransactionError = class extends Error {
  constructor(message, txid, code, programIds, serializedTx) {
    super(message);
    this.name = "TransactionError";
    this.txid = txid;
    this.code = code;
    this.programIds = programIds;
    this.serializedTx = serializedTx;
  }
};

// src/queries/useUltraSwapMutation.ts
var UltraSwapError = class extends Error {
  constructor(message, type2, txid) {
    super(message);
    this.type = type2;
    this.txid = txid;
  }
};
function useUltraSwapMutation() {
  const { wallet, signTransaction } = useWalletPassThrough();
  return (0, import_react_query2.useMutation)({
    mutationFn: (_0) => __async(this, [_0], function* ({
      setTxStatus,
      setLastSwapResult,
      fromTokenInfo,
      toTokenInfo,
      quoteResponseMeta
    }) {
      const publicKey = wallet == null ? void 0 : wallet.adapter.publicKey;
      if (!signTransaction || !publicKey) {
        throw new UltraSwapError(
          "Wallet not connected, or missing wallet functions",
          "WALLET_SIGNING_FAILED" /* WALLET_SIGNING_FAILED */
        );
      }
      setTxStatus({
        txid: "",
        status: "pending-approval"
      });
      const selectedQuote = quoteResponseMeta.original;
      const { transaction, requestId } = selectedQuote;
      if (!transaction) throw new Error("Missing transaction");
      const based64tx = import_buffer.Buffer.from(transaction, "base64");
      const versionedTransaction = import_web34.VersionedTransaction.deserialize(new Uint8Array(based64tx));
      const signedTransaction = yield signTransaction(versionedTransaction);
      const serializedTransaction = import_buffer.Buffer.from(signedTransaction.serialize()).toString("base64");
      setTxStatus({
        txid: "",
        status: "sending"
      });
      const response = yield ultraSwapService.submitSwap(serializedTransaction, requestId);
      const { signature, status } = response;
      if (status === "Failed") {
        throw new UltraSwapError(response.error, "FAILED" /* FAILED */, signature);
      }
      const { inputAmountResult, outputAmountResult } = response;
      setTxStatus({
        txid: signature,
        status: "success"
      });
      setLastSwapResult({
        swapResult: {
          txid: signature,
          inputAddress: new import_web34.PublicKey(fromTokenInfo.id),
          outputAddress: new import_web34.PublicKey(toTokenInfo.id),
          inputAmount: Number(inputAmountResult),
          outputAmount: Number(outputAmountResult)
        },
        quoteReponse: quoteResponseMeta
      });
      return signature;
    }),
    onError: (error, variables) => __async(this, null, function* () {
      const { setTxStatus, setLastSwapResult, quoteResponseMeta } = variables;
      if (error instanceof import_wallet_adapter2.WalletSignTransactionError) {
        const message = error.message || error.error || "Transaction cancelled";
        setLastSwapResult({
          swapResult: {
            error: new TransactionError(message)
          },
          quoteReponse: quoteResponseMeta
        });
        return;
      }
      if (error instanceof Error) {
        setLastSwapResult({
          swapResult: {
            error: new TransactionError(error.message)
          },
          quoteReponse: quoteResponseMeta
        });
        return;
      }
      if ("json" in error) {
        const json = yield error.json();
        setLastSwapResult({
          swapResult: {
            error: new TransactionError(json.error || "Unknown error")
          },
          quoteReponse: quoteResponseMeta
        });
        setTxStatus({
          txid: json.txid || "",
          status: "fail"
        });
      }
    })
  });
}

// src/hooks/useBalances.ts
var import_react_query3 = require("@tanstack/react-query");
var import_react6 = require("react");
var ULTRA_NATIVESOL_ID = "SOL";
var WSOL_ID = WRAPPED_SOL_MINT.toString();
var WRAPPED_SOL_ID = "wSOL";
function combineSolAmounts(data) {
  const balances = structuredClone(data);
  if (balances[WSOL_ID]) {
    balances[WRAPPED_SOL_ID] = __spreadValues({}, balances[WSOL_ID]);
  }
  if (balances[ULTRA_NATIVESOL_ID]) {
    balances[WSOL_ID] = __spreadValues({}, balances[ULTRA_NATIVESOL_ID]);
  }
  return balances;
}
var useBalances = () => {
  const { publicKey } = useWalletPassThrough();
  const address = (0, import_react6.useMemo)(() => {
    if (!publicKey) return "";
    return publicKey.toString();
  }, [publicKey]);
  return (0, import_react_query3.useQuery)({
    queryKey: ["ultra", "balances", address],
    queryFn: (_0) => __async(void 0, [_0], function* ({ signal }) {
      return yield ultraSwapService.getBalance(address, signal);
    }),
    enabled: !!address && getTerminalInView(),
    cacheTime: 2e4,
    staleTime: 2e4,
    refetchOnWindowFocus: false,
    refetchIntervalInBackground: false,
    refetchOnMount: false,
    select: (data) => combineSolAmounts(data)
  });
};

// src/hooks/useAsset.ts
var import_react_query4 = require("@tanstack/react-query");

// src/contexts/SearchService.ts
var BASE_URL = "https://datapi.jup.ag";
var SearchService = class {
  constructor() {
    this.ROUTE = {
      SEARCH: `${BASE_URL}/v1/assets`
    };
  }
  search(query) {
    return __async(this, null, function* () {
      const response = yield fetch(`${this.ROUTE.SEARCH}/search?query=${query}`, {
        credentials: "include"
      });
      const data = yield response.json();
      return data;
    });
  }
};
var searchService = new SearchService();

// src/hooks/useAsset.ts
var ASSET_QUERY_KEY = ["search", "assets"];
var useAsset = (mint) => {
  return (0, import_react_query4.useQuery)({
    queryKey: [...ASSET_QUERY_KEY, mint],
    queryFn: () => searchService.search(mint),
    enabled: !!mint,
    staleTime: Infinity,
    select: (data) => data[0]
  });
};

// src/contexts/SwapContext.tsx
var import_jsx_runtime7 = require("react/jsx-runtime");
var SwapContext = (0, import_react7.createContext)(null);
function useSwapContext() {
  const context = (0, import_react7.useContext)(SwapContext);
  if (!context) throw new Error("Missing SwapContextProvider");
  return context;
}
var INITIAL_FORM = {
  fromMint: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
  toMint: WRAPPED_SOL_MINT.toString(),
  fromValue: "",
  toValue: ""
};
var SwapContextProvider = (props) => {
  var _a2, _b;
  const { displayMode, scriptDomain: scriptDomain2, formProps: originalFormProps, children, enableWalletPassthrough } = props;
  const { screen: screen2 } = useScreenState();
  const { wallet } = useWalletPassThrough();
  const { refetch: refetchBalances } = useBalances();
  const isToPairFocused = (0, import_react7.useRef)(false);
  const walletPublicKey = (0, import_react7.useMemo)(() => {
    var _a3;
    return (_a3 = wallet == null ? void 0 : wallet.adapter.publicKey) == null ? void 0 : _a3.toString();
  }, [wallet == null ? void 0 : wallet.adapter.publicKey]);
  const formProps = (0, import_react7.useMemo)(() => __spreadValues(__spreadValues({}, INITIAL_FORM), originalFormProps), [originalFormProps]);
  const [form, setForm] = (0, import_react7.useState)({
    fromMint: (_a2 = formProps == null ? void 0 : formProps.initialInputMint) != null ? _a2 : "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
    toMint: (_b = formProps == null ? void 0 : formProps.initialOutputMint) != null ? _b : WRAPPED_SOL_MINT.toString(),
    fromValue: "",
    toValue: ""
  });
  const { data: fromTokenInfo } = useAsset(form.fromMint);
  const { data: toTokenInfo } = useAsset(form.toMint);
  (0, import_react7.useEffect)(() => {
    if (formProps.fixedMint) {
      if (formProps.fixedMint !== formProps.initialInputMint && formProps.fixedMint !== formProps.initialOutputMint) {
        console.error("fixedMint is not the same as the initial input or output mint");
      }
    }
  }, [formProps.fixedMint, formProps.initialInputMint, formProps.initialOutputMint]);
  const [errors, setErrors] = (0, import_react7.useState)({});
  const setupInitialAmount = (0, import_react7.useCallback)(() => {
    if (!(formProps == null ? void 0 : formProps.initialAmount) || !fromTokenInfo || !toTokenInfo) return;
    const toUiAmount = () => {
      if (!formProps.initialAmount) {
        return;
      }
      if (!fromTokenInfo) return;
      const value = new import_decimal2.default(formProps.initialAmount).div(Math.pow(10, fromTokenInfo.decimals)).toFixed();
      return value;
    };
    setTimeout(() => {
      setForm((prev) => {
        var _a3;
        return __spreadProps(__spreadValues({}, prev), { fromValue: (_a3 = toUiAmount()) != null ? _a3 : "" });
      });
    }, 0);
  }, [formProps.initialAmount, fromTokenInfo, toTokenInfo]);
  (0, import_react7.useEffect)(() => {
    setupInitialAmount();
  }, [formProps.initialAmount, setupInitialAmount]);
  const debouncedForm = useDebounce(form, 250);
  const amount = (0, import_react7.useMemo)(() => {
    if (!fromTokenInfo || !toTokenInfo) {
      return import_jsbi2.default.BigInt(0);
    }
    if (isToPairFocused.current === true) {
      if (!debouncedForm.toValue || !hasNumericValue(debouncedForm.toValue)) {
        return import_jsbi2.default.BigInt(0);
      }
      return import_jsbi2.default.BigInt(new import_decimal2.default(debouncedForm.toValue).mul(Math.pow(10, toTokenInfo.decimals)).floor().toFixed());
    } else {
      if (!debouncedForm.fromValue || !hasNumericValue(debouncedForm.fromValue)) {
        return import_jsbi2.default.BigInt(0);
      }
      return import_jsbi2.default.BigInt(
        new import_decimal2.default(debouncedForm.fromValue).mul(Math.pow(10, fromTokenInfo.decimals)).floor().toFixed()
      );
    }
  }, [debouncedForm.fromValue, debouncedForm.toValue, fromTokenInfo, toTokenInfo]);
  const [txStatus, setTxStatus] = (0, import_react7.useState)(void 0);
  const {
    data: ogQuoteResponseMeta,
    isFetching: loading,
    error: quoteError,
    refetch: refresh,
    errorUpdatedAt,
    dataUpdatedAt,
    isSuccess,
    isError
  } = useQuoteQuery(
    {
      inputMint: debouncedForm.fromMint,
      outputMint: debouncedForm.toMint,
      amount: amount.toString(),
      taker: walletPublicKey,
      swapMode: isToPairFocused.current ? "ExactOut" : "ExactIn",
      referralAccount: formProps.referralAccount,
      referralFee: formProps.referralFee
    },
    // Stop refetching when transaction is in progress
    !txStatus
  );
  (0, import_react7.useEffect)(() => {
    if (quoteError) {
      if (typeof quoteError === "string") {
        setErrors({
          fromValue: { title: quoteError, message: "" }
        });
        return;
      }
      setErrors({
        fromValue: { title: "Error fetching route. Try changing your input", message: "" }
      });
      return;
    }
    setErrors({});
  }, [quoteError]);
  const lastRefreshTimestamp = (0, import_react7.useMemo)(() => {
    if (loading) {
      return (/* @__PURE__ */ new Date()).getTime();
    }
    if (isError) {
      return new Date(errorUpdatedAt).getTime();
    }
    if (isSuccess) {
      return new Date(dataUpdatedAt).getTime();
    }
    return void 0;
  }, [loading, errorUpdatedAt, dataUpdatedAt, isError, isSuccess]);
  const [quoteResponseMeta, setQuoteResponseMeta] = (0, import_react7.useState)(null);
  (0, import_react7.useEffect)(() => {
    if (!ogQuoteResponseMeta) {
      setQuoteResponseMeta(null);
      return;
    }
    setQuoteResponseMeta(ogQuoteResponseMeta);
  }, [ogQuoteResponseMeta]);
  (0, import_react7.useEffect)(() => {
    if (!form.fromValue && !form.toValue && !quoteResponseMeta) {
      setForm((prev) => __spreadProps(__spreadValues({}, prev), { fromValue: "", toValue: "" }));
      return;
    }
    setForm((prev) => {
      const newValue = __spreadValues({}, prev);
      if (!fromTokenInfo || !toTokenInfo) return prev;
      const { outAmount, inAmount } = (quoteResponseMeta == null ? void 0 : quoteResponseMeta.quoteResponse) || {};
      if (!isToPairFocused.current) {
        newValue.toValue = outAmount ? new import_decimal2.default(outAmount.toString()).div(__pow(10, toTokenInfo.decimals)).toFixed() : "";
      } else {
        newValue.fromValue = inAmount ? new import_decimal2.default(inAmount.toString()).div(__pow(10, fromTokenInfo.decimals)).toFixed() : "";
      }
      return newValue;
    });
  }, [form.fromValue, form.toValue, fromTokenInfo, quoteResponseMeta, toTokenInfo]);
  const [lastSwapResult, setLastSwapResult] = (0, import_react7.useState)(null);
  const { mutateAsync: ultraSwapMutation } = useUltraSwapMutation();
  const onSubmit = (0, import_react7.useCallback)(() => __async(void 0, null, function* () {
    if (!walletPublicKey || !(wallet == null ? void 0 : wallet.adapter) || !quoteResponseMeta) {
      return null;
    }
    setTxStatus({
      txid: "",
      status: "loading"
    });
    try {
      if (!fromTokenInfo) throw new Error("Missing fromTokenInfo");
      if (!toTokenInfo) throw new Error("Missing toTokenInfo");
      yield ultraSwapMutation({
        quoteResponseMeta,
        fromTokenInfo,
        toTokenInfo,
        setTxStatus,
        setLastSwapResult
      });
    } catch (error) {
      console.log("Swap error", error);
    }
  }), [walletPublicKey, wallet == null ? void 0 : wallet.adapter, quoteResponseMeta, ultraSwapMutation, fromTokenInfo, toTokenInfo]);
  const reset = (0, import_react7.useCallback)(
    ({ resetValues } = { resetValues: false }) => {
      if (resetValues) {
        setForm(INITIAL_FORM);
        setupInitialAmount();
      } else {
        setForm((prev) => __spreadProps(__spreadValues({}, prev), { toValue: "" }));
      }
      setQuoteResponseMeta(null);
      setErrors({});
      setLastSwapResult(null);
      setTxStatus(void 0);
      refetchBalances();
    },
    [refetchBalances, setupInitialAmount]
  );
  (0, import_react7.useEffect)(() => {
    if (typeof window.Jupiter.onFormUpdate === "function") {
      window.Jupiter.onFormUpdate(form);
    }
  }, [form]);
  (0, import_react7.useEffect)(() => {
    if (typeof window.Jupiter.onScreenUpdate === "function") {
      window.Jupiter.onScreenUpdate(screen2);
    }
  }, [screen2]);
  return /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
    SwapContext.Provider,
    {
      value: {
        form,
        setForm,
        errors,
        setErrors,
        fromTokenInfo,
        toTokenInfo,
        quoteResponseMeta,
        setQuoteResponseMeta,
        onSubmit,
        lastSwapResult,
        reset,
        refresh,
        loading,
        quoteError,
        lastRefreshTimestamp,
        isToPairFocused,
        displayMode,
        formProps,
        scriptDomain: scriptDomain2,
        swapping: {
          txStatus
        },
        enableWalletPassthrough
      },
      children
    }
  );
};

// src/components/Header.tsx
var import_react11 = require("react");

// src/icons/RefreshSVG.tsx
var import_jsx_runtime8 = require("react/jsx-runtime");
var RefreshSVG = ({ width = "12", height = "12" }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsxs)("svg", { width, height, viewBox: "0 0 12 12", fill: "inherit", xmlns: "http://www.w3.org/2000/svg", children: [
    /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("g", { clipPath: "url(#clip0_841_4053)", children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
      "path",
      {
        d: "M11.6466 4.23513V0.706082L10.4111 1.94156C9.3173 0.741165 7.72912 0 6 0C2.6827 0 0 2.6827 0 6C0 9.3173 2.68203 12 6 12C7.69405 12 9.21142 11.2939 10.3059 10.165L9.31797 9.14128C8.50601 10.0234 7.30561 10.5879 6 10.5879C3.45892 10.5879 1.41216 8.5411 1.41216 6.00002C1.41216 3.45894 3.45892 1.41218 6 1.41218C7.34135 1.41218 8.57615 2.01238 9.42317 2.92954L8.11757 4.23515L11.6466 4.23513Z",
        fill: "inherit"
      }
    ) }),
    /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("defs", { children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("clipPath", { id: "clip0_841_4053", children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("rect", { width, height, fill: "white" }) }) })
  ] });
};
var RefreshSVG_default = RefreshSVG;

// src/components/WalletComponents/components/WalletModalButton.tsx
var import_react8 = require("react");
var import_jsx_runtime9 = require("react/jsx-runtime");
var WalletModalButton = () => {
  const { connecting } = useWalletPassThrough();
  const { setScreen } = useScreenState();
  const handleClick = (0, import_react8.useCallback)((event) => {
    if (window.Jupiter.enableWalletPassthrough && window.Jupiter.onRequestConnectWallet) {
      window.Jupiter.onRequestConnectWallet();
    } else {
      setScreen("Wallet");
    }
  }, [setScreen]);
  return /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(
    "button",
    {
      type: "button",
      className: "py-2 px-3 h-7 flex items-center rounded-2xl text-xs bg-interactive text-primary-text",
      onClick: handleClick,
      children: connecting ? /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { children: /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { children: "Connecting..." }) }) : /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { children: /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { children: "Connect Wallet" }) })
    }
  );
};

// src/components/WalletComponents/WalletButton.tsx
var import_react9 = require("react");

// src/components/CurrentUserBadge.tsx
var import_jsx_runtime10 = require("react/jsx-runtime");
var CurrentUserBadge = () => {
  var _a2;
  const { publicKey, wallet } = useWalletPassThrough();
  if (!wallet || !publicKey) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime10.jsxs)("div", { className: "flex items-center bg-interactive py-2 px-3 rounded-2xl h-7", children: [
    /* @__PURE__ */ (0, import_jsx_runtime10.jsx)(
      "div",
      {
        className: "w-4 h-4 rounded-full bg-interactive dark:bg-white-10 flex justify-center items-center",
        style: { position: "relative" },
        children: /* @__PURE__ */ (0, import_jsx_runtime10.jsx)("img", { alt: "Wallet logo", width: 16, height: 16, src: (_a2 = wallet == null ? void 0 : wallet.adapter) == null ? void 0 : _a2.icon })
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime10.jsx)("div", { className: "ml-2", children: /* @__PURE__ */ (0, import_jsx_runtime10.jsx)("div", { className: "text-xs text-primary-text", children: shortenAddress(`${publicKey}`, 2) }) })
  ] });
};

// src/components/WalletComponents/WalletButton.tsx
var import_jsx_runtime11 = require("react/jsx-runtime");
var WalletButton = () => {
  const { publicKey, connected, connecting, disconnect } = useWalletPassThrough();
  const [active, setActive] = (0, import_react9.useState)(false);
  const ref = (0, import_react9.useRef)(null);
  const { screen: screen2, setScreen } = useScreenState();
  const base58 = (0, import_react9.useMemo)(() => publicKey == null ? void 0 : publicKey.toBase58(), [publicKey]);
  const onClickDisconnect = () => {
    setActive(false);
    disconnect();
  };
  const closePopup = () => {
    setActive(false);
  };
  useOutsideClick(ref, closePopup);
  if (!connected && !connecting || !base58) {
    return /* @__PURE__ */ (0, import_jsx_runtime11.jsx)("div", { onClick: () => setScreen("Wallet"), className: "text-primary-text", children: /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(WalletModalButton, {}) });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime11.jsxs)("div", { className: "cursor-pointer relative", children: [
    /* @__PURE__ */ (0, import_jsx_runtime11.jsx)("div", { onClick: () => setActive(!active), children: /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(CurrentUserBadge, {}) }),
    screen2 === "Initial" ? /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(
      "ul",
      {
        "aria-label": "dropdown-list",
        className: active ? "absolute block top-10 right-0 text-sm bg-black rounded-lg p-2 text-primary-text dark:bg-white dark:text-black" : "hidden",
        ref,
        role: "menu",
        children: /* @__PURE__ */ (0, import_jsx_runtime11.jsx)("li", { onClick: onClickDisconnect, role: "menuitem", children: /* @__PURE__ */ (0, import_jsx_runtime11.jsx)("span", { children: "Disconnect" }) })
      }
    ) : null
  ] });
};

// src/contexts/BrandingProvider.tsx
var import_react10 = require("react");
var import_jsx_runtime12 = require("react/jsx-runtime");
var BrandingContext = (0, import_react10.createContext)({});
var BrandingProvider = (props) => {
  const { branding, children } = props;
  const logoUri = (0, import_react10.useMemo)(() => (branding == null ? void 0 : branding.logoUri) ? branding.logoUri : "https://jup.ag/svg/jupiter-logo.svg", [branding == null ? void 0 : branding.logoUri]);
  const name = (0, import_react10.useMemo)(() => (branding == null ? void 0 : branding.name) ? branding.name : "Jupiter", [branding == null ? void 0 : branding.name]);
  return /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(BrandingContext.Provider, { value: { logoUri, name }, children });
};
function useBranding() {
  const context = (0, import_react10.useContext)(BrandingContext);
  return context;
}

// src/components/Header.tsx
var import_jsx_runtime13 = require("react/jsx-runtime");
var Header = () => {
  const { refresh, enableWalletPassthrough } = useSwapContext();
  const { refetch: refetchBalances } = useBalances();
  const { logoUri, name } = useBranding();
  const onRefresh = (0, import_react11.useCallback)(() => {
    refetchBalances();
    refresh();
  }, [refetchBalances, refresh]);
  return /* @__PURE__ */ (0, import_jsx_runtime13.jsx)("div", { className: "mt-2 h-7 pl-3 pr-2", children: /* @__PURE__ */ (0, import_jsx_runtime13.jsxs)("div", { className: "w-full flex items-center justify-between ", children: [
    /* @__PURE__ */ (0, import_jsx_runtime13.jsxs)("div", { className: "flex items-center space-x-2", children: [
      /* @__PURE__ */ (0, import_jsx_runtime13.jsx)("img", { src: logoUri, alt: "Terminal Branding", className: "w-6 h-6" }),
      /* @__PURE__ */ (0, import_jsx_runtime13.jsx)("span", { className: "font-bold text-sm text-primary-text", children: name })
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime13.jsxs)("div", { className: "flex space-x-1 items-center", children: [
      /* @__PURE__ */ (0, import_jsx_runtime13.jsx)(
        "button",
        {
          type: "button",
          className: "p-2 h-7 w-7 flex items-center justify-center  rounded-full  bg-interactive text-primary-text fill-current",
          onClick: onRefresh,
          children: /* @__PURE__ */ (0, import_jsx_runtime13.jsx)(RefreshSVG_default, {})
        }
      ),
      !enableWalletPassthrough && /* @__PURE__ */ (0, import_jsx_runtime13.jsx)(WalletButton, {})
    ] })
  ] }) });
};
var Header_default = Header;

// src/components/screens/InitialScreen.tsx
var import_react24 = require("react");

// src/components/Form.tsx
var import_react19 = require("react");
var import_react_number_format = require("react-number-format");

// src/misc/constants.ts
var import_web35 = require("@solana/web3.js");
var MINIMUM_SOL_BALANCE = 0.01;
var ROUTE_CACHE_DURATION = 2e4;
var MAX_INPUT_LIMIT = 1e14;
var WRAPPED_SOL_MINT2 = new import_web35.PublicKey("So11111111111111111111111111111111111111112");

// src/components/Coinbalance.tsx
var React6 = __toESM(require("react"));
var import_jsx_runtime14 = require("react/jsx-runtime");
var CoinBalance = (props) => {
  const { data: balances } = useBalances();
  const { connected } = useWalletPassThrough();
  const formattedBalance = React6.useMemo(() => {
    if (!balances) {
      return "0";
    }
    const accBalanceObj = balances[props.mintAddress];
    if (!accBalanceObj) return "0";
    const balance = accBalanceObj.uiAmount.toString();
    return balance;
  }, [balances, props.mintAddress]);
  if (props.hideZeroBalance && (formattedBalance === "0" || !formattedBalance)) return null;
  if (!connected) return null;
  return /* @__PURE__ */ (0, import_jsx_runtime14.jsx)("span", { translate: "no", children: formattedBalance });
};
var Coinbalance_default = CoinBalance;

// src/components/JupButton.tsx
var import_react12 = __toESM(require("react"));
var import_jsx_runtime15 = require("react/jsx-runtime");
var JupButton = import_react12.default.forwardRef(
  ({ onClick, disabled, children, className = "", size = "md" }, ref) => {
    const contentClass = (() => {
      if (size === "sm") {
        return "px-4 py-2.5 text-xs";
      }
      if (size === "md") {
        return "px-4 py-3 text-sm font-semibold";
      }
      if (size === "lg") {
        return "p-5 text-md font-semibold";
      }
    })();
    return /* @__PURE__ */ (0, import_jsx_runtime15.jsx)(
      "button",
      {
        type: "button",
        ref,
        className: cn("rounded-xl", {
          relative: true,
          "opacity-50 cursor-not-allowed": disabled,
          [className]: true
        }),
        disabled,
        onClick,
        children: /* @__PURE__ */ (0, import_jsx_runtime15.jsx)("div", { className: `${contentClass} h-full w-full leading-none`, children })
      }
    );
  }
);
JupButton.displayName = "JupButton";
var JupButton_default = JupButton;

// src/components/TokenIcon.tsx
var React8 = __toESM(require("react"));

// src/icons/WarningIcon.tsx
var import_jsx_runtime16 = require("react/jsx-runtime");
var WarningIcon = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime16.jsxs)("svg", __spreadProps(__spreadValues({ width: "12", height: "12", viewBox: "0 0 20 20", fill: "none", xmlns: "http://www.w3.org/2000/svg" }, props), { children: [
    /* @__PURE__ */ (0, import_jsx_runtime16.jsx)("rect", { x: "8", y: "6", width: "4", height: "10", fill: "white" }),
    /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(
      "path",
      {
        fillRule: "evenodd",
        clipRule: "evenodd",
        d: "M17.8634 15.6962C17.5539 15.1565 11.3038 4.37041 10.7692 3.44957C10.4179 2.8452 9.58075 2.85511 9.22912 3.44957C8.83702 4.11227 2.55515 14.9396 2.11839 15.7235C1.7984 16.2978 2.15189 17.0509 2.88009 17.0509H17.0974C17.7586 17.0509 18.2502 16.3695 17.8635 15.696L17.8634 15.6962ZM10.0005 15.6277C9.50937 15.6277 9.11108 15.2297 9.11108 14.7383C9.11108 14.247 9.50937 13.8489 10.0005 13.8489C10.4918 13.8489 10.8899 14.247 10.8899 14.7383C10.8899 15.2297 10.4918 15.6277 10.0005 15.6277ZM10.5341 12.7817C10.5341 13.1374 10.3562 13.3154 10.0005 13.3154C9.64474 13.3154 9.46681 13.1375 9.46681 12.7817L8.93314 7.62324C8.93314 7.08957 9.28887 6.5559 10.0005 6.5559C10.7119 6.5559 11.0678 7.08957 11.0678 7.62324L10.5341 12.7817Z",
        fill: "currentColor"
      }
    )
  ] }));
};
var WarningIcon_default = WarningIcon;

// src/components/TokenIcon.tsx
var import_jsx_runtime17 = require("react/jsx-runtime");
function genImageTransformURL(url) {
  const qs = new URLSearchParams({
    w: "48",
    h: "48",
    url: (() => {
      if (url.endsWith("/")) return url.slice(0, -1);
      return url;
    })()
  });
  return `https://wsrv.nl/?${qs.toString()}`;
}
var UnknownTokenImage = ({ width, height, imageUrl }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime17.jsxs)(
    "svg",
    {
      width,
      height,
      viewBox: "0 0 24 24",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      "data-url": imageUrl,
      "data-transformed-url": genImageTransformURL((imageUrl == null ? void 0 : imageUrl.toString()) || ""),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(
          "path",
          {
            d: "M12 0C18.6271 0 24 5.37288 24 12C24 18.6271 18.6269 24 12 24C5.37312 24 0 18.6286 0 12C0 5.37144 5.37216 0 12 0Z",
            fill: "#23C1AA"
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(
          "path",
          {
            d: "M10.79 14.55H12.89V14.355C12.89 13.925 13.01 13.55 13.25 13.23C13.49 12.91 13.765 12.605 14.075 12.315C14.315 12.085 14.545 11.85 14.765 11.61C14.985 11.36 15.165 11.09 15.305 10.8C15.455 10.5 15.53 10.16 15.53 9.78C15.53 9.25 15.395 8.75 15.125 8.28C14.855 7.8 14.45 7.41 13.91 7.11C13.38 6.8 12.725 6.645 11.945 6.645C11.305 6.645 10.725 6.765 10.205 7.005C9.69504 7.245 9.27504 7.575 8.94504 7.995C8.62504 8.415 8.42004 8.905 8.33004 9.465L10.415 9.99C10.475 9.61 10.64 9.31 10.91 9.09C11.19 8.86 11.515 8.745 11.885 8.745C12.315 8.745 12.64 8.85 12.86 9.06C13.09 9.26 13.205 9.52 13.205 9.84C13.205 10.15 13.09 10.425 12.86 10.665C12.63 10.895 12.37 11.155 12.08 11.445C11.77 11.765 11.475 12.14 11.195 12.57C10.925 13 10.79 13.545 10.79 14.205V14.55ZM10.73 18H12.98V15.75H10.73V18Z",
            fill: "white"
          }
        )
      ]
    }
  );
};
var TokenIcon = ({
  info,
  width = 32,
  height = 32,
  enableUnknownTokenWarning = true
}) => {
  const [transformerHasError, setTransformerHasError] = React8.useState(false);
  const [hasError, setHasError] = React8.useState(false);
  React8.useEffect(() => {
    setHasError(false);
  }, [info]);
  const imageUrl = React8.useMemo(() => {
    var _a2;
    try {
      if (!(info == null ? void 0 : info.icon)) return void 0;
      if (typeof window !== "undefined" && ((_a2 = info == null ? void 0 : info.icon) == null ? void 0 : _a2.startsWith("/")))
        return new URL(window.location.origin + info.icon);
      return (info == null ? void 0 : info.icon) ? new URL(info.icon) : void 0;
    } catch (error) {
      return void 0;
    }
  }, [info == null ? void 0 : info.icon]);
  const isUnknown = React8.useMemo(() => {
    if (!enableUnknownTokenWarning) return false;
    if (info) {
      return checkIsUnknownToken(info);
    }
  }, [enableUnknownTokenWarning, info]);
  const ImageComp = React8.useMemo(() => {
    if (!imageUrl || !info || hasError && transformerHasError) {
      return /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(UnknownTokenImage, { width, height, imageUrl: (imageUrl == null ? void 0 : imageUrl.toString()) || "" });
    }
    if (transformerHasError) {
      return (
        // eslint-disable-next-line
        /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(
          "img",
          {
            src: info.icon,
            alt: info.symbol,
            width,
            height,
            style: { maxWidth: width, maxHeight: height },
            className: `object-cover rounded-full`,
            onError: () => {
              setHasError(true);
            }
          }
        )
      );
    }
    return (
      // eslint-disable-next-line
      /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(
        "img",
        {
          src: genImageTransformURL(info.icon || ""),
          alt: info.symbol,
          width,
          height,
          style: { maxWidth: width, maxHeight: height },
          className: `object-cover rounded-full`,
          onError: () => {
            setTransformerHasError(true);
          }
        }
      )
    );
  }, [hasError, height, imageUrl, info, transformerHasError, width]);
  return /* @__PURE__ */ (0, import_jsx_runtime17.jsxs)("span", { className: "relative", children: [
    ImageComp,
    isUnknown && /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(
      WarningIcon_default,
      {
        width: Math.max(width * 0.6, 16),
        height: Math.max(height * 0.6, 16),
        className: "absolute -p-1 text-warning -bottom-[2px] -right-[5px]"
      }
    )
  ] });
};
var TokenIcon_default = TokenIcon;

// src/icons/WalletIcon.tsx
var import_jsx_runtime18 = require("react/jsx-runtime");
var WalletIcon = ({ width = 20, height = 20 }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime18.jsx)("svg", { width, height, viewBox: "0 0 11 11", fill: "none", xmlns: "http://www.w3.org/2000/svg", children: /* @__PURE__ */ (0, import_jsx_runtime18.jsx)(
    "path",
    {
      d: "M2.0625 11H10.3125V2.75H8.59375V0H2.0625C0.923158 0 0 0.923158 0 2.0625V8.9375C0 10.0768 0.923158 11 2.0625 11ZM8.9375 4.125V9.625H2.0625C1.6825 9.625 1.375 9.3175 1.375 8.9375V4.00486C1.59589 4.08408 1.82818 4.12437 2.0625 4.12504L8.9375 4.125ZM2.0625 1.375H7.21875V2.75H2.0625C1.6825 2.75 1.375 2.4425 1.375 2.0625C1.375 1.6825 1.6825 1.375 2.0625 1.375Z",
      fill: "white",
      fillOpacity: "0.25"
    }
  ) });
};
var WalletIcon_default = WalletIcon;

// src/components/CoinBalanceUSD.tsx
var import_decimal3 = __toESM(require("decimal.js"));
var import_react13 = require("react");
var import_jsx_runtime19 = require("react/jsx-runtime");
var CoinBalanceUSD = (props) => {
  const { tokenInfo, amount, maxDecimals, prefix = "" } = props;
  const tokenPrice = tokenInfo.usdPrice || 0;
  const amountInUSD = (0, import_react13.useMemo)(() => {
    if (!amount || !hasNumericValue(amount)) return new import_decimal3.default(0);
    return new import_decimal3.default(amount).mul(tokenPrice);
  }, [amount, tokenPrice]);
  return /* @__PURE__ */ (0, import_jsx_runtime19.jsxs)(import_jsx_runtime19.Fragment, { children: [
    prefix,
    "$",
    formatNumber.format(amountInUSD, maxDecimals || 2)
  ] });
};

// src/components/PriceInfo/index.tsx
var import_decimal6 = __toESM(require("decimal.js"));
var import_jsbi3 = __toESM(require("jsbi"));
var import_react14 = require("react");

// src/components/ExchangeRate.tsx
var import_decimal5 = __toESM(require("decimal.js"));
var React9 = __toESM(require("react"));

// src/components/PrecisionTickSize/index.tsx
var import_decimal4 = __toESM(require("decimal.js"));
var import_jsx_runtime20 = require("react/jsx-runtime");
function generateSubscriptNumbers(x) {
  const subscriptNumbers = ["\u2080", "\u2081", "\u2082", "\u2083", "\u2084", "\u2085", "\u2086", "\u2087", "\u2088", "\u2089"];
  const xString = x.toString();
  let result = "";
  for (let i = 0; i < xString.length; i++) {
    const digit = parseInt(xString.charAt(i), 10);
    const subscriptNumber = subscriptNumbers[digit];
    result += subscriptNumber;
  }
  return result;
}
var usePrecisionTick = (value) => {
  const firstSD = import_decimal4.default.abs(import_decimal4.default.ceil(new import_decimal4.default(-1).mul(import_decimal4.default.log10(value)))).toNumber();
  const [prefix, suffix] = [
    new import_decimal4.default(value).toFixed().slice(0, firstSD + 2),
    // +2 to account for 0.
    new import_decimal4.default(value).toFixed().slice(firstSD + 1)
    // +1 to account for 0. - and slice index
  ];
  return [firstSD, prefix, suffix];
};
var PrecisionTickSize = ({ value, maxSuffix }) => {
  const [firstSD, _, suffix] = usePrecisionTick(value);
  if (firstSD <= 5) {
    return /* @__PURE__ */ (0, import_jsx_runtime20.jsx)("span", { children: value.toFixed(6) });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime20.jsxs)("span", { className: "flex items-center h-4", children: [
    "0.0",
    /* @__PURE__ */ (0, import_jsx_runtime20.jsx)("span", { className: "mb-3 text-xl mx-0.5", children: generateSubscriptNumbers(firstSD - 1) }),
    suffix.slice(0, maxSuffix)
  ] });
};
var PrecisionTickSize_default = PrecisionTickSize;

// src/components/ExchangeRate.tsx
var import_jsx_runtime21 = require("react/jsx-runtime");
var calculateRate = ({ inAmount, inputDecimal, outAmount, outputDecimal }, reverse) => {
  const input = fromLamports(inAmount, inputDecimal);
  const output = fromLamports(outAmount, outputDecimal);
  const rate = !reverse ? new import_decimal5.default(input).div(output) : new import_decimal5.default(output).div(input);
  if (Number.isNaN(rate.toNumber())) {
    return new import_decimal5.default(0);
  }
  return rate;
};
var ApproxSVG = ({ width = 16, height = 16 }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("svg", { width, height, viewBox: "0 0 16 16", fill: "none", xmlns: "http://www.w3.org/2000/svg", children: /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(
    "path",
    {
      d: "M10.8573 8.18429L13.6323 5.95933L10.8573 3.73438V5.31937H3.32735V6.59937H10.8573V8.18429ZM5.14223 7.81429L2.36719 10.0393L5.14223 12.2642V10.6792H12.6722V9.39922H5.14223V7.81429Z",
      fill: "currentColor"
    }
  ) });
};
var ExchangeRate = ({
  className,
  textClassName,
  loading = false,
  fromTokenInfo,
  rateParams,
  toTokenInfo,
  reversible = true
}) => {
  const [reverse, setReverse] = React9.useState(reversible != null ? reversible : true);
  const rate = React9.useMemo(() => calculateRate(rateParams, reverse), [reverse, rateParams]);
  const onReverse = React9.useCallback((event) => {
    event.stopPropagation();
    setReverse((prevState) => !prevState);
  }, []);
  return /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)(
    "div",
    {
      className: cn(className, "flex cursor-pointer text-primary-text text-xs align-center"),
      onClick: onReverse,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("span", { className: cn(textClassName, "max-w-full flex whitespace-nowrap"), children: reverse ? /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)(import_jsx_runtime21.Fragment, { children: [
          "1 ",
          fromTokenInfo.symbol,
          " \u2248",
          /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("div", { className: "flex ml-0.5", children: rate.gt(1e-5) ? `${formatNumber.format(rate, 6)} ${toTokenInfo.symbol}` : /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)(import_jsx_runtime21.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(PrecisionTickSize_default, { value: rate.toNumber(), maxSuffix: 6 }),
            " ",
            toTokenInfo.symbol
          ] }) })
        ] }) : /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)(import_jsx_runtime21.Fragment, { children: [
          "1 ",
          toTokenInfo.symbol,
          " \u2248",
          /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("div", { className: "flex ml-0.5", children: rate.gt(1e-5) ? `${formatNumber.format(rate, 6)} ${fromTokenInfo.symbol}` : /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)(import_jsx_runtime21.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(PrecisionTickSize_default, { value: rate.toNumber(), maxSuffix: 6 }),
            " ",
            fromTokenInfo.symbol
          ] }) })
        ] }) }),
        reversible ? /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("div", { className: "ml-2", children: /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(ApproxSVG, {}) }) : null
      ]
    }
  );
};
var ExchangeRate_default = ExchangeRate;

// src/icons/UltraIcon.tsx
var import_jsx_runtime22 = require("react/jsx-runtime");
var UltraIcon = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime22.jsxs)("svg", __spreadProps(__spreadValues({ viewBox: "0 0 10 10", fill: "none", xmlns: "http://www.w3.org/2000/svg", height: "1em", width: "1em" }, props), { children: [
    /* @__PURE__ */ (0, import_jsx_runtime22.jsxs)("g", { clipPath: "url(#clip0_1785_33901)", children: [
      /* @__PURE__ */ (0, import_jsx_runtime22.jsx)(
        "path",
        {
          d: "M7.20001 7.86127C7.45429 7.86127 7.69815 7.96228 7.87795 8.14208C8.05775 8.32188 8.15876 8.56574 8.15876 8.82002C8.15876 8.56574 8.25977 8.32188 8.43957 8.14208C8.61937 7.96228 8.86324 7.86127 9.11751 7.86127C8.86324 7.86127 8.61937 7.76026 8.43957 7.58046C8.25977 7.40066 8.15876 7.15679 8.15876 6.90252C8.15876 7.15679 8.05775 7.40066 7.87795 7.58046C7.69815 7.76026 7.45429 7.86127 7.20001 7.86127ZM7.20001 2.10877C7.45429 2.10877 7.69815 2.20978 7.87795 2.38958C8.05775 2.56938 8.15876 2.81324 8.15876 3.06752C8.15876 2.81324 8.25977 2.56938 8.43957 2.38958C8.61937 2.20978 8.86324 2.10877 9.11751 2.10877C8.86324 2.10877 8.61937 2.00776 8.43957 1.82796C8.25977 1.64816 8.15876 1.40429 8.15876 1.15002C8.15876 1.40429 8.05775 1.64816 7.87795 1.82796C7.69815 2.00776 7.45429 2.10877 7.20001 2.10877Z",
          stroke: "currentColor",
          strokeWidth: "0.95875",
          strokeLinecap: "round",
          strokeLinejoin: "round"
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime22.jsx)(
        "path",
        {
          d: "M4.09971 1.42435C4.08095 1.37314 4.04691 1.32892 4.00219 1.29768C3.95748 1.26644 3.90425 1.24969 3.84971 1.24969C3.79517 1.24969 3.74194 1.26644 3.69723 1.29768C3.65251 1.32892 3.61847 1.37314 3.59971 1.42435L3.34534 2.11935C3.1457 2.6643 2.82969 3.15919 2.41931 3.56957C2.00893 3.97996 1.51403 4.29596 0.969084 4.4956L0.274086 4.74998C0.222871 4.76874 0.178655 4.80279 0.147417 4.8475C0.116179 4.89221 0.0994263 4.94543 0.0994263 4.99998C0.0994263 5.05452 0.116179 5.10775 0.147417 5.15246C0.178655 5.19717 0.222871 5.23122 0.274086 5.24998L0.969084 5.50435C1.51403 5.70399 2.00893 6.02001 2.41931 6.43039C2.82969 6.84077 3.1457 7.33566 3.34534 7.88061L3.59971 8.5756C3.61847 8.62681 3.65251 8.67103 3.69723 8.70227C3.74194 8.7335 3.79517 8.75026 3.84971 8.75026C3.90425 8.75026 3.95748 8.7335 4.00219 8.70227C4.04691 8.67103 4.08095 8.62681 4.09971 8.5756L4.35409 7.88061C4.55372 7.33566 4.86973 6.84077 5.28011 6.43039C5.69049 6.02001 6.18538 5.70399 6.73033 5.50435L7.42533 5.24998C7.47654 5.23122 7.52076 5.19717 7.552 5.15246C7.58323 5.10775 7.59999 5.05452 7.59999 4.99998C7.59999 4.94543 7.58323 4.89221 7.552 4.8475C7.52076 4.80279 7.47654 4.76874 7.42533 4.74998L6.73033 4.4956C6.18538 4.29596 5.8125 3.97995 5.40212 3.56957C4.99174 3.15919 4.55372 2.6643 4.35409 2.11935L4.09971 1.42435Z",
          fill: "currentColor"
        }
      )
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime22.jsx)("defs", { children: /* @__PURE__ */ (0, import_jsx_runtime22.jsx)("clipPath", { id: "clip0_1785_33901", children: /* @__PURE__ */ (0, import_jsx_runtime22.jsx)("rect", { width: "10", height: "10", fill: "white" }) }) })
  ] }));
};

// src/components/PriceInfo/TransactionFee.tsx
var import_jsx_runtime23 = require("react/jsx-runtime");
var TransactionFee = ({ gasFee, gasless }) => {
  if (!gasFee) return null;
  return /* @__PURE__ */ (0, import_jsx_runtime23.jsxs)("div", { className: "flex items-center justify-between text-xs", children: [
    /* @__PURE__ */ (0, import_jsx_runtime23.jsx)("div", { className: "flex w-[50%] text-primary-text/50", children: /* @__PURE__ */ (0, import_jsx_runtime23.jsx)("span", { children: "Network Fee" }) }),
    /* @__PURE__ */ (0, import_jsx_runtime23.jsxs)("div", { className: "flex items-center gap-1", children: [
      gasless && /* @__PURE__ */ (0, import_jsx_runtime23.jsxs)(import_jsx_runtime23.Fragment, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime23.jsx)(UltraIcon, { color: "#FDB022", width: 12, height: 12 }),
        /* @__PURE__ */ (0, import_jsx_runtime23.jsx)("div", { className: "text-xs text-primary-text", children: "Gasless" })
      ] }),
      /* @__PURE__ */ (0, import_jsx_runtime23.jsxs)(
        "div",
        {
          className: cn("text-primary-text", {
            "line-through": gasless
          }),
          children: [
            gasFee,
            " SOL"
          ]
        }
      )
    ] })
  ] });
};
var TransactionFee_default = TransactionFee;

// src/queries/useUltraRouter.ts
var import_react_query5 = require("@tanstack/react-query");
var useUltraRouters = (options) => {
  return (0, import_react_query5.useQuery)(__spreadValues({
    queryKey: ["routers"],
    queryFn: () => ultraSwapService.getRouters(),
    cacheTime: Infinity,
    staleTime: Infinity
  }, options));
};

// src/components/PriceInfo/index.tsx
var import_jsx_runtime24 = require("react/jsx-runtime");
var Index = ({
  quoteResponse,
  fromTokenInfo,
  toTokenInfo,
  loading,
  containerClassName
}) => {
  const rateParams = {
    inAmount: (quoteResponse == null ? void 0 : quoteResponse.quoteResponse.inAmount) || import_jsbi3.default.BigInt(0),
    // If there's no selectedRoute, we will use first route value to temporarily calculate
    inputDecimal: fromTokenInfo.decimals,
    outAmount: (quoteResponse == null ? void 0 : quoteResponse.quoteResponse.outAmount) || import_jsbi3.default.BigInt(0),
    // If there's no selectedRoute, we will use first route value to temporarily calculate
    outputDecimal: toTokenInfo.decimals
  };
  const { data: routerInfo } = useUltraRouters({
    select: (data) => {
      if (!quoteResponse) {
        return null;
      }
      return data.find((router2) => router2.id === quoteResponse.quoteResponse.router);
    }
  });
  const priceImpact = formatNumber.format(
    new import_decimal6.default((quoteResponse == null ? void 0 : quoteResponse.quoteResponse.priceImpactPct) || 0).mul(100).toDP(2)
  );
  const priceImpactText = Number(priceImpact) < 0.01 ? void 0 : `-${priceImpact}%`;
  const fee = (0, import_react14.useMemo)(() => {
    if (!quoteResponse) {
      return 0;
    }
    return quoteResponse.quoteResponse.feeBps / 100;
  }, [quoteResponse]);
  const router = (0, import_react14.useMemo)(() => {
    if (!quoteResponse) {
      return;
    }
    return quoteResponse.quoteResponse.router;
  }, [quoteResponse]);
  const gasFee = (0, import_react14.useMemo)(() => {
    if (quoteResponse) {
      const { prioritizationFeeLamports } = quoteResponse.quoteResponse;
      if (prioritizationFeeLamports) {
        return prioritizationFeeLamports / 1e9;
      }
    }
    return 0;
  }, [quoteResponse]);
  return /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)("div", { className: cn("mt-4 space-y-4 border border-white/5 rounded-xl", containerClassName), children: [
    /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)("div", { className: "flex items-center justify-between text-xs", children: [
      /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("div", { className: "text-primary-text/50", children: /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("span", { children: "Rate" }) }),
      import_jsbi3.default.greaterThan(rateParams.inAmount, import_jsbi3.default.BigInt(0)) && import_jsbi3.default.greaterThan(rateParams.outAmount, import_jsbi3.default.BigInt(0)) ? /* @__PURE__ */ (0, import_jsx_runtime24.jsx)(
        ExchangeRate_default,
        {
          loading,
          rateParams,
          fromTokenInfo,
          toTokenInfo,
          reversible: true
        }
      ) : /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("span", { className: "text-primary-text/50", children: "-" })
    ] }),
    priceImpactText && /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)("div", { className: "flex items-center justify-between text-xs text-primary-text/50", children: [
      /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("span", { children: "Price Impact" }) }),
      /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("div", { className: "text-primary-text", children: priceImpactText })
    ] }),
    router && /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)("div", { className: "flex items-center justify-between text-xs", children: [
      /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("div", { className: "text-primary-text/50", children: /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("span", { children: "Router" }) }),
      /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("div", { className: "flex items-center gap-1", children: routerInfo && /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)(import_jsx_runtime24.Fragment, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("img", { src: routerInfo.icon, alt: quoteResponse.quoteResponse.router, width: 10, height: 10 }),
        /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("div", { className: "text-primary-text", children: routerInfo.name })
      ] }) })
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)("div", { className: "flex items-center justify-between text-xs", children: [
      /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("div", { className: "text-primary-text/50", children: /* @__PURE__ */ (0, import_jsx_runtime24.jsx)("span", { children: "Fee" }) }),
      /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)("div", { className: "text-primary-text", children: [
        fee,
        "%"
      ] })
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime24.jsx)(TransactionFee_default, { gasFee, gasless: quoteResponse == null ? void 0 : quoteResponse.quoteResponse.gasless })
  ] });
};
var PriceInfo_default = Index;

// src/components/SwitchPairButton.tsx
var import_jsx_runtime25 = require("react/jsx-runtime");
var IconSwitchPairDark = () => /* @__PURE__ */ (0, import_jsx_runtime25.jsxs)("svg", { width: 20, height: 20, viewBox: "0 0 21 22", fill: "white", fillOpacity: 0.5, xmlns: "http://www.w3.org/2000/svg", children: [
  /* @__PURE__ */ (0, import_jsx_runtime25.jsx)(
    "path",
    {
      d: "M6.51043 7.47998V14.99H7.77043V7.47998L9.66043 9.36998L10.5505 8.47994L7.5859 5.51453C7.3398 5.26925 6.94114 5.26925 6.69504 5.51453L3.73047 8.47994L4.62051 9.36998L6.51043 7.47998Z",
      fill: "white"
    }
  ),
  /* @__PURE__ */ (0, import_jsx_runtime25.jsx)(
    "path",
    {
      d: "M14.4902 14.52V7.01001H13.2302V14.52L11.3402 12.63L10.4502 13.5201L13.4148 16.4855C13.6609 16.7308 14.0595 16.7308 14.3056 16.4855L17.2702 13.5201L16.3802 12.63L14.4902 14.52Z",
      fill: "white"
    }
  )
] });
var SwitchPairButton = ({
  className,
  onClick,
  disabled
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime25.jsx)("div", { className: "flex justify-center bg-background rounded-full", children: /* @__PURE__ */ (0, import_jsx_runtime25.jsx)(
    "div",
    {
      onClick,
      className: cn(
        "border-[3px] border-background fill-current text-black bg-interactive hover:bg-interactive/80 dark:text-primary-text-35  dark:hover:border-primary dark:border dark:border-white-35 h-8 w-8 rounded-full flex items-center justify-center cursor-pointer",
        className,
        {
          "opacity-50 cursor-not-allowed": disabled
        }
      ),
      children: /* @__PURE__ */ (0, import_jsx_runtime25.jsx)(IconSwitchPairDark, {})
    }
  ) });
};
var SwitchPairButton_default = SwitchPairButton;

// src/components/useTimeDiff/useTimeDiff.tsx
var import_react15 = __toESM(require("react"));
var useTimeDiff = () => {
  const { lastRefreshTimestamp } = useSwapContext();
  const [hasExpired, setHasExpired] = import_react15.default.useState(false);
  const [timeDiff, setTimeDiff] = (0, import_react15.useState)(0);
  (0, import_react15.useEffect)(() => {
    const intervalId = setInterval(() => {
      if (!lastRefreshTimestamp) {
        return;
      }
      const value = Date.now() > lastRefreshTimestamp + ROUTE_CACHE_DURATION;
      const elapsedSeconds = (Date.now() - (lastRefreshTimestamp + ROUTE_CACHE_DURATION)) / 1e3;
      setTimeDiff(elapsedSeconds / (ROUTE_CACHE_DURATION / 1e3) * 100);
      setHasExpired(value);
    }, 1e3);
    return () => clearInterval(intervalId);
  }, [lastRefreshTimestamp]);
  return [hasExpired, timeDiff];
};
var useTimeDiff_default = useTimeDiff;

// src/components/Form.tsx
var import_decimal7 = __toESM(require("decimal.js"));

// src/types/constants.ts
var SwapMode = {
  ExactInOrOut: "ExactInOrOut",
  ExactIn: "ExactIn",
  ExactOut: "ExactOut"
};

// src/components/JupShield.tsx
var import_react_query6 = require("@tanstack/react-query");

// src/hooks/useMobile.ts
var import_react16 = require("react");
var import_react_use = require("react-use");
var useMobile = () => {
  const [isDesktop, setIsDesktop] = (0, import_react16.useState)(false);
  (0, import_react_use.useIsomorphicLayoutEffect)(() => {
    function updateSize() {
      const desktopQuery = window.matchMedia("(min-width: 1024px)");
      setIsDesktop(desktopQuery.matches);
    }
    updateSize();
    window.addEventListener("resize", updateSize);
    return () => window.removeEventListener("resize", updateSize);
  }, []);
  return !isDesktop;
};

// src/components/Popover/Popover.tsx
var import_react17 = require("react");
var import_react_popper = require("react-popper");
var import_lodash = __toESM(require("lodash.debounce"));
var import_jsx_runtime26 = require("react/jsx-runtime");
var Popover = ({
  placement = "auto",
  trigger = "click",
  persistOnClick = true,
  strategy = "fixed",
  arrow = false,
  buttonContent,
  popoverContent,
  contentClassName,
  buttonContentClassName,
  isOpen,
  onClose,
  matchWidth,
  offset,
  drawShades = false
}) => {
  const isLocalMode = (0, import_react17.useMemo)(
    () => typeof isOpen === "undefined",
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  const isMobile = useMobile();
  const [localOpen, setLocalOpen] = (0, import_react17.useState)(isLocalMode ? false : true);
  const [referenceElement, setReferenceElement] = (0, import_react17.useState)(null);
  const [popperElement, setPopperElement] = (0, import_react17.useState)(null);
  const shouldShowArrow = (0, import_react17.useMemo)(() => Boolean(popoverContent) && arrow, [popoverContent, arrow]);
  const modifiers = offset ? [
    {
      name: "offset",
      options: {
        offset
      }
    }
  ] : void 0;
  const { styles, attributes: attributes2 } = (0, import_react_popper.usePopper)(referenceElement, popperElement, {
    strategy,
    placement,
    modifiers
  });
  const handleClose = (0, import_react17.useCallback)(
    (0, import_lodash.default)(
      () => {
        if (isLocalMode) {
          setLocalOpen(false);
        } else if (onClose) {
          onClose();
        }
      },
      // persistOnClick makes sure Tooltip stays open when user click on it
      // to make sure that if user hover to content, we don't close it
      trigger === "hover" && persistOnClick ? 50 : 0
    ),
    [setLocalOpen, isLocalMode, onClose, trigger, persistOnClick]
  );
  const handleOpen = (0, import_react17.useCallback)(() => {
    handleClose.cancel();
    setLocalOpen(true);
  }, [setLocalOpen, handleClose]);
  useOutsideClick({ current: popperElement }, handleClose);
  const onClick = () => {
    if (isLocalMode) {
      if (persistOnClick && trigger === "hover" || trigger === "click") {
        handleOpen();
      }
    } else if (isOpen && onClose) {
      onClose();
    }
  };
  const shouldRenderContent = isLocalMode ? localOpen : isOpen;
  const hoverProps = trigger === "hover" ? isMobile ? {
    onTouchStart: handleOpen,
    onTouchEnd: handleClose
  } : {
    onMouseEnter: handleOpen,
    onMouseLeave: handleClose
  } : {};
  return /* @__PURE__ */ (0, import_jsx_runtime26.jsxs)(import_jsx_runtime26.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime26.jsx)(
      "div",
      __spreadProps(__spreadValues({
        ref: setReferenceElement,
        onClick,
        className: cn(buttonContentClassName, {
          "z-50": shouldRenderContent
        })
      }, hoverProps), {
        children: buttonContent
      })
    ),
    shouldRenderContent && /* @__PURE__ */ (0, import_jsx_runtime26.jsxs)(
      "div",
      __spreadProps(__spreadValues(__spreadValues({
        id: "tooltip",
        ref: setPopperElement,
        style: __spreadProps(__spreadValues({}, styles.popper), { maxWidth: matchWidth ? referenceElement == null ? void 0 : referenceElement.clientWidth : "auto" })
      }, attributes2.popper), hoverProps), {
        className: cn(
          "rounded-lg w-auto bg-none shadow-xl dark:bg-white/5 backdrop-blur-xl transition-opacity opacity-0",
          contentClassName,
          {
            "z-50 opacity-100": shouldRenderContent
          }
        ),
        children: [
          popoverContent,
          shouldShowArrow ? /* @__PURE__ */ (0, import_jsx_runtime26.jsx)(
            "div",
            {
              id: "arrow",
              className: "before:absolute absolute before:w-2 w-2 before:h-2 h-2 before:bg-inherit bg-inherit before:visible before:rotate-45 invisible",
              "data-popper-arrow": true
            }
          ) : null
        ]
      })
    ),
    /* @__PURE__ */ (0, import_jsx_runtime26.jsx)(
      "div",
      {
        className: cn("fixed top-0 left-0 w-full h-full transition-all opacity-0 pointer-events-none z-[-1]", {
          "!backdrop-blur-xxs !bg-black/20 opacity-100 pointer-events-auto z-40": drawShades && shouldRenderContent
        }),
        onClick: handleClose
      }
    )
  ] });
};
var Popover_default = Popover;

// src/components/Tooltip/PopoverTooltip.tsx
var import_jsx_runtime27 = require("react/jsx-runtime");
var PopoverTooltip = ({
  content,
  disabled = false,
  variant = "dark",
  children,
  offset,
  placement,
  persistOnClick,
  buttonContentClassName,
  contentClassName,
  drawShades,
  strategy
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime27.jsx)(
    Popover_default,
    {
      placement: placement || "top",
      buttonContent: children,
      buttonContentClassName,
      offset: offset || [0, 2.5],
      contentClassName,
      drawShades,
      popoverContent: content && /* @__PURE__ */ (0, import_jsx_runtime27.jsx)(
        "div",
        {
          className: cn("p-2 rounded-lg max-w-[360px] w-full text-xs", {
            "bg-white text-black": variant === "light",
            "bg-black text-primary-text": variant === "dark",
            "group-hover:visible group-hover:z-50": !disabled
          }),
          children: content
        }
      ),
      id: "TooltipPopover",
      trigger: persistOnClick ? "click" : "hover",
      persistOnClick,
      strategy
    }
  );
};
var PopoverTooltip_default = PopoverTooltip;

// src/components/JupShield.tsx
var import_react18 = require("react");

// src/components/Plural/index.tsx
var import_jsx_runtime28 = require("react/jsx-runtime");
var Plural = ({ value, one, other }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime28.jsx)(import_jsx_runtime28.Fragment, { children: value === 1 ? one.replace("#", value.toString()) : other.replace("#", value.toString()) });
};
var Plural_default = Plural;

// src/icons/InfoIcon.tsx
var import_jsx_runtime29 = require("react/jsx-runtime");
var InfoIcon = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime29.jsx)("svg", __spreadProps(__spreadValues({ xmlns: "http://www.w3.org/2000/svg", height: "1em", width: "1em", viewBox: "0 0 256 256" }, props), { children: /* @__PURE__ */ (0, import_jsx_runtime29.jsx)(
    "path",
    {
      fill: "currentColor",
      d: "M128 24a104 104 0 1 0 104 104A104.11 104.11 0 0 0 128 24m0 192a88 88 0 1 1 88-88a88.1 88.1 0 0 1-88 88m-8-80V80a8 8 0 0 1 16 0v56a8 8 0 0 1-16 0m20 36a12 12 0 1 1-12-12a12 12 0 0 1 12 12"
    }
  ) }));
};
var InfoIcon_default = InfoIcon;

// src/components/JupShield.tsx
var import_jsx_runtime30 = require("react/jsx-runtime");
var parseShieldWarningtoSentenceCase = (warning) => {
  const str = warning.type.replace(/_/g, " ");
  return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
};
var JupShieldIcon = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime30.jsx)("svg", __spreadProps(__spreadValues({ xmlns: "http://www.w3.org/2000/svg", height: "1em", width: "1em", viewBox: "0 0 24 24" }, props), { children: /* @__PURE__ */ (0, import_jsx_runtime30.jsxs)("g", { fill: "none", children: [
    /* @__PURE__ */ (0, import_jsx_runtime30.jsx)(
      "path",
      {
        stroke: "currentColor",
        strokeWidth: "1.5",
        d: "M3 10.417c0-3.198 0-4.797.378-5.335c.377-.537 1.88-1.052 4.887-2.081l.573-.196C10.405 2.268 11.188 2 12 2s1.595.268 3.162.805l.573.196c3.007 1.029 4.51 1.544 4.887 2.081C21 5.62 21 7.22 21 10.417v1.574c0 5.638-4.239 8.375-6.899 9.536C13.38 21.842 13.02 22 12 22s-1.38-.158-2.101-.473C7.239 20.365 3 17.63 3 11.991z"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime30.jsx)("path", { stroke: "currentColor", strokeLinecap: "round", strokeWidth: "1.5", d: "M12 8v4" }),
    /* @__PURE__ */ (0, import_jsx_runtime30.jsx)("circle", { cx: "12", cy: "15", r: "1", fill: "currentColor" })
  ] }) }));
};
var Warnings = ({
  warning,
  children,
  isHighRisk
}) => {
  const warningTitle = (0, import_react18.useCallback)((warning2) => {
    return parseShieldWarningtoSentenceCase(warning2);
  }, []);
  return /* @__PURE__ */ (0, import_jsx_runtime30.jsxs)("div", { className: "flex gap-1", children: [
    children,
    /* @__PURE__ */ (0, import_jsx_runtime30.jsxs)("div", { className: "flex flex-1 flex-col", children: [
      /* @__PURE__ */ (0, import_jsx_runtime30.jsx)("span", { className: cn("text-neutral-300", isHighRisk && "text-warning"), children: warningTitle(warning) }),
      /* @__PURE__ */ (0, import_jsx_runtime30.jsx)("span", { className: cn("text-neutral-500", isHighRisk && "text-amber-200"), children: warning.message })
    ] })
  ] }, warning.type);
};
var JupShield = ({ tokenAddress }) => {
  const isMobile = useMobile();
  const { data, isFetching } = (0, import_react_query6.useQuery)({
    queryKey: ["shield", tokenAddress],
    queryFn: () => ultraSwapService.getShield([tokenAddress]),
    cacheTime: 5 * 6e4,
    staleTime: 5 * 6e4,
    keepPreviousData: true,
    select: (data2) => {
      const warnings = data2.warnings[tokenAddress];
      return {
        isNotVerified: warnings.find((warning) => warning.type === "NOT_VERIFIED"),
        totalWarnings: warnings.length,
        highRiskWarnings: warnings.filter(
          (warning) => warning.severity === "critical" /* CRITICAL */ || warning.severity === "warning" /* WARNING */
        ),
        otherWarnings: warnings.filter((warning) => warning.severity === "info" /* INFO */)
      };
    }
  });
  if (!data || data.totalWarnings === 0) return null;
  const { isNotVerified, totalWarnings, highRiskWarnings, otherWarnings } = data;
  if (!isNotVerified) return null;
  return /* @__PURE__ */ (0, import_jsx_runtime30.jsx)(
    PopoverTooltip_default,
    {
      persistOnClick: isMobile,
      placement: "bottom",
      drawShades: true,
      buttonContentClassName: "!cursor-help",
      offset: [130, 5],
      content: /* @__PURE__ */ (0, import_jsx_runtime30.jsxs)("div", { className: "flex flex-col gap-y-2 p-0 ", children: [
        /* @__PURE__ */ (0, import_jsx_runtime30.jsxs)("div", { className: "text-sm font-semibold text-warning", children: [
          totalWarnings,
          " JupShield ",
          /* @__PURE__ */ (0, import_jsx_runtime30.jsx)(Plural_default, { one: "Warning", other: "Warnings", value: totalWarnings })
        ] }),
        highRiskWarnings.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime30.jsx)("div", { className: "flex flex-col gap-y-2 rounded-md bg-warning/[8%] p-2", children: highRiskWarnings.map((warning) => /* @__PURE__ */ (0, import_jsx_runtime30.jsx)(Warnings, { warning, isHighRisk: true, children: /* @__PURE__ */ (0, import_jsx_runtime30.jsx)(InfoIcon_default, { width: 15, height: 15, className: "text-warning" }) }, warning.type)) }),
        otherWarnings.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime30.jsx)("div", { className: cn("flex flex-col gap-y-2", highRiskWarnings.length > 0 && "p-2"), children: otherWarnings.map((warning) => /* @__PURE__ */ (0, import_jsx_runtime30.jsx)(Warnings, { warning, isHighRisk: false, children: /* @__PURE__ */ (0, import_jsx_runtime30.jsx)(InfoIcon_default, { width: 15, height: 15 }) }, warning.type)) })
      ] }),
      children: /* @__PURE__ */ (0, import_jsx_runtime30.jsxs)("div", { className: cn("mt-1 flex items-center gap-x-1", isFetching && "blur-sm"), children: [
        /* @__PURE__ */ (0, import_jsx_runtime30.jsx)(JupShieldIcon, { width: 15, height: 15, className: "text-warning" }),
        /* @__PURE__ */ (0, import_jsx_runtime30.jsxs)(
          "span",
          {
            className: cn(
              "!text-xxs font-semibold text-warning underline decoration-warning decoration-dashed underline-offset-4"
            ),
            children: [
              totalWarnings,
              " Warnings"
            ]
          }
        )
      ] })
    }
  );
};
var JupShield_default = JupShield;

// src/components/Form.tsx
var import_jsx_runtime31 = require("react/jsx-runtime");
var FormInputContainer = ({ tokenInfo, onBalanceClick, title, pairSelectDisabled, onClickSelectPair, children, value }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)(
    "div",
    {
      className: cn(
        "border border-transparent bg-module rounded-xl transition-all",
        "py-3 px-4 flex flex-col dark:text-primary-text  gap-y-2",
        "group focus-within:border-primary/50 focus-within:shadow-swap-input-dark rounded-xl"
      ),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)("div", { className: "flex justify-between items-center text-xs text-primary-text", children: [
          /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("div", { children: title }),
          tokenInfo && /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)(
            "div",
            {
              className: cn("flex  space-x-1 text-xs items-center text-primary-text/50 fill-current cursor-pointer"),
              onClick: (e) => {
                onBalanceClick(e);
              },
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(WalletIcon_default, { width: 10, height: 10 }),
                /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(Coinbalance_default, { mintAddress: tokenInfo.id, hideZeroBalance: false }),
                /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("span", { children: tokenInfo.symbol })
              ]
            }
          )
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)("div", { className: "flex", children: [
          /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)("div", { children: [
            /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)(
              "button",
              {
                type: "button",
                className: cn("py-2 px-3 rounded-lg flex items-center bg-interactive text-primary-text", {
                  "hover:bg-interactive/80": !pairSelectDisabled
                }),
                disabled: pairSelectDisabled,
                onClick: onClickSelectPair,
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("div", { className: "h-5 w-5", children: /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(TokenIcon_default, { info: tokenInfo, width: 20, height: 20 }) }),
                  /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("div", { className: "ml-4 mr-2 font-semibold", translate: "no", children: /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("div", { className: "truncate", children: tokenInfo == null ? void 0 : tokenInfo.symbol }) }),
                  pairSelectDisabled ? null : /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("span", { className: "text-primary-text/25 fill-current", children: /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(ChevronDownIcon_default, {}) })
                ]
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("div", { className: "flex justify-between items-center h-[20px]", children: (tokenInfo == null ? void 0 : tokenInfo.id) && /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(JupShield_default, { tokenAddress: tokenInfo.id }) })
          ] }),
          /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)("div", { className: "flex flex-col items-end justify-between w-full", children: [
            children,
            /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("span", { className: "text-xs text-primary-text/50", children: tokenInfo && /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(CoinBalanceUSD, { tokenInfo, amount: value }) })
          ] })
        ] })
      ]
    }
  );
};
var Form = ({ onSubmit, isDisabled, setSelectPairSelector }) => {
  const { publicKey } = useWalletPassThrough();
  const { data: balances } = useBalances();
  const {
    form,
    setForm,
    fromTokenInfo,
    toTokenInfo,
    quoteResponseMeta,
    formProps: { fixedAmount, swapMode, fixedMint },
    loading,
    refresh,
    quoteError,
    errors,
    isToPairFocused,
    onSubmit: onSubmitUltra
  } = useSwapContext();
  const [hasExpired, timeDiff] = useTimeDiff_default();
  const { setScreen } = useScreenState();
  (0, import_react19.useEffect)(() => {
    if (hasExpired) {
      refresh();
    }
  }, [hasExpired]);
  const shouldDisabledFromSelector = (0, import_react19.useMemo)(() => {
    if ((fromTokenInfo == null ? void 0 : fromTokenInfo.id) === fixedMint) {
      return true;
    }
    return false;
  }, [fixedMint, fromTokenInfo == null ? void 0 : fromTokenInfo.id]);
  const shouldDisabledToSelector = (0, import_react19.useMemo)(() => {
    if ((toTokenInfo == null ? void 0 : toTokenInfo.id) === fixedMint) {
      return true;
    }
    return false;
  }, [fixedMint, toTokenInfo == null ? void 0 : toTokenInfo.id]);
  const walletPublicKey = (0, import_react19.useMemo)(() => publicKey == null ? void 0 : publicKey.toString(), [publicKey]);
  const onChangeFromValue = ({ value }) => {
    if (value === "") {
      setForm((form2) => __spreadProps(__spreadValues({}, form2), { fromValue: "", toValue: "" }));
      return;
    }
    const isInvalid = Number.isNaN(value);
    if (isInvalid) return;
    setForm((form2) => __spreadProps(__spreadValues({}, form2), { fromValue: value }));
  };
  const onChangeToValue = ({ value }) => {
    if (value === "") {
      setForm((form2) => __spreadProps(__spreadValues({}, form2), { fromValue: "", toValue: "" }));
      return;
    }
    const isInvalid = Number.isNaN(value);
    if (isInvalid) return;
    setForm((form2) => __spreadProps(__spreadValues({}, form2), { toValue: value }));
  };
  const balance = (0, import_react19.useMemo)(() => {
    if (!(fromTokenInfo == null ? void 0 : fromTokenInfo.id)) return null;
    if (!balances) return null;
    const accBalanceObj = balances[fromTokenInfo.id];
    if (!accBalanceObj) return null;
    return accBalanceObj.uiAmount.toString();
  }, [balances, fromTokenInfo == null ? void 0 : fromTokenInfo.id]);
  const onClickMax = (0, import_react19.useCallback)(
    (e) => {
      e.preventDefault();
      if (!balance) return;
      if ((fromTokenInfo == null ? void 0 : fromTokenInfo.id) === WRAPPED_SOL_MINT.toBase58()) {
        setForm((prev) => __spreadProps(__spreadValues({}, prev), {
          fromValue: new import_decimal7.default(balance).gt(MINIMUM_SOL_BALANCE) ? new import_decimal7.default(balance).minus(MINIMUM_SOL_BALANCE).toFixed(9) : "0"
        }));
      } else {
        setForm((prev) => __spreadProps(__spreadValues({}, prev), {
          fromValue: balance
        }));
      }
    },
    [balance, fromTokenInfo == null ? void 0 : fromTokenInfo.id, setForm]
  );
  const onClickSwitchPair = () => {
    setForm((prev) => __spreadProps(__spreadValues({}, prev), {
      fromValue: "",
      toValue: "",
      fromMint: prev.toMint,
      toMint: prev.fromMint
    }));
  };
  const { inputAmountDisabled, outputAmountDisabled } = (0, import_react19.useMemo)(() => {
    const result = { inputAmountDisabled: true, outputAmountDisabled: true };
    if (!fixedAmount) {
      if (swapMode === SwapMode.ExactIn) {
        result.inputAmountDisabled = false;
      } else if (swapMode === SwapMode.ExactOut) {
        result.outputAmountDisabled = false;
      } else {
        result.inputAmountDisabled = false;
        result.outputAmountDisabled = false;
      }
    }
    return result;
  }, [fixedAmount, swapMode]);
  const onClickSelectFromMint = (0, import_react19.useCallback)(() => {
    if (shouldDisabledFromSelector) return;
    setSelectPairSelector("fromMint");
  }, [shouldDisabledFromSelector, setSelectPairSelector]);
  const onClickSelectToMint = (0, import_react19.useCallback)(() => {
    if (shouldDisabledToSelector) return;
    setSelectPairSelector("toMint");
  }, [shouldDisabledToSelector, setSelectPairSelector]);
  const thousandSeparator = (0, import_react19.useMemo)(() => detectedSeparator === "," ? "." : ",", []);
  const withValueLimit = (0, import_react19.useCallback)(
    ({ floatValue }) => !floatValue || floatValue <= MAX_INPUT_LIMIT,
    []
  );
  const handleClick = (0, import_react19.useCallback)(
    (event) => {
      if (window.Jupiter.enableWalletPassthrough && window.Jupiter.onRequestConnectWallet) {
        window.Jupiter.onRequestConnectWallet();
      } else {
        setScreen("Wallet");
      }
    },
    [setScreen]
  );
  const shouldButtonDisabled = (0, import_react19.useMemo)(() => {
    if (isDisabled || loading || !!errors.fromValue) {
      return true;
    }
    return false;
  }, [isDisabled, loading, errors.fromValue]);
  return /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)("div", { className: "h-full flex flex-col items-center justify-center", children: [
    /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("div", { className: "w-full mt-2 rounded-xl flex flex-col px-2", children: /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)("div", { className: "flex-col", children: [
      /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(
        FormInputContainer,
        {
          tokenInfo: fromTokenInfo,
          onBalanceClick: (e) => {
            isToPairFocused.current = false;
            onClickMax(e);
          },
          title: "Selling",
          pairSelectDisabled: shouldDisabledFromSelector,
          onClickSelectPair: onClickSelectFromMint,
          value: form.fromValue,
          children: (fromTokenInfo == null ? void 0 : fromTokenInfo.decimals) && /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(
            import_react_number_format.NumericFormat,
            {
              disabled: fixedAmount || swapMode === SwapMode.ExactOut,
              value: typeof form.fromValue === "undefined" ? "" : form.fromValue,
              decimalScale: fromTokenInfo.decimals,
              thousandSeparator,
              allowNegative: false,
              valueIsNumericString: true,
              inputMode: "decimal",
              onValueChange: onChangeFromValue,
              placeholder: "0.00",
              className: cn(
                "w-full h-[40px] bg-transparent text-primary-text text-right font-semibold text-xl placeholder:text-primary-text/50",
                {
                  "cursor-not-allowed": inputAmountDisabled || swapMode === SwapMode.ExactOut
                }
              ),
              onKeyDown: () => {
                isToPairFocused.current = false;
              },
              decimalSeparator: detectedSeparator,
              isAllowed: withValueLimit
            }
          )
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("div", { className: "relative z-10 -my-3 flex justify-center", children: /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(SwitchPairButton_default, { onClick: onClickSwitchPair, className: cn("transition-all") }) }),
      /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(
        FormInputContainer,
        {
          tokenInfo: toTokenInfo,
          onBalanceClick: onClickMax,
          title: "Buying",
          pairSelectDisabled: shouldDisabledToSelector,
          onClickSelectPair: onClickSelectToMint,
          value: form.toValue,
          children: (toTokenInfo == null ? void 0 : toTokenInfo.decimals) && /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(
            import_react_number_format.NumericFormat,
            {
              inputMode: "decimal",
              disabled: outputAmountDisabled || swapMode === SwapMode.ExactIn,
              value: typeof form.toValue === "undefined" ? "" : form.toValue,
              decimalScale: toTokenInfo.decimals,
              thousandSeparator,
              allowNegative: false,
              valueIsNumericString: true,
              onValueChange: onChangeToValue,
              className: cn(
                "h-[40px] w-full bg-transparent text-primary-text text-right font-semibold text-lg placeholder:text-primary-text/50",
                {
                  "cursor-not-allowed": outputAmountDisabled || swapMode === SwapMode.ExactIn
                }
              ),
              placeholder: swapMode === SwapMode.ExactOut ? "Enter desired amount" : "0.00",
              decimalSeparator: detectedSeparator,
              isAllowed: withValueLimit,
              onKeyDown: (e) => {
                if (e.metaKey || e.ctrlKey || e.key === "Meta" || e.key === "Control" || e.key === "Alt" || e.key === "Shift") {
                  return;
                }
                isToPairFocused.current = true;
              }
            }
          )
        }
      )
    ] }) }),
    /* @__PURE__ */ (0, import_jsx_runtime31.jsxs)("div", { className: "w-full px-2", children: [
      !walletPublicKey ? /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(JupButton_default, { size: "lg", className: "w-full mt-4 bg-primary !text-uiv2-text/75", onClick: handleClick, children: "Connect Wallet" }) : /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(
        JupButton_default,
        {
          size: "lg",
          className: cn("w-full mt-4 disabled:opacity-50 !text-uiv2-text/75 !bg-primary "),
          onClick: () => {
            onSubmit();
            onSubmitUltra();
          },
          disabled: shouldButtonDisabled,
          children: loading ? /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("span", { children: "Loading" }) : errors.fromValue ? /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("span", { children: errors.fromValue.title }) : /* @__PURE__ */ (0, import_jsx_runtime31.jsx)("span", { children: "Swap" })
        }
      ),
      quoteResponseMeta && fromTokenInfo && toTokenInfo ? /* @__PURE__ */ (0, import_jsx_runtime31.jsx)(
        PriceInfo_default,
        {
          quoteResponse: quoteResponseMeta,
          fromTokenInfo,
          toTokenInfo,
          loading
        }
      ) : null
    ] })
  ] });
};
var Form_default = Form;

// src/components/FormPairSelector.tsx
var import_react23 = require("react");
var import_react_virtualized_auto_sizer = __toESM(require("react-virtualized-auto-sizer"));
var import_react_window = require("react-window");

// src/icons/LeftArrowIcon.tsx
var import_jsx_runtime32 = require("react/jsx-runtime");
var LeftArrowIcon = ({ width = "12", height = "12" }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime32.jsx)("svg", { width, height, viewBox: "0 0 53 52", fill: "none", xmlns: "http://www.w3.org/2000/svg", children: /* @__PURE__ */ (0, import_jsx_runtime32.jsx)(
    "path",
    {
      d: "M17.3336 23.8334H43.7062V28.1667H17.3336L28.9555 39.7887L25.8919 42.8524L9.03955 26L25.8919 9.14771L28.9555 12.2114L17.3336 23.8334Z",
      fill: "currentColor"
    }
  ) });
};
var LeftArrowIcon_default = LeftArrowIcon;

// src/icons/SearchIcon.tsx
var import_jsx_runtime33 = require("react/jsx-runtime");
var SearchIcon = () => {
  return /* @__PURE__ */ (0, import_jsx_runtime33.jsx)("div", { className: "flex items-center text-primary-text fill-current", children: /* @__PURE__ */ (0, import_jsx_runtime33.jsx)("svg", { width: "18", height: "18", viewBox: "0 0 18 18", fill: "inherit", xmlns: "http://www.w3.org/2000/svg", children: /* @__PURE__ */ (0, import_jsx_runtime33.jsx)(
    "path",
    {
      d: "M7.30327 14.6058C8.75327 14.6074 10.1705 14.1746 11.3729 13.3637L15.5971 17.5871C16.1463 18.1371 17.0377 18.1371 17.5877 17.5871C18.1377 17.0371 18.1377 16.1457 17.5877 15.5964L13.3643 11.3722C14.5823 9.55661 14.9229 7.28943 14.2909 5.19563C13.6596 3.10183 12.1229 1.40183 10.1033 0.56283C8.08365 -0.276231 5.79385 -0.16607 3.86505 0.86283C1.93537 1.89251 0.569053 3.73243 0.140853 5.87683C-0.286487 8.02143 0.269759 10.2448 1.65725 11.9354C3.04397 13.6261 5.11665 14.6064 7.30325 14.6058H7.30327ZM7.30327 1.68943C8.79233 1.68865 10.2197 2.28005 11.2729 3.33319C12.3252 4.38631 12.9166 5.81359 12.9166 7.30279C12.9166 8.79199 12.3252 10.2192 11.2729 11.2724C10.2198 12.3247 8.79247 12.9162 7.30327 12.9162C5.81407 12.9162 4.38687 12.3247 3.33367 11.2724C2.28133 10.2193 1.68913 8.79199 1.68991 7.30279C1.69148 5.81451 2.28287 4.38719 3.33523 3.33479C4.38759 2.28239 5.81483 1.69103 7.30323 1.68947L7.30327 1.68943Z",
      fill: "inherit",
      fillOpacity: "0.25"
    }
  ) }) });
};
var SearchIcon_default = SearchIcon;

// src/components/FormPairSelector.tsx
var import_lodash2 = __toESM(require("lodash.debounce"));

// src/components/FormPairRow.tsx
var import_react21 = __toESM(require("react"));
var import_decimal8 = __toESM(require("decimal.js"));

// src/contexts/preferredExplorer/index.tsx
var import_react20 = require("react");
var import_jsx_runtime34 = require("react/jsx-runtime");
var AVAILABLE_EXPLORER = [
  {
    name: "Solana Explorer",
    url: "https://explorer.solana.com/",
    get: (txid, cluster = "mainnet-beta") => {
      if (cluster !== "mainnet-beta") return `https://explorer.solana.com/tx/${txid}?cluster=${cluster}`;
      return `https://explorer.solana.com/tx/${txid}`;
    },
    getToken: (mint, cluster = "mainnet-beta") => {
      if (cluster !== "mainnet-beta") return `https://explorer.solana.com/address/${mint}?cluster=${cluster}`;
      return `https://explorer.solana.com/address/${mint}`;
    }
  },
  {
    name: "Solscan",
    url: "https://solscan.io/",
    get: (txid, cluster = "mainnet-beta") => {
      if (cluster !== "mainnet-beta") return `https://solscan.io/tx/${txid}?cluster=${cluster}`;
      return `https://solscan.io/tx/${txid}`;
    },
    getToken: (mint, cluster = "mainnet-beta") => {
      if (cluster !== "mainnet-beta") return `https://solscan.io/token/${mint}?cluster=${cluster}`;
      return `https://solscan.io/token/${mint}`;
    }
  },
  {
    name: "Solana Beach",
    url: "https://solanabeach.io/",
    get: (txid, cluster = "mainnet-beta") => {
      if (cluster !== "mainnet-beta") return `https://solanabeach.io/transaction/${txid}?cluster=${cluster}`;
      return `https://solanabeach.io/transaction/${txid}`;
    },
    getToken: (mint, cluster = "mainnet-beta") => {
      if (cluster !== "mainnet-beta") return `https://solanabeach.io/address/${mint}?cluster=${cluster}`;
      return `https://solanabeach.io/address/${mint}`;
    }
  },
  {
    name: "SolanaFM",
    url: "https://solana.fm/",
    get: (txid, cluster = "mainnet-beta") => {
      if (cluster === "devnet") return `https://solana.fm/tx/${txid}?cluster=devnet-solana`;
      if (cluster === "testnet") return `https://solana.fm/tx/${txid}?cluster=testnet-qn1`;
      return `https://solana.fm/tx/${txid}`;
    },
    getToken: (mint, cluster = "mainnet-beta") => {
      if (cluster === "devnet") return `https://solana.fm/address/${mint}?cluster=devnet-solana`;
      if (cluster === "testnet") return `https://solana.fm/address/${mint}?cluster=testnet-qn1`;
      return `https://solana.fm/address/${mint}`;
    }
  }
];
var PreferredExplorerContext = (0, import_react20.createContext)({
  explorer: AVAILABLE_EXPLORER[0].name,
  getExplorer: (txid, cluster) => "",
  getTokenExplorer: (mint, cluster) => "",
  setExplorer: (explorer) => {
  }
});
var PreferredExplorerProvider = ({
  defaultExplorer,
  children
}) => {
  const [explorer, setExplorer] = (0, import_react20.useState)(defaultExplorer != null ? defaultExplorer : AVAILABLE_EXPLORER[0].name);
  const explorerObject = (0, import_react20.useMemo)(() => {
    return AVAILABLE_EXPLORER.find((e) => e.name === explorer) || AVAILABLE_EXPLORER[0];
  }, [explorer]);
  const getExplorer = (0, import_react20.useCallback)(
    (txid, cluster) => explorerObject.get(txid, cluster),
    [explorerObject]
  );
  const getTokenExplorer = (0, import_react20.useCallback)(
    (mint, cluster) => explorerObject.getToken(mint, cluster),
    [explorerObject]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime34.jsx)(
    PreferredExplorerContext.Provider,
    {
      value: { explorer, getExplorer, getTokenExplorer, setExplorer: (explorer2) => setExplorer(explorer2) },
      children
    }
  );
};
function usePreferredExplorer() {
  const context = (0, import_react20.useContext)(PreferredExplorerContext);
  return context;
}

// src/icons/ExternalIcon.tsx
var import_jsx_runtime35 = require("react/jsx-runtime");
var ExternalIcon = () => /* @__PURE__ */ (0, import_jsx_runtime35.jsx)("svg", { width: "10", height: "10", viewBox: "0 0 10 10", fill: "none", xmlns: "http://www.w3.org/2000/svg", children: /* @__PURE__ */ (0, import_jsx_runtime35.jsx)(
  "path",
  {
    d: "M4 2V3H1.5V8.5H7V6H8V9C8 9.13261 7.94732 9.25979 7.85355 9.35355C7.75979 9.44732 7.63261 9.5 7.5 9.5H1C0.867392 9.5 0.740215 9.44732 0.646447 9.35355C0.552678 9.25979 0.5 9.13261 0.5 9V2.5C0.5 2.36739 0.552678 2.24021 0.646447 2.14645C0.740215 2.05268 0.867392 2 1 2H4ZM9.5 0.5V4.5H8.5V2.2065L4.6035 6.1035L3.8965 5.3965L7.7925 1.5H5.5V0.5H9.5Z",
    fill: "currentColor"
  }
) });
var ExternalIcon_default = ExternalIcon;

// src/components/TokenLink.tsx
var import_jsx_runtime36 = require("react/jsx-runtime");
var TokenLink = ({ tokenInfo, className }) => {
  const { getTokenExplorer } = usePreferredExplorer();
  return /* @__PURE__ */ (0, import_jsx_runtime36.jsxs)(
    "a",
    {
      target: "_blank",
      rel: "noreferrer",
      className: cn(
        "flex items-center bg-black/25 text-primary-text/75 px-2 py-0.5 space-x-1 rounded cursor-pointer",
        className
      ),
      href: getTokenExplorer(tokenInfo.id),
      onClick: (e) => e.stopPropagation(),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime36.jsx)("div", { className: "text-xxs", children: shortenAddress(tokenInfo.id) }),
        /* @__PURE__ */ (0, import_jsx_runtime36.jsx)(ExternalIcon_default, {})
      ]
    }
  );
};
TokenLink.displayName = "TokenLink";
var TokenLink_default = TokenLink;

// src/components/CheckedBadge.tsx
var import_jsx_runtime37 = require("react/jsx-runtime");
var CheckedBadge = ({ width = 24, height = 24 }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime37.jsxs)(
    "svg",
    {
      width,
      height,
      viewBox: "0 0 24 24",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime37.jsx)(
          "path",
          {
            d: "M13.8179 4.54512L13.6275 4.27845C12.8298 3.16176 11.1702 3.16176 10.3725 4.27845L10.1821 4.54512C9.76092 5.13471 9.05384 5.45043 8.33373 5.37041L7.48471 5.27608C6.21088 5.13454 5.13454 6.21088 5.27608 7.48471L5.37041 8.33373C5.45043 9.05384 5.13471 9.76092 4.54512 10.1821L4.27845 10.3725C3.16176 11.1702 3.16176 12.8298 4.27845 13.6275L4.54512 13.8179C5.13471 14.2391 5.45043 14.9462 5.37041 15.6663L5.27608 16.5153C5.13454 17.7891 6.21088 18.8655 7.48471 18.7239L8.33373 18.6296C9.05384 18.5496 9.76092 18.8653 10.1821 19.4549L10.3725 19.7215C11.1702 20.8382 12.8298 20.8382 13.6275 19.7215L13.8179 19.4549C14.2391 18.8653 14.9462 18.5496 15.6663 18.6296L16.5153 18.7239C17.7891 18.8655 18.8655 17.7891 18.7239 16.5153L18.6296 15.6663C18.5496 14.9462 18.8653 14.2391 19.4549 13.8179L19.7215 13.6275C20.8382 12.8298 20.8382 11.1702 19.7215 10.3725L19.4549 10.1821C18.8653 9.76092 18.5496 9.05384 18.6296 8.33373L18.7239 7.48471C18.8655 6.21088 17.7891 5.13454 16.5153 5.27608L15.6663 5.37041C14.9462 5.45043 14.2391 5.13471 13.8179 4.54512Z",
            stroke: "currentColor",
            strokeWidth: "2",
            strokeLinecap: "round",
            strokeLinejoin: "round"
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime37.jsx)(
          "path",
          {
            d: "M9 12L10.8189 13.8189V13.8189C10.9189 13.9189 11.0811 13.9189 11.1811 13.8189V13.8189L15 10",
            stroke: "currentColor",
            strokeWidth: "2",
            strokeLinecap: "round",
            strokeLinejoin: "round"
          }
        )
      ]
    }
  );
};
var CheckedBadge_default = CheckedBadge;

// src/queries/useLstApy.ts
var import_react_query7 = require("@tanstack/react-query");
function useLstApyFetcher() {
  return (0, import_react_query7.useQuery)(
    ["lst-apy"],
    () => __async(this, null, function* () {
      const lstApy = yield fetch(`https://worker.jup.ag/lst-apys`);
      const apyResult = yield lstApy.json();
      return apyResult;
    }),
    {
      retry: 3,
      keepPreviousData: true,
      staleTime: 3e5
      // 5m
    }
  );
}

// src/components/FormPairRow.tsx
var import_react_query8 = require("@tanstack/react-query");
var import_jsx_runtime38 = require("react/jsx-runtime");
var PAIR_ROW_HEIGHT = 72;
var LSTTag = ({ mintAddress }) => {
  const { data: lstApy } = useLstApyFetcher();
  const apy = (0, import_react21.useMemo)(() => {
    if (!lstApy) return;
    const value = lstApy.apys[mintAddress];
    if (value && hasNumericValue(value)) {
      return new import_decimal8.default(value).mul(100).toDP(2).toString();
    }
    return;
  }, [lstApy, mintAddress]);
  return /* @__PURE__ */ (0, import_jsx_runtime38.jsxs)("p", { className: "rounded-md text-xxs leading-none transition-all py-0.5 px-1 text-primary/50 border border-primary/50 font-semibold", children: [
    "LST ",
    apy ? `${apy}%` : ""
  ] });
};
var MultiTags = ({ item }) => {
  const { data: balances } = useBalances();
  const isLoading = (0, import_react21.useRef)(false);
  const isLoaded = (0, import_react21.useRef)(false);
  const [renderedTag, setRenderedTag] = import_react21.default.useState({
    isVerified: false,
    isLST: false,
    // isUnknown: false,
    isToken2022: false,
    isFrozen: false
  });
  (0, import_react21.useEffect)(() => {
    if (isLoaded.current || isLoading.current) return;
    isLoading.current = true;
    setTimeout(() => {
      var _a2, _b;
      const result = {
        isVerified: checkIsStrictOrVerified(item),
        isLST: Boolean((_a2 = item.tags) == null ? void 0 : _a2.includes("lst")),
        // isUnknown: checkIsUnknownToken(item),
        isToken2022: Boolean(checkIsToken2022(item)),
        isFrozen: ((_b = balances == null ? void 0 : balances[item.id]) == null ? void 0 : _b.isFrozen) || false
      };
      setRenderedTag(result);
      isLoading.current = false;
      isLoaded.current = true;
    }, 0);
  }, []);
  const remainingTags = [];
  if (!renderedTag) return null;
  const { isToken2022, isFrozen, isLST } = renderedTag;
  return /* @__PURE__ */ (0, import_jsx_runtime38.jsxs)("div", { className: "flex justify-end gap-x-1", children: [
    isFrozen && /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("p", { className: "border rounded-md text-xxs leading-none transition-all py-0.5 px-1 border-warning/50 text-warning/50", children: "Frozen" }),
    isToken2022 && /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("p", { className: "rounded-md text-xxs leading-none transition-all py-0.5 px-1 bg-black/10 font-semibold text-primary-text/20", children: "Token2022" }),
    remainingTags == null ? void 0 : remainingTags.map((tag, idx) => /* @__PURE__ */ (0, import_jsx_runtime38.jsx)(
      "div",
      {
        className: "rounded-md text-xxs leading-none transition-all py-0.5 px-1 bg-black/10 font-semibold text-primary-text/20",
        children: tag
      },
      idx
    )),
    isLST && /* @__PURE__ */ (0, import_jsx_runtime38.jsx)(LSTTag, { mintAddress: item.id })
  ] });
};
var FormPairRow = (props) => {
  var _a2;
  const {
    item,
    style,
    onSubmit,
    suppressCloseModal,
    showExplorer = true,
    enableUnknownTokenWarning = true
  } = props;
  const queryClient = (0, import_react_query8.useQueryClient)();
  const onClick = import_react21.default.useCallback(() => {
    queryClient.setQueryData([...ASSET_QUERY_KEY, item.id], [item]);
    onSubmit(item);
    if (suppressCloseModal) return;
  }, [onSubmit, item, suppressCloseModal, queryClient]);
  const { data: balances } = useBalances();
  const balance = (_a2 = balances == null ? void 0 : balances[item.id]) == null ? void 0 : _a2.uiAmount;
  const usdValue = balance ? new import_decimal8.default(balance).mul(item.usdPrice || 0) : void 0;
  const usdValueDisplay = usdValue && usdValue.gte(0.01) ? `$${formatNumber.format(usdValue, 2)}` : "";
  return /* @__PURE__ */ (0, import_jsx_runtime38.jsx)(
    "li",
    {
      className: `rounded cursor-pointer px-5 my-1 list-none flex w-full items-center bg-interactive hover:bg-interactive/80`,
      style: __spreadValues({ maxHeight: PAIR_ROW_HEIGHT - 4, height: PAIR_ROW_HEIGHT - 4 }, style),
      onClick,
      translate: "no",
      children: /* @__PURE__ */ (0, import_jsx_runtime38.jsxs)("div", { className: "flex h-full w-full items-center space-x-4", children: [
        /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("div", { className: "flex-shrink-0", children: /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("div", { className: "bg-gray-200 rounded-full", children: /* @__PURE__ */ (0, import_jsx_runtime38.jsx)(TokenIcon_default, { info: item, width: 36, height: 36, enableUnknownTokenWarning }) }) }),
        /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("div", { className: "flex-1 min-w-0", children: /* @__PURE__ */ (0, import_jsx_runtime38.jsxs)("div", { className: "flex flex-col gap-y-px", children: [
          /* @__PURE__ */ (0, import_jsx_runtime38.jsxs)("div", { className: "flex items-center", children: [
            /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("p", { className: "text-sm font-medium text-primary-text truncate", children: item.symbol }),
            checkIsStrictOrVerified(item) && /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("p", { className: "rounded-md text-xxs leading-none transition-all py-0.5 px-1 text-primary", children: /* @__PURE__ */ (0, import_jsx_runtime38.jsx)(CheckedBadge_default, { width: 18, height: 18 }) })
          ] }),
          /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("p", { className: "text-xs text-primary-text/50 dark:text-primary-text-35 truncate", children: item.id === WRAPPED_SOL_MINT.toBase58() ? "Solana" : item.name }),
          showExplorer ? /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("div", { className: "-ml-1 z-10 w-fit", onClick: (e) => e.stopPropagation(), children: /* @__PURE__ */ (0, import_jsx_runtime38.jsx)(TokenLink_default, { tokenInfo: item }) }) : null
        ] }) }),
        /* @__PURE__ */ (0, import_jsx_runtime38.jsxs)("div", { className: "text-xs text-primary-text/50 text-right h-full flex flex-col justify-evenly", children: [
          /* @__PURE__ */ (0, import_jsx_runtime38.jsx)(Coinbalance_default, { mintAddress: item.id, hideZeroBalance: true }),
          usdValueDisplay ? /* @__PURE__ */ (0, import_jsx_runtime38.jsx)("p", { children: usdValueDisplay }) : null,
          /* @__PURE__ */ (0, import_jsx_runtime38.jsx)(MultiTags, __spreadValues({}, props))
        ] })
      ] })
    }
  );
};
var FormPairRow_default = FormPairRow;

// src/hooks/useSearch.ts
var import_react22 = require("react");
var import_react_query9 = require("@tanstack/react-query");
var CHUNK_SIZE = 100;
var useSearch = (mintAddresses, options = {}) => {
  var _a2, _b;
  const mintAddressesString = (0, import_react22.useMemo)(() => mintAddresses.join(","), [mintAddresses]);
  return (0, import_react_query9.useQuery)({
    queryKey: ["search", mintAddressesString],
    queryFn: () => __async(void 0, null, function* () {
      if (mintAddresses.length === 0) {
        return searchService.search("");
      }
      const chunks = [];
      for (let i = 0; i < mintAddresses.length; i += CHUNK_SIZE) {
        chunks.push(mintAddresses.slice(i, i + CHUNK_SIZE));
      }
      const chunkPromises = chunks.map(
        (chunk) => searchService.search(chunk.join(","))
      );
      const results = yield Promise.allSettled(chunkPromises);
      return results.flatMap((result) => result.status === "fulfilled" ? result.value : []);
    }),
    enabled: (_a2 = options.enabled) != null ? _a2 : true,
    staleTime: (_b = options.staleTime) != null ? _b : Infinity
  });
};

// src/components/FormPairSelector.tsx
var import_jsx_runtime39 = require("react/jsx-runtime");
var PAIR_ROW_HEIGHT2 = 72;
var SEARCH_BOX_HEIGHT = 56;
var rowRenderer = (0, import_react23.memo)((props) => {
  const { data, index, style } = props;
  const item = data.searchResult[index];
  return /* @__PURE__ */ (0, import_jsx_runtime39.jsx)(FormPairRow_default, { item, style, onSubmit: data.onSubmit }, item.address);
}, import_react_window.areEqual);
var FormPairSelector = ({ onSubmit, onClose }) => {
  const [search, setSearch] = (0, import_react23.useState)("");
  const { data: balances = {} } = useBalances();
  const { data: blueChipTokens = [] } = useSearch([]);
  const { data: searchTokens = [], isLoading } = useSearch([search], { staleTime: 5 * 60 * 1e3 });
  const { data: userBalanceTokens = [] } = useSearch([Object.keys(balances).join(",")]);
  const searchResult = (0, import_react23.useMemo)(() => {
    if (!search) {
      return sortByUserBalance([...userBalanceTokens, ...blueChipTokens], balances);
    } else {
      return sortByUserBalance(searchTokens, balances);
    }
  }, [blueChipTokens, balances, searchTokens, search, userBalanceTokens]);
  const triggerSearch = (0, import_react23.useCallback)(
    (0, import_lodash2.default)((value) => __async(void 0, null, function* () {
      setSearch(value);
    }), 200),
    [blueChipTokens]
  );
  const onChange = (0, import_react23.useCallback)(
    (e) => {
      setSearch("");
      triggerSearch(e.target.value);
    },
    [triggerSearch]
  );
  const listRef = (0, import_react23.createRef)();
  const inputRef = (0, import_react23.createRef)();
  (0, import_react23.useEffect)(() => {
    var _a2;
    return (_a2 = inputRef.current) == null ? void 0 : _a2.focus();
  }, [inputRef]);
  return /* @__PURE__ */ (0, import_jsx_runtime39.jsxs)("div", { className: "flex flex-col h-full w-full py-4 px-2 bg-black", children: [
    /* @__PURE__ */ (0, import_jsx_runtime39.jsxs)("div", { className: "flex w-full justify-between", children: [
      /* @__PURE__ */ (0, import_jsx_runtime39.jsx)("div", { className: "text-primary-text fill-current w-6 h-6 cursor-pointer", onClick: onClose, children: /* @__PURE__ */ (0, import_jsx_runtime39.jsx)(LeftArrowIcon_default, { width: 24, height: 24 }) }),
      /* @__PURE__ */ (0, import_jsx_runtime39.jsx)("div", { className: "text-primary-text", children: "Select Token" }),
      /* @__PURE__ */ (0, import_jsx_runtime39.jsx)("div", { className: " w-6 h-6" })
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime39.jsxs)(
      "div",
      {
        className: "flex px-5 mt-4 w-[98%] rounded-xl bg-module",
        style: { height: SEARCH_BOX_HEIGHT, maxHeight: SEARCH_BOX_HEIGHT },
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime39.jsx)(SearchIcon_default, {}),
          /* @__PURE__ */ (0, import_jsx_runtime39.jsx)(
            "input",
            {
              autoComplete: "off",
              className: "w-full rounded-xl ml-4 truncate bg-transparent text-primary-text placeholder:text-primary-text/50",
              placeholder: `Search`,
              onChange: (e) => onChange(e),
              ref: inputRef
            }
          )
        ]
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime39.jsxs)("div", { className: "mt-2", style: { flexGrow: 1 }, children: [
      searchResult && searchResult.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime39.jsx)(import_react_virtualized_auto_sizer.default, { children: ({ height, width }) => {
        return /* @__PURE__ */ (0, import_jsx_runtime39.jsx)(
          import_react_window.FixedSizeList,
          {
            ref: listRef,
            height,
            itemCount: searchResult.length,
            itemSize: PAIR_ROW_HEIGHT2,
            width: width - 2,
            itemData: {
              searchResult,
              onSubmit
            },
            className: cn("overflow-y-scroll mr-1 min-h-[12rem] px-5 webkit-scrollbar"),
            children: rowRenderer
          }
        );
      } }),
      isLoading ? /* @__PURE__ */ (0, import_jsx_runtime39.jsx)("div", { className: "mt-4 mb-4 text-center text-primary-text/50", children: /* @__PURE__ */ (0, import_jsx_runtime39.jsx)("span", { children: "Loading..." }) }) : searchResult && searchResult.length === 0 ? /* @__PURE__ */ (0, import_jsx_runtime39.jsx)("div", { className: "mt-4 mb-4 text-center text-primary-text/50", children: /* @__PURE__ */ (0, import_jsx_runtime39.jsx)("span", { children: "No tokens found" }) }) : /* @__PURE__ */ (0, import_jsx_runtime39.jsx)(import_jsx_runtime39.Fragment, {})
    ] })
  ] });
};
var FormPairSelector_default = FormPairSelector;

// src/components/screens/InitialScreen.tsx
var import_decimal9 = __toESM(require("decimal.js"));
var import_jsx_runtime40 = require("react/jsx-runtime");
var InitialScreen = () => {
  const { data: balances } = useBalances();
  const {
    form,
    setForm,
    setErrors,
    quoteResponseMeta,
    loading
  } = useSwapContext();
  const { setScreen } = useScreenState();
  const balance = (0, import_react24.useMemo)(() => {
    var _a2;
    if (!balances) return 0;
    return ((_a2 = balances[form.fromMint]) == null ? void 0 : _a2.uiAmount) || 0;
  }, [balances, form.fromMint]);
  const [isDisabled, setIsDisabled] = (0, import_react24.useState)(false);
  const { data: asset } = useAsset(form.fromMint);
  (0, import_react24.useEffect)(() => {
    if (!form.fromValue || !form.fromMint || !form.toMint || !form.toValue || !quoteResponseMeta || loading) {
      setErrors({});
      setIsDisabled(true);
      return;
    }
    if (new import_decimal9.default(form.fromValue).gt(balance)) {
      setErrors({
        fromValue: { title: `Insufficient ${asset == null ? void 0 : asset.symbol}`, message: "" }
      });
      setIsDisabled(true);
      return;
    }
    setErrors({});
    setIsDisabled(false);
  }, [form, asset, balance, quoteResponseMeta, loading, setErrors]);
  const [selectPairSelector, setSelectPairSelector] = (0, import_react24.useState)(null);
  const onSelectMint = (0, import_react24.useCallback)(
    (tokenInfo) => __async(void 0, null, function* () {
      if (selectPairSelector === "fromMint") {
        setForm((prev) => __spreadValues(__spreadProps(__spreadValues({}, prev), {
          fromMint: tokenInfo.id,
          fromValue: ""
        }), prev.toMint === tokenInfo.id ? { toMint: prev.fromMint } : void 0));
      } else {
        setForm((prev) => __spreadValues(__spreadProps(__spreadValues({}, prev), {
          toMint: tokenInfo.id,
          toValue: ""
        }), prev.fromMint === tokenInfo.id ? { fromMint: prev.toMint } : void 0));
      }
      setSelectPairSelector(null);
    }),
    [selectPairSelector, setForm]
  );
  const onSubmitToConfirmation = (0, import_react24.useCallback)(() => {
    setScreen("Swapping");
  }, [setScreen]);
  return /* @__PURE__ */ (0, import_jsx_runtime40.jsxs)(import_jsx_runtime40.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime40.jsx)(
      "form",
      {
        onSubmit: onSubmitToConfirmation,
        className: cn({
          hidden: Boolean(selectPairSelector)
        }),
        children: /* @__PURE__ */ (0, import_jsx_runtime40.jsx)(
          Form_default,
          {
            onSubmit: onSubmitToConfirmation,
            isDisabled,
            setSelectPairSelector
          }
        )
      }
    ),
    selectPairSelector !== null ? /* @__PURE__ */ (0, import_jsx_runtime40.jsx)("div", { className: "absolute top-0 left-0 h-full w-full bg-black rounded-lg overflow-hidden", children: /* @__PURE__ */ (0, import_jsx_runtime40.jsx)(
      FormPairSelector_default,
      {
        onSubmit: onSelectMint,
        onClose: () => setSelectPairSelector(null)
      }
    ) }) : null
  ] });
};
var InitialScreen_default = InitialScreen;

// src/components/screens/ReviewOrderScreen.tsx
var import_react25 = require("react");
var import_jsx_runtime41 = require("react/jsx-runtime");
var ConfirmationScreen = () => {
  const {
    fromTokenInfo,
    toTokenInfo,
    onSubmit: onSubmitJupiter,
    quoteResponseMeta,
    loading,
    refresh
  } = useSwapContext();
  const [hasExpired] = useTimeDiff_default();
  const { setScreen } = useScreenState();
  const onGoBack = () => {
    refresh();
    setScreen("Initial");
  };
  const onSubmit = (0, import_react25.useCallback)(() => __async(void 0, null, function* () {
    setScreen("Swapping");
    onSubmitJupiter();
  }), [onSubmitJupiter, setScreen]);
  return /* @__PURE__ */ (0, import_jsx_runtime41.jsxs)("div", { className: "flex flex-col h-full w-full py-4 px-2", children: [
    /* @__PURE__ */ (0, import_jsx_runtime41.jsxs)("div", { className: "flex w-full justify-between", children: [
      /* @__PURE__ */ (0, import_jsx_runtime41.jsx)("div", { className: "text-primary-text fill-current w-6 h-6 cursor-pointer", onClick: onGoBack, children: /* @__PURE__ */ (0, import_jsx_runtime41.jsx)(LeftArrowIcon_default, { width: 24, height: 24 }) }),
      /* @__PURE__ */ (0, import_jsx_runtime41.jsx)("div", { className: "text-primary-text", children: "Review Order" }),
      /* @__PURE__ */ (0, import_jsx_runtime41.jsx)("div", { className: " w-6 h-6" })
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime41.jsx)("div", { children: quoteResponseMeta && fromTokenInfo && toTokenInfo ? /* @__PURE__ */ (0, import_jsx_runtime41.jsx)(
      PriceInfo_default,
      {
        quoteResponse: quoteResponseMeta,
        fromTokenInfo,
        toTokenInfo,
        loading,
        containerClassName: "bg-modiu; border-none"
      }
    ) : null }),
    hasExpired ? /* @__PURE__ */ (0, import_jsx_runtime41.jsx)(JupButton_default, { size: "lg", className: "w-full mt-4 disabled:opacity-50 !p-0", onClick: onGoBack, children: /* @__PURE__ */ (0, import_jsx_runtime41.jsx)("span", { className: "text-sm", children: "Refresh" }) }) : /* @__PURE__ */ (0, import_jsx_runtime41.jsx)(
      JupButton_default,
      {
        size: "lg",
        className: cn(
          "w-full mt-4 disabled:opacity-50 !text-uiv2-text/75 leading-none !max-h-14 bg-gradient-to-r from-[#00BEF0] to-[#C7F284]"
        ),
        onClick: onSubmit,
        children: /* @__PURE__ */ (0, import_jsx_runtime41.jsx)("span", { children: "Confirm" })
      }
    )
  ] });
};
var ReviewOrderScreen_default = ConfirmationScreen;

// src/components/screens/SwappingScreen.tsx
var import_react26 = require("react");

// src/components/Spinner/index.tsx
var import_jsx_runtime42 = require("react/jsx-runtime");
var Spinner = ({
  className,
  baseColor = "currentColor",
  spinnerColor = "currentColor",
  width = 16,
  height = 16
}) => /* @__PURE__ */ (0, import_jsx_runtime42.jsx)("div", { className: `rounded-full flex items-center justify-center ${className}`, style: { width, height }, children: /* @__PURE__ */ (0, import_jsx_runtime42.jsxs)(
  "svg",
  {
    className: "animate-spin h-5 w-5 text-black dark:text-primary-text",
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 24 24",
    children: [
      /* @__PURE__ */ (0, import_jsx_runtime42.jsx)("circle", { className: "opacity-25", cx: "12", cy: "12", r: "10", stroke: baseColor, strokeWidth: "4" }),
      /* @__PURE__ */ (0, import_jsx_runtime42.jsx)(
        "path",
        {
          className: "opacity-75",
          fill: spinnerColor,
          d: "M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
        }
      )
    ]
  }
) });
var Spinner_default = Spinner;

// src/icons/SuccessIcon.tsx
var import_jsx_runtime43 = require("react/jsx-runtime");
var SuccessIcon = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime43.jsx)("svg", __spreadProps(__spreadValues({ xmlns: "http://www.w3.org/2000/svg", width: "256", height: "256", viewBox: "0 0 256 256" }, props), { children: /* @__PURE__ */ (0, import_jsx_runtime43.jsx)(
    "path",
    {
      fill: "currentColor",
      d: "M128 24a104 104 0 1 0 104 104A104.11 104.11 0 0 0 128 24m45.66 85.66l-56 56a8 8 0 0 1-11.32 0l-24-24a8 8 0 0 1 11.32-11.32L112 148.69l50.34-50.35a8 8 0 0 1 11.32 11.32"
    }
  ) }));
};
var SuccessIcon_default = SuccessIcon;

// src/icons/JupiterLogoV2.tsx
var import_jsx_runtime44 = require("react/jsx-runtime");
var JupiterLogoV2 = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)("svg", __spreadProps(__spreadValues({ xmlns: "http://www.w3.org/2000/svg", width: "128", height: "128", viewBox: "0 0 33 32", fill: "none" }, props), { children: [
    /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("g", { clipPath: "url(#clip0_11565_169621)", children: /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)("g", { filter: "url(#filter0_d_11565_169621)", children: [
      /* @__PURE__ */ (0, import_jsx_runtime44.jsx)(
        "path",
        {
          d: "M3.09074 25.1666C4.44267 27.0471 6.17683 28.6205 8.1795 29.7838C10.1822 30.947 12.4081 31.6738 14.7114 31.9165C13.5264 30.1333 11.8039 28.4928 9.65354 27.2438C7.50318 25.9948 5.22592 25.3125 3.09074 25.1666Z",
          fill: "url(#paint0_linear_11565_169621)"
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsx)(
        "path",
        {
          d: "M12.543 22.2705C8.40015 19.8636 3.91612 19.2502 0.707663 20.3338C1.0174 21.3575 1.42589 22.3487 1.92738 23.2934C4.71498 23.2288 7.75856 23.9859 10.5906 25.6308C13.4227 27.2757 15.5888 29.5459 16.9143 32C17.9839 31.9672 19.0479 31.8309 20.0913 31.5932C19.4426 28.2698 16.6849 24.6779 12.543 22.2705Z",
          fill: "url(#paint1_linear_11565_169621)"
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsx)(
        "path",
        {
          d: "M32.2852 12.5009C31.7585 10.3584 30.8054 8.34403 29.4829 6.57804C28.1604 4.81205 26.4956 3.33067 24.5879 2.22235C22.6802 1.11403 20.5687 0.401504 18.3796 0.127309C16.1904 -0.146885 13.9684 0.0228794 11.8463 0.626465C15.3915 1.06033 19.3267 2.39122 23.1859 4.63324C27.0452 6.87525 30.1533 9.63411 32.2852 12.5009Z",
          fill: "url(#paint2_linear_11565_169621)"
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsx)(
        "path",
        {
          d: "M27.1271 20.3583C25.3124 17.3446 22.2038 14.4588 18.3743 12.2342C14.5449 10.0095 10.4991 8.7388 6.98531 8.65474C3.894 8.58152 1.57389 9.48017 0.621548 11.1197C0.616125 11.1294 0.608532 11.1386 0.602566 11.1484C0.516877 11.4559 0.44312 11.7639 0.37587 12.0731C1.70568 11.5481 3.24645 11.2558 4.95969 11.2232C8.76959 11.1517 13.0334 12.3703 16.9681 14.6562C20.9027 16.9422 24.0759 20.0438 25.9003 23.3878C26.7182 24.8944 27.2285 26.3777 27.4308 27.7948C27.6662 27.5844 27.8972 27.3669 28.1212 27.1408C28.1272 27.1305 28.131 27.1196 28.1369 27.1088C29.0893 25.4677 28.721 23.0076 27.1271 20.3583Z",
          fill: "url(#paint3_linear_11565_169621)"
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsx)(
        "path",
        {
          d: "M15.4609 17.2485C9.59662 13.8416 3.11626 13.3079 0 15.6855C0.00612096 16.4297 0.0630166 17.1726 0.170292 17.9091C1.08699 17.6312 2.03177 17.4562 2.98718 17.3874C6.46952 17.1254 10.3087 18.0957 13.7927 20.1207C17.2766 22.1458 20.023 25.0018 21.5209 28.1543C21.935 29.018 22.2508 29.9254 22.4624 30.8595C23.1555 30.5878 23.8294 30.2694 24.4794 29.9066C25.0011 26.0213 21.3268 20.656 15.4609 17.2485Z",
          fill: "url(#paint4_linear_11565_169621)"
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsx)(
        "path",
        {
          d: "M30.1434 15.3141C28.3082 12.3036 25.1724 9.40969 21.3158 7.17039C17.4593 4.93109 13.3977 3.64033 9.87257 3.53674C7.1853 3.45919 5.10382 4.11053 4.02457 5.34109C8.50588 4.58182 14.4168 5.85794 20.146 9.18625C25.8753 12.5146 29.9135 17.0181 31.4722 21.2868C32.0064 19.7406 31.5416 17.6098 30.1434 15.3141Z",
          fill: "url(#paint5_linear_11565_169621)"
        }
      )
    ] }) }),
    /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)("defs", { children: [
      /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)(
        "filter",
        {
          id: "filter0_d_11565_169621",
          x: "-22.7449",
          y: "-20.4704",
          width: "77.7749",
          height: "77.4898",
          filterUnits: "userSpaceOnUse",
          colorInterpolationFilters: "sRGB",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("feFlood", { floodOpacity: "0", result: "BackgroundImageFix" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)(
              "feColorMatrix",
              {
                in: "SourceAlpha",
                type: "matrix",
                values: "0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0",
                result: "hardAlpha"
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("feOffset", { dy: "2.27449" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("feGaussianBlur", { stdDeviation: "11.3724" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("feComposite", { in2: "hardAlpha", operator: "out" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("feColorMatrix", { type: "matrix", values: "0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.1 0" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("feBlend", { mode: "normal", in2: "BackgroundImageFix", result: "effect1_dropShadow_11565_169621" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("feBlend", { mode: "normal", in: "SourceGraphic", in2: "effect1_dropShadow_11565_169621", result: "shape" })
          ]
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)(
        "linearGradient",
        {
          id: "paint0_linear_11565_169621",
          x1: "21.5",
          y1: "6.5",
          x2: "6.66667",
          y2: "32",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "0.0001", stopColor: "currentColor" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "1", stopColor: "currentColor" })
          ]
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)(
        "linearGradient",
        {
          id: "paint1_linear_11565_169621",
          x1: "21.5",
          y1: "6.5",
          x2: "6.66667",
          y2: "32",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "0.0001", stopColor: "currentColor" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "1", stopColor: "currentColor" })
          ]
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)(
        "linearGradient",
        {
          id: "paint2_linear_11565_169621",
          x1: "21.5",
          y1: "6.5",
          x2: "6.66667",
          y2: "32",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "0.0001", stopColor: "currentColor" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "1", stopColor: "currentColor" })
          ]
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)(
        "linearGradient",
        {
          id: "paint3_linear_11565_169621",
          x1: "21.5",
          y1: "6.5",
          x2: "6.66667",
          y2: "32",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "0.0001", stopColor: "currentColor" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "1", stopColor: "currentColor" })
          ]
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)(
        "linearGradient",
        {
          id: "paint4_linear_11565_169621",
          x1: "21.5",
          y1: "6.5",
          x2: "6.66667",
          y2: "32",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "0.0001", stopColor: "currentColor" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "1", stopColor: "currentColor" })
          ]
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsxs)(
        "linearGradient",
        {
          id: "paint5_linear_11565_169621",
          x1: "21.5",
          y1: "6.5",
          x2: "6.66667",
          y2: "32",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "0.0001", stopColor: "currentColor" }),
            /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("stop", { offset: "1", stopColor: "currentColor" })
          ]
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("clipPath", { id: "clip0_11565_169621", children: /* @__PURE__ */ (0, import_jsx_runtime44.jsx)("rect", { width: "32.2852", height: "32", fill: "white" }) })
    ] })
  ] }));
};
var JupiterLogoV2_default = JupiterLogoV2;

// src/components/screens/SwappingScreen.tsx
var import_jsx_runtime45 = require("react/jsx-runtime");
var ErrorIcon = () => {
  return /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("svg", { width: "40", height: "40", viewBox: "0 0 40 40", fill: "none", xmlns: "http://www.w3.org/2000/svg", children: [
    /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("g", { clipPath: "url(#clip0_7547_116874)", children: [
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("circle", { cx: "20", cy: "20", r: "20", fill: "#F04A44" }),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(
        "path",
        {
          d: "M19.8444 25.4321C18.6773 25.4332 17.7205 24.5092 17.6793 23.3431L17.1718 9.04107C17.1507 8.45326 17.3706 7.88344 17.7786 7.46056C18.1867 7.03768 18.7492 6.7998 19.337 6.7998H20.3519C20.9397 6.7998 21.5021 7.03768 21.9102 7.46056C22.3183 7.88344 22.5382 8.45329 22.5171 9.04107L22.0096 23.3431C21.9684 24.5092 21.0116 25.4332 19.8444 25.4321Z",
          fill: "white"
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(
        "path",
        {
          d: "M22.8893 30.4989C22.8893 32.1809 21.5266 33.5436 19.8446 33.5436C18.1626 33.5436 16.7998 32.1809 16.7998 30.4989C16.7998 28.8169 18.1626 27.4541 19.8446 27.4541C21.5266 27.4541 22.8893 28.8169 22.8893 30.4989Z",
          fill: "white"
        }
      )
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("defs", { children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("clipPath", { id: "clip0_7547_116874", children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("rect", { width: "40", height: "40", fill: "white" }) }) })
  ] });
};
var SwappingScreen = () => {
  const {
    displayMode,
    lastSwapResult,
    reset,
    swapping: { txStatus },
    fromTokenInfo,
    toTokenInfo,
    refresh
  } = useSwapContext();
  const { screen: screen2, setScreen } = useScreenState();
  const [errorMessage, setErrorMessage] = (0, import_react26.useState)("");
  const onSwapMore = () => {
    reset();
    setErrorMessage("");
    setScreen("Initial");
    refresh();
  };
  const onGoBack = () => {
    reset({ resetValues: false });
    setErrorMessage("");
    setScreen("Initial");
    refresh();
  };
  (0, import_react26.useEffect)(() => {
    var _a2, _b, _c, _d;
    if (screen2 !== "Swapping") return;
    if ((lastSwapResult == null ? void 0 : lastSwapResult.swapResult) && "error" in (lastSwapResult == null ? void 0 : lastSwapResult.swapResult)) {
      setErrorMessage(((_b = (_a2 = lastSwapResult == null ? void 0 : lastSwapResult.swapResult) == null ? void 0 : _a2.error) == null ? void 0 : _b.message) || "");
      if (window.Jupiter.onSwapError) {
        window.Jupiter.onSwapError({
          error: (_c = lastSwapResult == null ? void 0 : lastSwapResult.swapResult) == null ? void 0 : _c.error,
          quoteResponseMeta: lastSwapResult == null ? void 0 : lastSwapResult.quoteReponse
        });
      }
      return;
    } else if ((lastSwapResult == null ? void 0 : lastSwapResult.swapResult) && "txid" in (lastSwapResult == null ? void 0 : lastSwapResult.swapResult)) {
      if (window.Jupiter.onSuccess) {
        window.Jupiter.onSuccess({
          txid: (_d = lastSwapResult == null ? void 0 : lastSwapResult.swapResult) == null ? void 0 : _d.txid,
          swapResult: lastSwapResult == null ? void 0 : lastSwapResult.swapResult,
          quoteResponseMeta: lastSwapResult == null ? void 0 : lastSwapResult.quoteReponse
        });
      }
      return;
    }
  }, [lastSwapResult, screen2]);
  const onClose = () => {
    if (!displayMode || displayMode === "modal") {
      window.Jupiter.close();
    }
    reset();
    setScreen("Initial");
  };
  const { explorer, getExplorer } = usePreferredExplorer();
  const isLoading = (txStatus == null ? void 0 : txStatus.status) === "loading" || (txStatus == null ? void 0 : txStatus.status) === "pending-approval" || (txStatus == null ? void 0 : txStatus.status) === "sending";
  const Content2 = () => {
    return /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)(import_jsx_runtime45.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "flex w-full justify-center", children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "text-primary-text", children: "Swapping" }) }),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "flex w-full justify-center items-center mt-9", children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "h-16 w-16 animate-hue duration-100", children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(JupiterLogoV2_default, { width: 64, height: 64 }) }) }),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "flex flex-col w-full justify-center items-center px-5 mt-7", children: isLoading && /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: "flex items-center w-full rounded-xl p-4 bg-module mb-2", children: [
        /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(Spinner_default, { spinnerColor: "white" }),
        /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "ml-4 flex w-full justify-between", children: /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("span", { className: "text-primary-text text-sm", children: [
          txStatus.status === "loading" && "Preparing transactions",
          txStatus.status === "pending-approval" && "Pending Approval",
          txStatus.status === "sending" && "Swapping"
        ] }) })
      ] }) })
    ] });
  };
  const SuccessContent = () => {
    const { inputAmount, outputAmount, explorerLink } = (0, import_react26.useMemo)(() => {
      return {
        inputAmount: (lastSwapResult == null ? void 0 : lastSwapResult.swapResult) && "inputAmount" in (lastSwapResult == null ? void 0 : lastSwapResult.swapResult) ? lastSwapResult == null ? void 0 : lastSwapResult.swapResult.inputAmount : 0,
        outputAmount: (lastSwapResult == null ? void 0 : lastSwapResult.swapResult) && "outputAmount" in (lastSwapResult == null ? void 0 : lastSwapResult.swapResult) ? lastSwapResult == null ? void 0 : lastSwapResult.swapResult.outputAmount : 0,
        explorerLink: (lastSwapResult == null ? void 0 : lastSwapResult.swapResult) && "txid" in (lastSwapResult == null ? void 0 : lastSwapResult.swapResult) ? getExplorer(lastSwapResult == null ? void 0 : lastSwapResult.swapResult.txid) : ""
      };
    }, []);
    if (!fromTokenInfo || !toTokenInfo || !(lastSwapResult == null ? void 0 : lastSwapResult.quoteReponse)) {
      return null;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)(import_jsx_runtime45.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "flex justify-center mt-8", children: /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: " flex justify-center relative  items-center", children: [
        /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "bg-success bg-opacity-[15%]  animate-pulse  h-[60px] w-[60px] rounded-full" }),
        /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "rounded-full absolute  justify-center", children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(SuccessIcon_default, { className: "text-success", height: 56, width: 56 }) })
      ] }) }),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: "flex flex-col justify-center items-center gap-y-2", children: [
        /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: "mt-2 flex flex-col items-center justify-center text-center px-4", children: [
          /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("p", { className: "text-xs font-semibold text-primary-text", children: [
            "Swapped ",
            readableValue(inputAmount, fromTokenInfo.decimals),
            " ",
            fromTokenInfo.symbol,
            " to"
          ] }),
          /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("p", { className: "text-2xl font-semibold text-primary-text", children: [
            readableValue(outputAmount, toTokenInfo.decimals),
            " ",
            toTokenInfo.symbol
          ] })
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: " bg-module rounded-xl overflow-y-auto w-full webkit-scrollbar py-3 max-h-[260px]  px-3", children: [
          /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(
            PriceInfo_default,
            {
              quoteResponse: lastSwapResult == null ? void 0 : lastSwapResult.quoteReponse,
              fromTokenInfo,
              toTokenInfo,
              loading: false,
              containerClassName: " border-none mt-0"
            }
          ),
          explorerLink && /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: "flex items-center justify-between text-xs text-primary-text/50  mt-4", children: [
            /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("span", { children: "Transaction" }) }),
            /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)(
              "a",
              {
                href: explorerLink,
                target: "_blank",
                rel: "noopener noreferrer",
                className: "cursor-pointer text-primary-text ml-2 text-xs hover:underline",
                children: [
                  "View on ",
                  explorer
                ]
              }
            )
          ] })
        ] })
      ] }),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: "pb-4 flex space-x-2", children: [
        /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(
          JupButton_default,
          {
            size: "lg",
            className: "w-full mt-4 disabled:opacity-50 !text-uiv2-text/75 leading-none !max-h-14 bg-primary",
            onClick: onSwapMore,
            children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("span", { children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("span", { className: "text-sm", children: "Swap More" }) })
          }
        ),
        displayMode !== "integrated" ? /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(
          JupButton_default,
          {
            size: "lg",
            className: "w-full mt-4 disabled:opacity-50 leading-none !max-h-14 text-primary-text bg-interactive",
            onClick: onClose,
            children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("span", { className: "text-sm", children: "Close" })
          }
        ) : null
      ] })
    ] });
  };
  return /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: "flex flex-col h-full w-full px-2", children: [
    errorMessage || (txStatus == null ? void 0 : txStatus.status) === "fail" ? /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "", children: /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: "flex flex-col items-center justify-center text-center mt-12", children: [
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(ErrorIcon, {}),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("p", { className: "text-primary-text mt-2", children: "Swap Failed" }),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("p", { className: "text-primary-text/50 text-xs mt-2", children: "We were unable to complete the swap, please try again." }),
      errorMessage ? /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("p", { className: "text-primary-text/50 text-xs mt-2 break-all", children: errorMessage }) : "",
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(
        JupButton_default,
        {
          size: "lg",
          className: "w-full mt-6 disabled:opacity-50 !text-uiv2-text/75 leading-none !max-h-14 bg-primary",
          onClick: onGoBack,
          children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("span", { children: "Retry" })
        }
      )
    ] }) }) : null,
    !errorMessage && (txStatus == null ? void 0 : txStatus.status) === "timeout" ? /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("div", { className: "flex justify-center", children: /* @__PURE__ */ (0, import_jsx_runtime45.jsxs)("div", { className: "flex flex-col items-center justify-center text-center mt-12", children: [
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(ErrorIcon, {}),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("p", { className: "text-primary-text mt-2", children: "Transaction timed-out" }),
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("p", { className: "text-primary-text/50 text-xs mt-2", children: "We were unable to complete the swap, please try again." }),
      errorMessage ? /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("p", { className: "text-primary-text/50 text-xs mt-2", children: errorMessage }) : "",
      /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(
        JupButton_default,
        {
          size: "lg",
          className: "w-full mt-6 disabled:opacity-50 !text-uiv2-text/75 leading-none !max-h-14 bg-primary",
          onClick: onGoBack,
          children: /* @__PURE__ */ (0, import_jsx_runtime45.jsx)("span", { children: "Retry" })
        }
      )
    ] }) }) : null,
    !errorMessage && isLoading ? /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(Content2, {}) : null,
    !errorMessage && (txStatus == null ? void 0 : txStatus.status) === "success" ? /* @__PURE__ */ (0, import_jsx_runtime45.jsx)(SuccessContent, {}) : null
  ] });
};
var SwappingScreen_default = SwappingScreen;

// src/components/screens/WalletScreen.tsx
var import_wallet_adapter3 = require("@jup-ag/wallet-adapter");
var import_react27 = require("react");

// src/icons/CloseIcon.tsx
var import_jsx_runtime46 = require("react/jsx-runtime");
var CloseIcon = ({ width = 20, height = 20 }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime46.jsx)("svg", { width, height, viewBox: "0 0 20 21", fill: "none", xmlns: "http://www.w3.org/2000/svg", children: /* @__PURE__ */ (0, import_jsx_runtime46.jsx)(
    "path",
    {
      d: "M2.0336 16.2126L8.2336 10.0126L2.0336 3.81263C1.7961 3.57903 1.66172 3.25951 1.66016 2.92669C1.65938 2.59309 1.79141 2.27357 2.02734 2.03763C2.26328 1.80247 2.5828 1.67045 2.9164 1.67201C3.25 1.67357 3.56874 1.80795 3.80234 2.04623L9.99994 8.24623L16.1999 2.04623C16.4335 1.80795 16.7523 1.67357 17.0859 1.67201C17.4187 1.67045 17.739 1.80248 17.9749 2.03763C18.2109 2.27357 18.3429 2.59309 18.3413 2.92669C18.3406 3.25951 18.2062 3.57903 17.9687 3.81263L11.7663 10.0126L17.9663 16.2126C18.2038 16.4462 18.3382 16.7658 18.3397 17.0986C18.3405 17.4322 18.2085 17.7517 17.9725 17.9876C17.7366 18.2228 17.4171 18.3548 17.0835 18.3533C16.7499 18.3517 16.4311 18.2173 16.1975 17.979L9.99994 11.779L3.79994 17.979C3.31088 18.4611 2.52494 18.4579 2.039 17.9736C1.55384 17.4884 1.54994 16.7025 2.03119 16.2126L2.0336 16.2126Z",
      fill: "currentColor"
    }
  ) });
};
var CloseIcon_default = CloseIcon;

// src/components/screens/WalletScreen.tsx
var import_jsx_runtime47 = require("react/jsx-runtime");
var WalletScreen = () => {
  const { setScreen } = useScreenState();
  const { handleConnectClick } = (0, import_wallet_adapter3.useUnifiedWalletContext)();
  const { wallets } = useWalletPassThrough();
  const handleConnect = (event, wallet) => __async(void 0, null, function* () {
    try {
      yield handleConnectClick(event, wallet.adapter);
    } catch (error) {
      console.error("Failed to connect wallet:", error);
    } finally {
      setScreen("Initial");
    }
  });
  const numberOfWallets = (0, import_react27.useMemo)(() => {
    return wallets == null ? void 0 : wallets.length;
  }, [wallets]);
  const hasWallet = (0, import_react27.useMemo)(() => {
    return numberOfWallets > 0;
  }, [numberOfWallets]);
  return /* @__PURE__ */ (0, import_jsx_runtime47.jsxs)("div", { className: "text-primary-text p-2", children: [
    /* @__PURE__ */ (0, import_jsx_runtime47.jsxs)("div", { className: "flex justify-between items-center mb-2 h-[28px] ", children: [
      /* @__PURE__ */ (0, import_jsx_runtime47.jsxs)("h1", { className: "text-sm font-bold", children: [
        "Available Wallets (",
        numberOfWallets,
        ")"
      ] }),
      /* @__PURE__ */ (0, import_jsx_runtime47.jsx)(
        "button",
        {
          onClick: () => setScreen("Initial"),
          className: "text-primary-text hover:text-primary-text/50 transition-colors bg-interactive rounded-full p-1",
          children: /* @__PURE__ */ (0, import_jsx_runtime47.jsx)(CloseIcon_default, { width: 15, height: 15 })
        }
      )
    ] }),
    !hasWallet && /* @__PURE__ */ (0, import_jsx_runtime47.jsx)("div", { className: "text-sm font-medium text-primary-text flex items-center justify-center h-full py-4", children: "No wallet found" }),
    /* @__PURE__ */ (0, import_jsx_runtime47.jsx)("div", { className: "grid grid-cols-2 gap-2", children: wallets == null ? void 0 : wallets.map((wallet) => /* @__PURE__ */ (0, import_jsx_runtime47.jsx)(
      "button",
      {
        onClick: (e) => handleConnect(e, wallet),
        className: `w-full text-left bg-interactive p-4 rounded-lg transition-colors ${"hover:bg-interactive/80 cursor-pointer"}`,
        children: /* @__PURE__ */ (0, import_jsx_runtime47.jsx)("div", { className: "flex items-center justify-between", children: /* @__PURE__ */ (0, import_jsx_runtime47.jsxs)("div", { className: "flex items-center gap-3", children: [
          wallet.adapter.icon && /* @__PURE__ */ (0, import_jsx_runtime47.jsx)("img", { src: wallet.adapter.icon, alt: `${wallet.adapter.name} icon`, className: "w-8 h-8" }),
          /* @__PURE__ */ (0, import_jsx_runtime47.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime47.jsx)("h2", { className: "text-sm font-medium", children: wallet.adapter.name }) })
        ] }) })
      },
      wallet.adapter.name
    )) })
  ] });
};
var WalletScreen_default = WalletScreen;

// src/components/Jupiter.tsx
var import_jsx_runtime48 = require("react/jsx-runtime");
var Content = () => {
  const { screen: screen2 } = useScreenState();
  return /* @__PURE__ */ (0, import_jsx_runtime48.jsxs)("div", { id: "jupiter-terminal", className: " h-full bg-background relative flex flex-col justify-between", children: [
    /* @__PURE__ */ (0, import_jsx_runtime48.jsxs)("div", { children: [
      screen2 === "Initial" ? /* @__PURE__ */ (0, import_jsx_runtime48.jsxs)(import_jsx_runtime48.Fragment, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime48.jsx)(Header_default, {}),
        /* @__PURE__ */ (0, import_jsx_runtime48.jsx)(InitialScreen_default, {})
      ] }) : null,
      screen2 === "Confirmation" ? /* @__PURE__ */ (0, import_jsx_runtime48.jsx)(ReviewOrderScreen_default, {}) : null,
      screen2 === "Swapping" ? /* @__PURE__ */ (0, import_jsx_runtime48.jsx)(SwappingScreen_default, {}) : null,
      screen2 === "Wallet" ? /* @__PURE__ */ (0, import_jsx_runtime48.jsx)(WalletScreen_default, {}) : null
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime48.jsxs)("span", { className: "text-primary-text/50 text-xs p-2 flex-row flex gap-1  justify-center", children: [
      "Powered by",
      /* @__PURE__ */ (0, import_jsx_runtime48.jsxs)("a", { href: "https://jup.ag", target: "_blank", rel: "noreferrer noopener", className: "flex items-center gap-1 ", children: [
        /* @__PURE__ */ (0, import_jsx_runtime48.jsx)(JupiterLogoV2_default, { className: "text-primary-text/50", width: 15, height: 15 }),
        "Jupiter"
      ] })
    ] })
  ] });
};
var JupiterApp = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime48.jsx)(SwapContextProvider, __spreadProps(__spreadValues({}, props), { children: /* @__PURE__ */ (0, import_jsx_runtime48.jsx)(BrandingProvider, __spreadProps(__spreadValues({}, props), { children: /* @__PURE__ */ (0, import_jsx_runtime48.jsx)(Content, {}) })) }));
};
var Jupiter_default = JupiterApp;

// src/contexts/ContextProvider.tsx
var import_wallet_adapter4 = require("@jup-ag/wallet-adapter");
var import_react28 = __toESM(require("react"));
var import_react29 = require("react");
var import_jsx_runtime49 = require("react/jsx-runtime");
var noop = () => {
};
var WalletContextProvider = ({
  autoConnect,
  children
}) => {
  const enableWalletPassthrough = (() => {
    if (typeof window === "undefined") return void 0;
    return window.Jupiter.enableWalletPassthrough;
  })();
  const wallets = (0, import_react29.useMemo)(() => {
    if (enableWalletPassthrough) {
      return [];
    }
    return [];
  }, [enableWalletPassthrough]);
  const [showWalletStatus, setShowWalletStatus] = (0, import_react28.useState)({
    show: false,
    message: ""
  });
  const ShouldWrapWalletProvider = (0, import_react29.useMemo)(() => {
    return enableWalletPassthrough ? import_react28.default.Fragment : ({ children: children2 }) => /* @__PURE__ */ (0, import_jsx_runtime49.jsx)(
      import_wallet_adapter4.UnifiedWalletProvider,
      {
        wallets,
        config: {
          env: "mainnet-beta",
          autoConnect: typeof autoConnect !== "undefined" ? autoConnect : true,
          metadata: {
            name: "Jupiter Terminal",
            url: "https://terminal.jup.ag",
            description: "An open-sourced, lite version of Jupiter that provides end-to-end swap flow by linking it in your HTML. Check out the visual demo for the various integration modes below.          ",
            iconUrls: []
          },
          hardcodedWallets: [],
          walletPrecedence: [],
          notificationCallback: {
            onConnect: noop,
            onConnecting: noop,
            onDisconnect: noop,
            onNotInstalled: ({ walletName, metadata }) => {
              setShowWalletStatus({
                show: true,
                message: /* @__PURE__ */ (0, import_jsx_runtime49.jsxs)("p", { className: "space-y-1", children: [
                  walletName,
                  " is not installed.",
                  /* @__PURE__ */ (0, import_jsx_runtime49.jsxs)("p", { className: "space-x-1", children: [
                    /* @__PURE__ */ (0, import_jsx_runtime49.jsxs)(
                      "a",
                      {
                        className: "underline font-semibold",
                        target: "_blank",
                        rel: "noopener noreferrer",
                        href: metadata.url,
                        children: [
                          "Visit ",
                          walletName,
                          " website"
                        ]
                      }
                    ),
                    /* @__PURE__ */ (0, import_jsx_runtime49.jsx)("span", { children: "to install it." })
                  ] })
                ] })
              });
              setTimeout(() => {
                setShowWalletStatus({
                  show: false,
                  message: ""
                });
              }, 5e3);
            }
          },
          theme: "jupiter"
        },
        children: children2
      }
    );
  }, [autoConnect, enableWalletPassthrough, wallets]);
  return /* @__PURE__ */ (0, import_jsx_runtime49.jsxs)(import_jsx_runtime49.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime49.jsx)(ShouldWrapWalletProvider, { children }),
    showWalletStatus.show && showWalletStatus.message ? /* @__PURE__ */ (0, import_jsx_runtime49.jsx)("div", { className: "absolute bottom-2 w-full px-2", children: /* @__PURE__ */ (0, import_jsx_runtime49.jsx)("div", { className: "w-full h-full bg-white/10 rounded-lg p-2 text-warning text-xs", children: showWalletStatus.message }) }) : null
  ] });
};
var ContextProvider = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime49.jsx)(import_jsx_runtime49.Fragment, { children: /* @__PURE__ */ (0, import_jsx_runtime49.jsx)(WalletContextProvider, __spreadProps(__spreadValues({}, props), { children: /* @__PURE__ */ (0, import_jsx_runtime49.jsx)(PreferredExplorerProvider, { defaultExplorer: props.defaultExplorer, children: props.children }) })) });
};

// src/index.tsx
var import_react_query10 = require("@tanstack/react-query");
var import_react30 = require("react");
var import_jsx_runtime50 = require("react/jsx-runtime");
var App = () => {
  const queryClient = (0, import_react30.useMemo)(() => new import_react_query10.QueryClient(), []);
  const [props] = (0, import_jotai4.useAtom)(appProps);
  if (!props) return null;
  return /* @__PURE__ */ (0, import_jsx_runtime50.jsx)(import_react_query10.QueryClientProvider, { client: queryClient, children: /* @__PURE__ */ (0, import_jsx_runtime50.jsx)(ContextProvider, __spreadProps(__spreadValues({}, props), { children: /* @__PURE__ */ (0, import_jsx_runtime50.jsx)(WalletPassthroughProvider_default, { children: /* @__PURE__ */ (0, import_jsx_runtime50.jsx)(ScreenProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime50.jsx)(Jupiter_default, __spreadValues({}, props)) }) }) })) });
};
var RenderJupiter = () => {
  return /* @__PURE__ */ (0, import_jsx_runtime50.jsx)(import_jotai4.Provider, { store: typeof window !== "undefined" ? window.Jupiter.store : void 0, children: /* @__PURE__ */ (0, import_jsx_runtime50.jsx)(App, {}) });
};

// src/npmlibary.tsx
function init2(props) {
  return __async(this, null, function* () {
    window.Jupiter = { init: init2, resume, close, appProps, syncProps };
    window.JupiterRenderer = {
      RenderJupiter
    };
    yield init(props);
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  appProps,
  close,
  init,
  resume,
  syncProps
});
